--- a/nss_data_plane/nss_data_plane_gmac.c
+++ b/nss_data_plane/nss_data_plane_gmac.c
@@ -20,7 +20,7 @@
 #include "nss_tx_rx_common.h"
 #include <nss_gmac_api_if.h>
 
-#define NSS_DP_GMAC_SUPPORTED_FEATURES (NETIF_F_HIGHDMA | NETIF_F_HW_CSUM | NETIF_F_RXCSUM | NETIF_F_SG | NETIF_F_FRAGLIST | (NETIF_F_TSO | NETIF_F_TSO6 | NETIF_F_UFO))
+#define NSS_DP_GMAC_SUPPORTED_FEATURES (NETIF_F_HIGHDMA | NETIF_F_HW_CSUM | NETIF_F_RXCSUM | NETIF_F_SG | NETIF_F_FRAGLIST | (NETIF_F_TSO | NETIF_F_TSO6))
 #define NSS_DATA_PLANE_GMAC_MAX_INTERFACES 4
 
 static DEFINE_SPINLOCK(nss_data_plane_gmac_stats_lock);
--- a/nss_hal/fsm9010/nss_hal_pvt.c
+++ b/nss_hal/fsm9010/nss_hal_pvt.c
@@ -145,7 +145,7 @@ static struct nss_platform_data *__nss_h
 	npd->nphys = res_nphys.start;
 	npd->vphys = res_vphys.start;
 
-	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+	npd->nmap = nss_ioremap(npd->nphys, resource_size(&res_nphys));
 	if (!npd->nmap) {
 		nss_info_always("%px: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -291,7 +291,7 @@ static int __nss_hal_request_irq(struct
 	}
 
 	int_ctx->irq = npd->irq[irq_num];
-	netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi, 64);
+	netif_napi_add_weight(NSS_CTX_NAPI_NDEV(nss_ctx), &int_ctx->napi, nss_core_handle_napi, 64);
 	return 0;
 }
 
--- a/nss_hal/ipq50xx/nss_hal_pvt.c
+++ b/nss_hal/ipq50xx/nss_hal_pvt.c
@@ -185,13 +185,13 @@ static struct nss_platform_data *__nss_h
 	npd->nphys = res_nphys.start;
 	npd->qgic_phys = res_qgic_phys.start;
 
-	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+	npd->nmap = nss_ioremap(npd->nphys, resource_size(&res_nphys));
 	if (!npd->nmap) {
 		nss_info_always("%px: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
 	}
 
-	npd->qgic_map = ioremap_nocache(npd->qgic_phys, resource_size(&res_qgic_phys));
+	npd->qgic_map = nss_ioremap(npd->qgic_phys, resource_size(&res_qgic_phys));
 	if (!npd->qgic_map) {
 		nss_info_always("%px: nss%d: ioremap() fail for qgic map\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -349,7 +349,7 @@ static int __nss_hal_common_reset(struct
 
 	of_node_put(cmn);
 
-	nss_misc_reset = ioremap_nocache(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
+	nss_misc_reset = nss_ioremap(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
 	if (!nss_misc_reset) {
 		pr_err("%px: ioremap fail for nss_misc_reset\n", nss_dev);
 		return -EFAULT;
@@ -599,7 +599,7 @@ static int __nss_hal_request_irq(struct
 		return err;
 	}
 
-	netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, napi_poll_cb, napi_wgt);
+	netif_napi_add_weight(NSS_CTX_NAPI_NDEV(nss_ctx), &int_ctx->napi, napi_poll_cb, napi_wgt);
 	int_ctx->cause = cause;
 	err = request_irq(irq, nss_hal_handle_irq, 0, irq_name, int_ctx);
 	if (err) {
--- a/nss_hal/ipq60xx/nss_hal_pvt.c
+++ b/nss_hal/ipq60xx/nss_hal_pvt.c
@@ -64,6 +64,20 @@
 #define NSS_UTCM_CLK "nss-utcm-clk"
 
 /*
+ * N2H interrupts
+ */
+#define NSS_IRQ_NAME_EMPTY_BUF_SOS "nss_empty_buf_sos"
+#define NSS_IRQ_NAME_EMPTY_BUF_QUEUE "nss_empty_buf_queue"
+#define NSS_IRQ_NAME_TX_UNBLOCK "nss-tx-unblock"
+#define NSS_IRQ_NAME_QUEUE0 "nss_queue0"
+#define NSS_IRQ_NAME_QUEUE1 "nss_queue1"
+#define NSS_IRQ_NAME_QUEUE2 "nss_queue2"
+#define NSS_IRQ_NAME_QUEUE3 "nss_queue3"
+#define NSS_IRQ_NAME_COREDUMP_COMPLETE "nss_coredump_complete"
+#define NSS_IRQ_NAME_PAGED_EMPTY_BUF_SOS "nss_paged_empty_buf_sos"
+#define NSS_IRQ_NAME_PROFILE_DMA "nss_profile_dma"
+
+/*
  * Voltage values
  */
 #define NOMINAL_VOLTAGE 1
@@ -208,13 +222,13 @@ static struct nss_platform_data *__nss_h
 	npd->nphys = res_nphys.start;
 	npd->qgic_phys = res_qgic_phys.start;
 
-	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+	npd->nmap = nss_ioremap(npd->nphys, resource_size(&res_nphys));
 	if (!npd->nmap) {
 		nss_info_always("%px: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
 	}
 
-	npd->qgic_map = ioremap_nocache(npd->qgic_phys, resource_size(&res_qgic_phys));
+	npd->qgic_map = nss_ioremap(npd->qgic_phys, resource_size(&res_qgic_phys));
 	if (!npd->qgic_map) {
 		nss_info_always("%px: nss%d: ioremap() fail for qgic map\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -434,13 +448,13 @@ static int __nss_hal_common_reset(struct
 
 	of_node_put(cmn);
 
-	nss_misc_reset = ioremap_nocache(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
+	nss_misc_reset = nss_ioremap(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
 	if (!nss_misc_reset) {
 		pr_err("%px: ioremap fail for nss_misc_reset\n", nss_dev);
 		return -EFAULT;
 	}
 
-	nss_misc_reset_flag = ioremap_nocache(res_nss_misc_reset_flag.start, resource_size(&res_nss_misc_reset_flag));
+	nss_misc_reset_flag = nss_ioremap(res_nss_misc_reset_flag.start, resource_size(&res_nss_misc_reset_flag));
 	if (!nss_misc_reset_flag) {
 		pr_err("%px: ioremap fail for nss_misc_reset_flag\n", nss_dev);
 		return -EFAULT;
@@ -610,71 +624,94 @@ static void __nss_hal_send_interrupt(str
 static int __nss_hal_request_irq(struct nss_ctx_instance *nss_ctx, struct nss_platform_data *npd, int irq_num)
 {
 	struct int_ctx_instance *int_ctx = &nss_ctx->int_ctx[irq_num];
+	uint32_t cause, napi_wgt;
 	int err = -1, irq = npd->irq[irq_num];
+	int (*napi_poll_cb)(struct napi_struct *, int) = NULL;
+	const char *irq_name;
 
 	irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
 
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_SOS) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_non_queue, NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_EMPTY_BUFFERS_SOS;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_empty_buf_sos", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_QUEUE) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_EMPTY_BUFFER_RETURN_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_EMPTY_BUFFER_QUEUE;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_empty_buf_queue", int_ctx);
-	}
+	switch (irq_num) {
+	case NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_SOS:
+		napi_poll_cb = nss_core_handle_napi_non_queue;
+		napi_wgt = NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_EMPTY_BUFFERS_SOS;
+		irq_name = NSS_IRQ_NAME_EMPTY_BUF_SOS;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_QUEUE:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_EMPTY_BUFFER_RETURN_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_EMPTY_BUFFER_QUEUE;
+		irq_name = NSS_IRQ_NAME_EMPTY_BUF_QUEUE;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_TX_UNBLOCKED:
+		napi_poll_cb = nss_core_handle_napi_non_queue;
+		napi_wgt = NSS_TX_UNBLOCKED_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_TX_UNBLOCKED;
+		irq_name = NSS_IRQ_NAME_TX_UNBLOCK;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_0:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_0;
+		irq_name = NSS_IRQ_NAME_QUEUE0;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_1:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_1;
+		irq_name = NSS_IRQ_NAME_QUEUE1;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_2:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_2;
+		irq_name = NSS_IRQ_NAME_QUEUE2;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_3:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_3;
+		irq_name = NSS_IRQ_NAME_QUEUE3;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_COREDUMP_COMPLETE:
+		napi_poll_cb = nss_core_handle_napi_emergency;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_COREDUMP_COMPLETE;
+		irq_name = NSS_IRQ_NAME_COREDUMP_COMPLETE;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_PAGED_EMPTY_BUFFER_SOS:
+		napi_poll_cb = nss_core_handle_napi_non_queue;
+		napi_wgt = NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_PAGED_EMPTY_BUFFERS_SOS;
+		irq_name = NSS_IRQ_NAME_PAGED_EMPTY_BUF_SOS;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_PROFILE_DMA:
+		napi_poll_cb = nss_core_handle_napi_sdma;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_PROFILE_DMA;
+		irq_name = NSS_IRQ_NAME_PROFILE_DMA;
+		break;
 
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_TX_UNBLOCKED) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_non_queue, NSS_TX_UNBLOCKED_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_TX_UNBLOCKED;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss-tx-unblock", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_0) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_0;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue0", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_1) {
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_1;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue1", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_2) {
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_2;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue2", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_3) {
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_3;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue3", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_COREDUMP_COMPLETE) {
-		int_ctx->cause = NSS_N2H_INTR_COREDUMP_COMPLETE;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_emergency, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_coredump_complete", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_PAGED_EMPTY_BUFFER_SOS) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_non_queue, NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_PAGED_EMPTY_BUFFERS_SOS;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_paged_empty_buf_sos", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_PROFILE_DMA) {
-		int_ctx->cause = NSS_N2H_INTR_PROFILE_DMA;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_sdma, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_profile_dma", int_ctx);
+	default:
+		nss_warning("%px: nss%d: unsupported irq# %d\n", nss_ctx, nss_ctx->id, irq_num);
+		return err;
 	}
 
+	netif_napi_add_weight(NSS_CTX_NAPI_NDEV(nss_ctx), &int_ctx->napi, napi_poll_cb, napi_wgt);
+	int_ctx->cause = cause;
+	err = request_irq(irq, nss_hal_handle_irq, 0, irq_name, int_ctx);
 	if (err) {
+		nss_warning("%px: nss%d: request_irq failed for irq# %d\n", nss_ctx, nss_ctx->id, irq_num);
 		return err;
 	}
 
@@ -685,7 +722,7 @@ static int __nss_hal_request_irq(struct
 /*
  * __nss_hal_init_imem
  */
-void __nss_hal_init_imem(struct nss_ctx_instance *nss_ctx)
+static void __nss_hal_init_imem(struct nss_ctx_instance *nss_ctx)
 {
 	/*
 	 * Nothing to be done as there are no TCM in ipq60xx
@@ -695,7 +732,7 @@ void __nss_hal_init_imem(struct nss_ctx_
 /*
  * __nss_hal_init_utcm_shared
  */
-bool __nss_hal_init_utcm_shared(struct nss_ctx_instance *nss_ctx, uint32_t *meminfo_start)
+static bool __nss_hal_init_utcm_shared(struct nss_ctx_instance *nss_ctx, uint32_t *meminfo_start)
 {
 	struct nss_meminfo_ctx *mem_ctx = &nss_ctx->meminfo_ctx;
 	uint32_t utcm_shared_map_magic = meminfo_start[2];
--- a/nss_hal/ipq806x/nss_hal_pvt.c
+++ b/nss_hal/ipq806x/nss_hal_pvt.c
@@ -461,7 +461,7 @@ static struct nss_platform_data *__nss_h
 	npd->nphys = res_nphys.start;
 	npd->vphys = res_vphys.start;
 
-	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+	npd->nmap = nss_ioremap(npd->nphys, resource_size(&res_nphys));
 	if (!npd->nmap) {
 		nss_info_always("%px: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -477,10 +477,9 @@ static struct nss_platform_data *__nss_h
 	/*
 	 * Clear TCM memory used by this core
 	 */
-	for (i = 0; i < resource_size(&res_vphys) ; i += 4) {
+	for (i = 0; i < resource_size(&res_vphys) ; i += 4)
 		nss_write_32(npd->vmap, i, 0);
-		NSS_CORE_DMA_CACHE_MAINT((npd->vmap + i), 4, DMA_TO_DEVICE);
-	}
+
 	NSS_CORE_DSB();
 
 	/*
@@ -714,7 +713,7 @@ static int __nss_hal_common_reset(struct
 	}
 	of_node_put(cmn);
 
-	fpb_base = ioremap_nocache(res_nss_fpb_base.start, resource_size(&res_nss_fpb_base));
+	fpb_base = nss_ioremap(res_nss_fpb_base.start, resource_size(&res_nss_fpb_base));
 	if (!fpb_base) {
 		pr_err("%px: ioremap fail for nss_fpb_base\n", nss_dev);
 		return -EFAULT;
@@ -1186,7 +1185,7 @@ static int __nss_hal_request_irq(struct
 	}
 
 	int_ctx->irq = npd->irq[irq_num];
-	netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi, 64);
+	netif_napi_add_weight(NSS_CTX_NAPI_NDEV(nss_ctx), &int_ctx->napi, nss_core_handle_napi, 64);
 
 	return 0;
 }
--- a/nss_hal/ipq807x/nss_hal_pvt.c
+++ b/nss_hal/ipq807x/nss_hal_pvt.c
@@ -67,6 +67,20 @@
 #define NSS_NC_AXI_CLK "nss-nc-axi-clk"
 
 /*
+ * N2H interrupts
+ */
+#define NSS_IRQ_NAME_EMPTY_BUF_SOS "nss_empty_buf_sos"
+#define NSS_IRQ_NAME_EMPTY_BUF_QUEUE "nss_empty_buf_queue"
+#define NSS_IRQ_NAME_TX_UNBLOCK "nss-tx-unblock"
+#define NSS_IRQ_NAME_QUEUE0 "nss_queue0"
+#define NSS_IRQ_NAME_QUEUE1 "nss_queue1"
+#define NSS_IRQ_NAME_QUEUE2 "nss_queue2"
+#define NSS_IRQ_NAME_QUEUE3 "nss_queue3"
+#define NSS_IRQ_NAME_COREDUMP_COMPLETE "nss_coredump_complete"
+#define NSS_IRQ_NAME_PAGED_EMPTY_BUF_SOS "nss_paged_empty_buf_sos"
+#define NSS_IRQ_NAME_PROFILE_DMA "nss_profile_dma"
+
+/*
  * Voltage values
  */
 #define NOMINAL_VOLTAGE 1
@@ -237,7 +251,7 @@ static struct nss_platform_data *__nss_h
 	npd->vphys = res_vphys.start;
 	npd->qgic_phys = res_qgic_phys.start;
 
-	npd->nmap = ioremap_nocache(npd->nphys, resource_size(&res_nphys));
+	npd->nmap = nss_ioremap(npd->nphys, resource_size(&res_nphys));
 	if (!npd->nmap) {
 		nss_info_always("%px: nss%d: ioremap() fail for nphys\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -250,7 +264,7 @@ static struct nss_platform_data *__nss_h
 		goto out;
 	}
 
-	npd->qgic_map = ioremap_nocache(npd->qgic_phys, resource_size(&res_qgic_phys));
+	npd->qgic_map = nss_ioremap(npd->qgic_phys, resource_size(&res_qgic_phys));
 	if (!npd->qgic_map) {
 		nss_info_always("%px: nss%d: ioremap() fail for qgic map\n", nss_ctx, nss_ctx->id);
 		goto out;
@@ -259,10 +273,9 @@ static struct nss_platform_data *__nss_h
 	/*
 	 * Clear TCM memory used by this core
 	 */
-	for (i = 0; i < resource_size(&res_vphys) ; i += 4) {
+	for (i = 0; i < resource_size(&res_vphys) ; i += 4)
 		nss_write_32(npd->vmap, i, 0);
-		NSS_CORE_DMA_CACHE_MAINT((npd->vmap + i), 4, DMA_TO_DEVICE);
-	}
+
 	NSS_CORE_DSB();
 
 	/*
@@ -470,7 +483,7 @@ static int __nss_hal_common_reset(struct
 	}
 	of_node_put(cmn);
 
-	nss_misc_reset = ioremap_nocache(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
+	nss_misc_reset = nss_ioremap(res_nss_misc_reset.start, resource_size(&res_nss_misc_reset));
 	if (!nss_misc_reset) {
 		pr_err("%px: ioremap fail for nss_misc_reset\n", nss_dev);
 		return -EFAULT;
@@ -655,74 +668,96 @@ static void __nss_hal_send_interrupt(str
 static int __nss_hal_request_irq(struct nss_ctx_instance *nss_ctx, struct nss_platform_data *npd, int irq_num)
 {
 	struct int_ctx_instance *int_ctx = &nss_ctx->int_ctx[irq_num];
+	uint32_t cause, napi_wgt;
 	int err = -1, irq = npd->irq[irq_num];
+	int (*napi_poll_cb)(struct napi_struct *, int) = NULL;
+	const char *irq_name;
 
 	irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
 
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_SOS) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_non_queue, NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_EMPTY_BUFFERS_SOS;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_empty_buf_sos", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_QUEUE) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_EMPTY_BUFFER_RETURN_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_EMPTY_BUFFER_QUEUE;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_empty_buf_queue", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_TX_UNBLOCKED) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_non_queue, NSS_TX_UNBLOCKED_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_TX_UNBLOCKED;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss-tx-unblock", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_0) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_0;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue0", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_1) {
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_1;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue1", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_2) {
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_2;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue2", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_3) {
-		int_ctx->cause = NSS_N2H_INTR_DATA_QUEUE_3;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_queue, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_queue3", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_COREDUMP_COMPLETE) {
-		int_ctx->cause = NSS_N2H_INTR_COREDUMP_COMPLETE;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_emergency, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_coredump_complete", int_ctx);
-	}
-
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_PAGED_EMPTY_BUFFER_SOS) {
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_non_queue, NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT);
-		int_ctx->cause = NSS_N2H_INTR_PAGED_EMPTY_BUFFERS_SOS;
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_paged_empty_buf_sos", int_ctx);
-	}
+	switch (irq_num) {
+	case NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_SOS:
+		napi_poll_cb = nss_core_handle_napi_non_queue;
+		napi_wgt = NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_EMPTY_BUFFERS_SOS;
+		irq_name = NSS_IRQ_NAME_EMPTY_BUF_SOS;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_EMPTY_BUFFER_QUEUE:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_EMPTY_BUFFER_RETURN_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_EMPTY_BUFFER_QUEUE;
+		irq_name = NSS_IRQ_NAME_EMPTY_BUF_QUEUE;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_TX_UNBLOCKED:
+		napi_poll_cb = nss_core_handle_napi_non_queue;
+		napi_wgt = NSS_TX_UNBLOCKED_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_TX_UNBLOCKED;
+		irq_name = NSS_IRQ_NAME_TX_UNBLOCK;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_0:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_0;
+		irq_name = NSS_IRQ_NAME_QUEUE0;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_1:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_1;
+		irq_name = NSS_IRQ_NAME_QUEUE1;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_2:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_2;
+		irq_name = NSS_IRQ_NAME_QUEUE2;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_DATA_QUEUE_3:
+		napi_poll_cb = nss_core_handle_napi_queue;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_DATA_QUEUE_3;
+		irq_name = NSS_IRQ_NAME_QUEUE3;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_COREDUMP_COMPLETE:
+		napi_poll_cb = nss_core_handle_napi_emergency;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_COREDUMP_COMPLETE;
+		irq_name = NSS_IRQ_NAME_COREDUMP_COMPLETE;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_PAGED_EMPTY_BUFFER_SOS:
+		napi_poll_cb = nss_core_handle_napi_non_queue;
+		napi_wgt = NSS_EMPTY_BUFFER_SOS_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_PAGED_EMPTY_BUFFERS_SOS;
+		irq_name = NSS_IRQ_NAME_PAGED_EMPTY_BUF_SOS;
+		break;
+
+	case NSS_HAL_N2H_INTR_PURPOSE_PROFILE_DMA:
+		napi_poll_cb = nss_core_handle_napi_sdma;
+		napi_wgt = NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT;
+		cause = NSS_N2H_INTR_PROFILE_DMA;
+		irq_name = NSS_IRQ_NAME_PROFILE_DMA;
+		break;
 
-	if (irq_num == NSS_HAL_N2H_INTR_PURPOSE_PROFILE_DMA) {
-		int_ctx->cause = NSS_N2H_INTR_PROFILE_DMA;
-		netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, nss_core_handle_napi_sdma, NSS_DATA_COMMAND_BUFFER_PROCESSING_WEIGHT);
-		err = request_irq(irq, nss_hal_handle_irq, 0, "nss_profile_dma", int_ctx);
+	default:
+		nss_warning("%px: nss%d: unsupported irq# %d\n", nss_ctx, nss_ctx->id, irq_num);
+		return err;
 	}
 
+	netif_napi_add_weight(NSS_CTX_NAPI_NDEV(nss_ctx), &int_ctx->napi, napi_poll_cb, napi_wgt);
+	int_ctx->cause = cause;
+	err = request_irq(irq, nss_hal_handle_irq, 0, irq_name, int_ctx);
 	if (err) {
+		nss_warning("%px: nss%d: request_irq failed for irq# %d\n", nss_ctx, nss_ctx->id, irq_num);
 		return err;
 	}
-
 	int_ctx->irq = irq;
 	return 0;
 }
@@ -730,7 +765,7 @@ static int __nss_hal_request_irq(struct
 /*
  * __nss_hal_init_imem
  */
-void __nss_hal_init_imem(struct nss_ctx_instance *nss_ctx)
+static void __nss_hal_init_imem(struct nss_ctx_instance *nss_ctx)
 {
 	struct nss_meminfo_ctx *mem_ctx = &nss_ctx->meminfo_ctx;
 
@@ -745,7 +780,7 @@ void __nss_hal_init_imem(struct nss_ctx_
 /*
  * __nss_hal_init_utcm_shared
  */
-bool __nss_hal_init_utcm_shared(struct nss_ctx_instance *nss_ctx, uint32_t *meminfo_start)
+static bool __nss_hal_init_utcm_shared(struct nss_ctx_instance *nss_ctx, uint32_t *meminfo_start)
 {
 	/*
 	 * Nothing to be done as there are no TCM in ipq807x
--- a/nss_hal/ipq95xx/nss_hal_pvt.c
+++ b/nss_hal/ipq95xx/nss_hal_pvt.c
@@ -724,19 +724,19 @@ static int __nss_hal_clock_configure(str
 		}
 	}
 
-	nss_info_always("Supported Frequencies - ");
+	nss_info("Supported Frequencies - ");
 	for (i = 0; i < NSS_FREQ_MAX_SCALE; i++) {
 		switch (nss_runtime_samples.freq_scale[i].frequency) {
 		case NSS_FREQ_748:
-			nss_info_always("748 MHz ");
+			nss_info("748 MHz ");
 			break;
 
 		case NSS_FREQ_1497:
-			nss_info_always("1.497 GHz ");
+			nss_info("1.497 GHz ");
 			break;
 
 		case NSS_FREQ_1689:
-			nss_info_always("1.689 GHz ");
+			nss_info("1.689 GHz ");
 			break;
 
 		default:
@@ -890,9 +890,9 @@ static int __nss_hal_request_irq(struct
 	}
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
-	netif_napi_add(&nss_ctx->napi_ndev, &int_ctx->napi, napi_poll_cb, napi_wgt);
+	netif_napi_add(NSS_CTX_NAPI_NDEV(nss_ctx), &int_ctx->napi, napi_poll_cb, napi_wgt);
 #else
-	netif_napi_add_weight(&nss_ctx->napi_ndev, &int_ctx->napi, napi_poll_cb, napi_wgt);
+	netif_napi_add_weight(NSS_CTX_NAPI_NDEV(nss_ctx), &int_ctx->napi, napi_poll_cb, napi_wgt);
 #endif
 	int_ctx->cause = cause;
 	err = request_irq(irq, nss_hal_handle_irq, 0, irq_name, int_ctx);
--- a/nss_hal/nss_hal.c
+++ b/nss_hal/nss_hal.c
@@ -27,9 +27,9 @@
 #include <linux/firmware.h>
 #include <linux/of.h>
 #include <linux/irq.h>
+#include <linux/ethtool.h>
 
 #include "nss_hal.h"
-#include "nss_arch.h"
 #include "nss_core.h"
 #include "nss_tx_rx_common.h"
 #ifdef NSS_DATA_PLANE_GENERIC_SUPPORT
@@ -50,6 +50,57 @@
  */
 #define NSS_AP0_IMAGE "qca-nss0.bin"
 #define NSS_AP1_IMAGE "qca-nss1.bin"
+#define BUFFER_SIZE 8192
+
+// Function to search for the byte sequence in the buffer
+static unsigned char *search_sequence(const unsigned char *buffer,
+				      size_t buffer_size,
+				      const unsigned char *sequence,
+				      size_t sequence_size)
+{
+	for (size_t i = 0; i <= buffer_size - sequence_size; i++) {
+		if (memcmp(buffer + i, sequence, sequence_size) == 0) {
+			return (unsigned char *)(buffer + i);
+		}
+	}
+	return NULL;
+}
+
+static int nss_hal_firmware_info(struct platform_device *nss_dev,
+				 const struct firmware *fw)
+{
+	unsigned char *start_pos, *end_pos;
+	size_t i;
+	unsigned char start_sequence[] = { 0x56, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e, 0x3a, 0x20 };
+	unsigned char end_sequence[] = { 0x00 };
+	char version[256];
+	bool found = false;
+
+	// Search for the start sequence
+	start_pos = search_sequence(fw->data, fw->size, start_sequence, sizeof(start_sequence));
+	if (start_pos) {
+		start_pos += sizeof(start_sequence);
+
+		end_pos = search_sequence(start_pos, fw->size - (start_pos - fw->data), end_sequence, sizeof(end_sequence));
+		if (end_pos) {
+			// Convert the version information to a string
+			for (i = 0; start_pos + i < end_pos && i < sizeof(version) - 1; i++) {
+				version[i] = start_pos[i];
+			}
+			version[i] = '\0';
+
+			dev_info(&nss_dev->dev, "NSS fw version: %s\n", version);
+			found = true;
+		}
+	}
+
+	if (!found) {
+		dev_err(&nss_dev->dev, "Unable to get NSS fw version\n");
+		return -EINVAL;
+	}
+
+	return 0;
+}
 
 int nss_hal_firmware_load(struct nss_ctx_instance *nss_ctx, struct platform_device *nss_dev, struct nss_platform_data *npd)
 {
@@ -58,9 +109,9 @@ int nss_hal_firmware_load(struct nss_ctx
 	int rc;
 
 	if (nss_ctx->id == 0) {
-		rc = request_firmware(&nss_fw, NSS_AP0_IMAGE, &(nss_dev->dev));
+		rc = firmware_request_nowarn(&nss_fw, NSS_AP0_IMAGE, &(nss_dev->dev));
 	} else if (nss_ctx->id == 1) {
-		rc = request_firmware(&nss_fw, NSS_AP1_IMAGE, &(nss_dev->dev));
+		rc = firmware_request_nowarn(&nss_fw, NSS_AP1_IMAGE, &(nss_dev->dev));
 	} else {
 		nss_warning("%px: Invalid nss dev: %d\n", nss_ctx, nss_ctx->id);
 		return -EINVAL;
@@ -82,7 +133,11 @@ int nss_hal_firmware_load(struct nss_ctx
 		return rc;
 	}
 
-	nss_info("nss_driver - fw of size %d  bytes copied to load addr: %x, nss_id : %d\n", (int)nss_fw->size, npd->load_addr, nss_dev->id);
+	if (nss_ctx->id == 0) {
+	  nss_hal_firmware_info(nss_dev, nss_fw);
+	}
+
+	dev_info(&nss_dev->dev, "fw of size %d bytes copied to addr: %x, nss_id: %d\n", (int)nss_fw->size, npd->load_addr, nss_ctx->id);
 	memcpy_toio(load_mem, nss_fw->data, nss_fw->size);
 	release_firmware(nss_fw);
 	iounmap(load_mem);
@@ -305,10 +360,12 @@ int nss_hal_probe(struct platform_device
 	/*
 	 * Initialize the dummy netdevice.
 	 */
-	init_dummy_netdev(&nss_ctx->napi_ndev);
+	nss_ctx->napi_ndev = alloc_netdev_dummy(0);
+	if (!nss_ctx->napi_ndev)
+		goto err_init;
 
 	for (i = 0; i < npd->num_irq; i++) {
-		err = nss_hal_register_irq(nss_ctx, npd, &nss_ctx->napi_ndev, i);
+		err = nss_hal_register_irq(nss_ctx, npd, NSS_CTX_NAPI_NDEV(nss_ctx), i);
 		if (err) {
 			goto err_register_irq;
 		}
@@ -586,7 +643,9 @@ int nss_hal_probe(struct platform_device
 	if (npd->wifioffload_enabled == NSS_FEATURE_ENABLED) {
 		nss_top->wifi_handler_id = nss_dev->id;
 		nss_top->dynamic_interface_table[NSS_DYNAMIC_INTERFACE_TYPE_VAP] = nss_dev->id;
+#if defined(NSS_HAL_IPQ806x_SUPPORT)
 		nss_wifi_register_handler();
+#endif
 		nss_wifili_register_handler();
 #ifdef NSS_DRV_WIFI_EXT_VDEV_ENABLE
 		nss_wifi_ext_vdev_register_handler();
@@ -793,6 +852,7 @@ int nss_hal_probe(struct platform_device
 	}
 
 	nss_info("%px: All resources initialized and nss core%d has been brought out of reset", nss_ctx, nss_dev->id);
+	nss_bootwait();
 	goto out;
 
 err_register_irq:
@@ -800,6 +860,9 @@ err_register_irq:
 		nss_hal_clean_up_irq(&nss_ctx->int_ctx[i]);
 	}
 
+	free_netdev(nss_ctx->napi_ndev);
+	nss_ctx->napi_ndev = NULL;
+
 err_init:
 	if (nss_dev->dev.of_node) {
 		if (npd->nmap) {
@@ -842,6 +905,9 @@ int nss_hal_remove(struct platform_devic
 		nss_hal_clean_up_irq(&nss_ctx->int_ctx[i]);
 	}
 
+	free_netdev(nss_ctx->napi_ndev);
+	nss_ctx->napi_ndev = NULL;
+
 	/*
 	 * nss-drv is exiting, unregister and restore host data plane
 	 */
--- a/nss_phys_if.c
+++ b/nss_phys_if.c
@@ -402,7 +402,7 @@ nss_tx_status_t nss_phys_if_link_state(s
  * nss_phys_if_mac_addr()
  *	Send a MAC address to physical interface
  */
-nss_tx_status_t nss_phys_if_mac_addr(struct nss_ctx_instance *nss_ctx, uint8_t *addr, uint32_t if_num)
+nss_tx_status_t nss_phys_if_mac_addr(struct nss_ctx_instance *nss_ctx, const uint8_t *addr, uint32_t if_num)
 {
 	struct nss_phys_if_msg nim;
 	struct nss_if_mac_address_set *nmas;
--- a/nss_phys_if.h
+++ b/nss_phys_if.h
@@ -212,6 +212,13 @@ struct nss_ctx_instance *nss_phys_if_reg
 					uint32_t features);
 
 /**
+ * @brief Unregister GMAC packet/message callbacks.
+ *
+ * @param if_num GMAC i/f number
+ */
+void nss_phys_if_unregister(uint32_t if_num);
+
+/**
  * @brief Send GMAC packet
  *
  * @param nss_ctx NSS context
@@ -284,7 +291,7 @@ nss_tx_status_t nss_phys_if_link_state(s
  *
  * @return nss_tx_status_t Tx status
  */
-nss_tx_status_t nss_phys_if_mac_addr(struct nss_ctx_instance *nss_ctx, uint8_t *addr, uint32_t if_num);
+nss_tx_status_t nss_phys_if_mac_addr(struct nss_ctx_instance *nss_ctx, const uint8_t *addr, uint32_t if_num);
 
 /**
  * @brief Send MTU change notification to NSS
