--- a/nss_edma.c
+++ b/nss_edma.c
@@ -18,6 +18,7 @@
  * nss_edma.c
  *	NSS EDMA APIs
  */
+#include "nss_tx_rx_common.h"
 #include "nss_edma_stats.h"
 #include "nss_edma_strings.h"
 
--- a/nss_edma_stats.c
+++ b/nss_edma_stats.c
@@ -650,9 +650,8 @@ void nss_edma_stats_dentry_create(void)
 	/*
 	 *  edma error stats
 	 */
-	edma_err_stats_d = NULL;
 	edma_err_stats_d = debugfs_create_file("err_stats", 0400, edma_d, &nss_top_main, &nss_edma_err_stats_ops);
-	if (unlikely(edma_port_stats_d == NULL)) {
+	if (unlikely(edma_err_stats_d == NULL)) {
 		nss_warning("Failed to create qca-nss-drv/stats/edma/%d/err_stats file", 0);
 		return;
 	}
--- a/nss_edma_strings.c
+++ b/nss_edma_strings.c
@@ -18,6 +18,7 @@
 #include "nss_core.h"
 #include <nss_edma.h>
 #include "nss_strings.h"
+#include "nss_edma_strings.h"
 
 /*
  * nss_edma_strings_stats_node
--- a/nss_eth_rx.c
+++ b/nss_eth_rx.c
@@ -20,6 +20,7 @@
  */
 
 #include <nss_core.h>
+#include "nss_tx_rx_common.h"
 #include "nss_eth_rx_stats.h"
 #include "nss_eth_rx_strings.h"
 
--- a/nss_eth_rx_strings.c
+++ b/nss_eth_rx_strings.c
@@ -19,6 +19,7 @@
 #include "nss_stats.h"
 #include "nss_core.h"
 #include "nss_strings.h"
+#include "nss_eth_rx_strings.h"
 
 /*
  * nss_eth_rx_strings_stats
--- a/nss_igs.c
+++ b/nss_igs.c
@@ -163,7 +163,7 @@ EXPORT_SYMBOL(nss_igs_register_if);
  * nss_igs_get_context()
  *	Get the IGS context.
  */
-struct nss_ctx_instance *nss_igs_get_context()
+struct nss_ctx_instance *nss_igs_get_context(void)
 {
 	return (struct nss_ctx_instance *)&nss_top_main.nss[nss_top_main.igs_handler_id];
 }
@@ -177,8 +177,6 @@ EXPORT_SYMBOL(nss_igs_get_context);
 void nss_igs_module_save(struct tc_action_ops *act, struct module *module)
 {
 	nss_assert(act);
-	nss_assert(act->type == TCA_ACT_MIRRED_NSS);
-
 	nss_igs_module = module;
 }
 EXPORT_SYMBOL(nss_igs_module_save);
@@ -188,7 +186,7 @@ EXPORT_SYMBOL(nss_igs_module_save);
  * nss_igs_module_get()
  *	Get the ingress shaping module reference.
  */
-bool nss_igs_module_get()
+bool nss_igs_module_get(void)
 {
 	nss_assert(nss_igs_module);
 	return try_module_get(nss_igs_module);
@@ -199,7 +197,7 @@ EXPORT_SYMBOL(nss_igs_module_get);
  * nss_igs_module_put()
  *	Release the ingress shaping module reference.
  */
-void nss_igs_module_put()
+void nss_igs_module_put(void)
 {
 	nss_assert(nss_igs_module);
 	module_put(nss_igs_module);
