--- a/nss_ipv4.c
+++ b/nss_ipv4.c
@@ -599,7 +599,7 @@ void nss_ipv4_free_conn_tables(void)
  * nss_ipv4_accel_mode_cfg_handler()
  *	Configure acceleration mode for IPv4
  */
-static int nss_ipv4_accel_mode_cfg_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_ipv4_accel_mode_cfg_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[0];
@@ -643,7 +643,7 @@ static int nss_ipv4_accel_mode_cfg_handl
  * nss_ipv4_dscp_map_cfg_handler()
  *	Sysctl handler for dscp/pri mappings.
  */
-static int nss_ipv4_dscp_map_cfg_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_ipv4_dscp_map_cfg_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[0];
@@ -708,35 +708,7 @@ static struct ctl_table nss_ipv4_table[]
 		.maxlen		= sizeof(struct nss_dscp_map_entry),
 		.mode		= 0644,
 		.proc_handler	= &nss_ipv4_dscp_map_cfg_handler,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ipv4_dir[] = {
-	{
-		.procname		= "ipv4cfg",
-		.mode			= 0555,
-		.child			= nss_ipv4_table,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ipv4_root_dir[] = {
-	{
-		.procname		= "nss",
-		.mode			= 0555,
-		.child			= nss_ipv4_dir,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ipv4_root[] = {
-	{
-		.procname		= "dev",
-		.mode			= 0555,
-		.child			= nss_ipv4_root_dir,
-	},
-	{ }
+	}
 };
 
 static struct ctl_table_header *nss_ipv4_header;
@@ -753,7 +725,7 @@ void nss_ipv4_register_sysctl(void)
 	/*
 	 * Register sysctl table.
 	 */
-	nss_ipv4_header = register_sysctl_table(nss_ipv4_root);
+	nss_ipv4_header = register_sysctl("dev/nss/ipv4cfg", nss_ipv4_table);
 }
 
 /*
--- a/nss_ipv4_reasm_strings.c
+++ b/nss_ipv4_reasm_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_ipv4_reasm.h>
 #include "nss_strings.h"
+#include "nss_ipv4_reasm_strings.h"
 
 /*
  * nss_ipv4_reasm_strings_stats
--- a/nss_ipv4_strings.c
+++ b/nss_ipv4_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_ipv4.h>
 #include "nss_strings.h"
+#include "nss_ipv4_strings.h"
 
 /*
  * nss_ipv4_strings_exception_stats
--- a/nss_ipv6.c
+++ b/nss_ipv6.c
@@ -18,6 +18,7 @@
  * nss_ipv6.c
  *	NSS IPv6 APIs
  */
+#include "linux/ipv6.h"
 #include <nss_core.h>
 #include "nss_dscp_map.h"
 #include "nss_ipv6_stats.h"
@@ -377,7 +378,7 @@ EXPORT_SYMBOL(nss_ipv6_get_mgr);
  * nss_ipv6_register_handler()
  *	Register our handler to receive messages for this interface
  */
-void nss_ipv6_register_handler()
+void nss_ipv6_register_handler(void)
 {
 	struct nss_ctx_instance *nss_ctx = nss_ipv6_get_mgr();
 
@@ -594,7 +595,7 @@ void nss_ipv6_free_conn_tables(void)
  * nss_ipv6_accel_mode_cfg_handler()
  *	Configure acceleration mode for IPv6
  */
-static int nss_ipv6_accel_mode_cfg_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_ipv6_accel_mode_cfg_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[0];
@@ -638,7 +639,7 @@ static int nss_ipv6_accel_mode_cfg_handl
  * nss_ipv6_dscp_map_cfg_handler()
  *	Sysctl handler for dscp/pri mappings.
  */
-static int nss_ipv6_dscp_map_cfg_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_ipv6_dscp_map_cfg_handler(const struct ctl_table *ctl, int write, void *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[0];
@@ -702,35 +703,7 @@ static struct ctl_table nss_ipv6_table[]
 		.maxlen			= sizeof(struct nss_dscp_map_entry),
 		.mode			= 0644,
 		.proc_handler		= &nss_ipv6_dscp_map_cfg_handler,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ipv6_dir[] = {
-	{
-		.procname		= "ipv6cfg",
-		.mode			= 0555,
-		.child			= nss_ipv6_table,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ipv6_root_dir[] = {
-	{
-		.procname		= "nss",
-		.mode			= 0555,
-		.child			= nss_ipv6_dir,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ipv6_root[] = {
-	{
-		.procname		= "dev",
-		.mode			= 0555,
-		.child			= nss_ipv6_root_dir,
-	},
-	{ }
+	}
 };
 
 static struct ctl_table_header *nss_ipv6_header;
@@ -747,7 +720,7 @@ void nss_ipv6_register_sysctl(void)
 	/*
 	 * Register sysctl table.
 	 */
-	nss_ipv6_header = register_sysctl_table(nss_ipv6_root);
+	nss_ipv6_header = register_sysctl("dev/nss/ipv6cfg", nss_ipv6_table);
 }
 
 /*
--- a/nss_ipv6_reasm_strings.c
+++ b/nss_ipv6_reasm_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_ipv6_reasm.h>
 #include "nss_strings.h"
+#include "nss_ipv6_reasm_strings.h"
 
 /*
  * nss_ipv6_reasm_strings_stats
--- a/nss_ipv6_strings.c
+++ b/nss_ipv6_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_ipv6.h>
 #include "nss_strings.h"
+#include "nss_ipv6_strings.h"
 
 /*
  * nss_ipv6_strings_exception_stats
