--- a/nss_gre_log.c
+++ b/nss_gre_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_gre_log.h"
 
 #define NSS_GRE_LOG_MESSAGE_TYPE_INDEX(type) ((type) - NSS_IF_MAX_MSG_TYPES - 1)
 
@@ -151,7 +152,7 @@ static void nss_gre_log_verbose(struct n
  */
 void nss_gre_log_tx_msg(struct nss_gre_msg *ngm)
 {
-	if (ngm->cm.type >= NSS_GRE_MSG_MAX) {
+	if (ngm->cm.type < (NSS_IF_MAX_MSG_TYPES + 1) || ngm->cm.type >= (NSS_IF_MAX_MSG_TYPES + NSS_GRE_MSG_MAX + 1)) {
 		nss_warning("%px: Invalid message type\n", ngm);
 		return;
 	}
@@ -171,6 +172,11 @@ void nss_gre_log_rx_msg(struct nss_gre_m
 		return;
 	}
 
+	if (ngm->cm.type < (NSS_IF_MAX_MSG_TYPES + 1) || ngm->cm.type >= (NSS_IF_MAX_MSG_TYPES + NSS_GRE_MSG_MAX + 1)) {
+		nss_warning("%px: Invalid message type\n", ngm);
+		return;
+	}
+
 	if (ngm->cm.response == NSS_CMN_RESPONSE_NOTIFY || (ngm->cm.response == NSS_CMN_RESPONSE_ACK)) {
 		nss_info("%px: type[%d]:%s, response[%d]:%s\n", ngm, ngm->cm.type,
 			nss_gre_log_message_types_str[NSS_GRE_LOG_MESSAGE_TYPE_INDEX(ngm->cm.type)],
--- a/nss_gre_redir_lag_ds_log.c
+++ b/nss_gre_redir_lag_ds_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_gre_redir_lag_ds_log.h"
 
 /*
  * nss_gre_redir_lag_ds_log_message_types_str
--- a/nss_gre_redir_lag_us_log.c
+++ b/nss_gre_redir_lag_us_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_gre_redir_lag_us_log.h"
 
 /*
  * nss_gre_redir_lag_us_log_message_types_str
--- a/nss_gre_redir_log.c
+++ b/nss_gre_redir_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_gre_redir_log.h"
 
 /*
  * nss_gre_redir_log_message_types_str
--- a/nss_gre_redir_mark_log.c
+++ b/nss_gre_redir_mark_log.c
@@ -15,6 +15,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_gre_redir_mark_log.h"
 
 /*
  * nss_gre_redir_mark_log_message_types_str
--- a/nss_gre_tunnel_log.c
+++ b/nss_gre_tunnel_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_gre_tunnel_log.h"
 
 /*
  * nss_gre_tunnel_log_message_types_str
--- a/nss_l2tpv2.c
+++ b/nss_l2tpv2.c
@@ -31,7 +31,7 @@ static struct nss_l2tpv2_stats_session_d
  * nss_l2tpv2_session_debug_stats_sync
  *	Per session debug stats for l2tpv2
  */
-void nss_l2tpv2_session_debug_stats_sync(struct nss_ctx_instance *nss_ctx, struct nss_l2tpv2_sync_session_stats_msg *stats_msg, uint16_t if_num)
+static void nss_l2tpv2_session_debug_stats_sync(struct nss_ctx_instance *nss_ctx, struct nss_l2tpv2_sync_session_stats_msg *stats_msg, uint16_t if_num)
 {
 	int i;
 	spin_lock_bh(&nss_l2tpv2_session_debug_stats_lock);
--- a/nss_l2tpv2_log.c
+++ b/nss_l2tpv2_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_l2tpv2_log.h"
 
 /*
  * nss_l2tpv2_log_message_types_str
--- a/nss_l2tpv2_strings.c
+++ b/nss_l2tpv2_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_l2tpv2.h>
 #include "nss_strings.h"
+#include "nss_l2tpv2_strings.h"
 
 /*
  * nss_l2tpv2_strings_session_stats
--- a/nss_map_t.c
+++ b/nss_map_t.c
@@ -65,7 +65,7 @@ static bool nss_map_t_verify_if_num(uint
  * nss_map_t_instance_debug_stats_sync
  *	debug stats for map_t
  */
-void nss_map_t_instance_debug_stats_sync(struct nss_ctx_instance *nss_ctx, struct nss_map_t_sync_stats_msg *stats_msg, uint16_t if_num)
+static void nss_map_t_instance_debug_stats_sync(struct nss_ctx_instance *nss_ctx, struct nss_map_t_sync_stats_msg *stats_msg, uint16_t if_num)
 {
 	int i;
 	enum nss_dynamic_interface_type if_type;
@@ -378,7 +378,7 @@ EXPORT_SYMBOL(nss_map_t_unregister_if);
 /*
  * nss_get_map_t_context()
  */
-struct nss_ctx_instance *nss_map_t_get_context()
+struct nss_ctx_instance *nss_map_t_get_context(void)
 {
 	return (struct nss_ctx_instance *)&nss_top_main.nss[nss_top_main.map_t_handler_id];
 }
--- a/nss_map_t_log.c
+++ b/nss_map_t_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_map_t_log.h"
 
 /*
  * nss_map_t_log_message_types_str
--- a/nss_map_t_strings.c
+++ b/nss_map_t_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_map_t.h>
 #include "nss_strings.h"
+#include "nss_map_t_strings.h"
 
 /*
  * nss_map_t_strings_instance_stats
--- a/nss_pptp.c
+++ b/nss_pptp.c
@@ -43,7 +43,7 @@ static struct nss_pptp_pvt {
  * nss_pptp_session_debug_stats_sync
  *	Per session debug stats for pptp
  */
-void nss_pptp_session_debug_stats_sync(struct nss_ctx_instance *nss_ctx,
+static void nss_pptp_session_debug_stats_sync(struct nss_ctx_instance *nss_ctx,
 				struct nss_pptp_sync_session_stats_msg *stats_msg, uint16_t if_num)
 {
 	int i, j, if_type;
--- a/nss_pptp_log.c
+++ b/nss_pptp_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_pptp_log.h"
 
 /*
  * nss_pptp_log_message_types_str
--- a/nss_pptp_strings.c
+++ b/nss_pptp_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_pptp.h>
 #include "nss_strings.h"
+#include "nss_pptp_strings.h"
 
 /*
  * nss_pptp_strings_session_debug_stats
--- a/nss_pvxlan_log.c
+++ b/nss_pvxlan_log.c
@@ -23,6 +23,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_pvxlan_log.h"
 
 /*
  * nss_pvxlan_log_message_types_str
--- a/nss_qvpn_log.c
+++ b/nss_qvpn_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_qvpn_log.h"
 
 /*
  * nss_qvpn_log_message_types_str
--- a/nss_qvpn_stats.c
+++ b/nss_qvpn_stats.c
@@ -17,6 +17,8 @@
  **************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include "nss_core.h"
 #include <nss_qvpn.h>
 #include "nss_qvpn_stats.h"
--- a/nss_tun6rd_log.c
+++ b/nss_tun6rd_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_tun6rd_log.h"
 
 /*
  * nss_tun6rd_log_message_types_str
--- a/nss_tunipip6_log.c
+++ b/nss_tunipip6_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_tunipip6_log.h"
 
 /*
  * nss_tunipip6_log_message_types_str
--- a/nss_tunipip6_stats.c
+++ b/nss_tunipip6_stats.c
@@ -17,6 +17,8 @@
  **************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include "nss_core.h"
 #include "nss_tunipip6.h"
 #include "nss_stats.h"
--- a/nss_udp_st_log.c
+++ b/nss_udp_st_log.c
@@ -23,6 +23,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_udp_st_log.h"
 
 /*
  * nss_udp_st_log_message_types_str
--- a/nss_udp_st_strings.c
+++ b/nss_udp_st_strings.c
@@ -20,6 +20,7 @@
 #include "nss_stats.h"
 #include "nss_core.h"
 #include <nss_udp_st.h>
+#include "nss_udp_st_strings.h"
 #include "nss_strings.h"
 
 /*
--- a/nss_vxlan_log.c
+++ b/nss_vxlan_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_vxlan_log.h"
 
 /*
  * nss_vxlan_log_message_types_str
