--- a/nss_c2c_tx.c
+++ b/nss_c2c_tx.c
@@ -330,35 +330,7 @@ static struct ctl_table nss_c2c_tx_table
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &nss_c2c_tx_performance_test_handler,
-	},
-	{ }
-};
-
-static struct ctl_table nss_c2c_tx_dir[] = {
-	{
-		.procname		= "c2c_tx",
-		.mode			= 0555,
-		.child			= nss_c2c_tx_table,
-	},
-	{ }
-};
-
-static struct ctl_table nss_c2c_tx_root_dir[] = {
-	{
-		.procname		= "nss",
-		.mode			= 0555,
-		.child			= nss_c2c_tx_dir,
-	},
-	{ }
-};
-
-static struct ctl_table nss_c2c_tx_root[] = {
-	{
-		.procname		= "dev",
-		.mode			= 0555,
-		.child			= nss_c2c_tx_root_dir,
-	},
-	{ }
+	}
 };
 
 static struct ctl_table_header *nss_c2c_tx_header;
@@ -378,7 +350,7 @@ void nss_c2c_tx_register_sysctl(void)
 	/*
 	 * Register sysctl table.
 	 */
-	nss_c2c_tx_header = register_sysctl_table(nss_c2c_tx_root);
+	nss_c2c_tx_header = register_sysctl("dev/nss/c2c_tx", nss_c2c_tx_table);
 }
 
 /*
--- a/nss_c2c_tx_log.c
+++ b/nss_c2c_tx_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_c2c_tx_log.h"
 
 /*
  * nss_c2c_tx_log_message_types_str
--- a/nss_dma.c
+++ b/nss_dma.c
@@ -285,7 +285,7 @@ EXPORT_SYMBOL(nss_dma_get_context);
  * nss_dma_test_handler()
  *	Handles the performance test.
  */
-static int nss_dma_test_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_dma_test_handler(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_ctx_instance *nss_ctx = nss_dma_get_context();
 	int cur_state = test_cfg.run.val;
@@ -375,8 +375,7 @@ static struct ctl_table nss_dma_table[]
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= proc_dointvec,
-	},
-	{ }
+	}
 };
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
--- a/nss_dma_log.c
+++ b/nss_dma_log.c
@@ -22,6 +22,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_dma_log.h"
 
 /*
  * nss_dma_log_message_types_str
@@ -61,7 +62,7 @@ static void nss_dma_configure_msg(struct
 static void nss_dma_test_perf_msg(struct nss_dma_msg *ndm)
 {
 	struct nss_dma_test_cfg *ndtc = &ndm->msg.test_cfg;
-	struct nss_cmn_node_stats *ncns = &ndtc->node_stats;
+	struct nss_cmn_node_stats *ncns __maybe_unused = &ndtc->node_stats;
 
 	nss_trace("%px: NSS DMA test perf message: \n",ndm);
 	nss_trace("%px: processed (TX: %u, RX:%u, time:%u)\n", ndm, ncns->tx_packets, ncns->rx_packets, ndtc->time_delta);
--- a/nss_dma_stats.c
+++ b/nss_dma_stats.c
@@ -17,6 +17,8 @@
  **************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include "nss_core.h"
 #include "nss_dma.h"
 #include "nss_dma_stats.h"
--- a/nss_dma_strings.c
+++ b/nss_dma_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include "nss_strings.h"
 #include "nss_dma_stats.h"
+#include "nss_dma_strings.h"
 
 /*
  * nss_dma_strings_stats
--- a/nss_freq.c
+++ b/nss_freq.c
@@ -202,7 +202,7 @@ static void nss_freq_compute_cpu_usage(s
  * nss_freq_scale_frequency()
  * 	Frequency scaling algorithm to scale frequency.
  */
-void nss_freq_scale_frequency(struct nss_ctx_instance *nss_ctx, uint32_t inst_cnt)
+static void nss_freq_scale_frequency(struct nss_ctx_instance *nss_ctx, uint32_t inst_cnt)
 {
 	uint32_t b_index;
 	uint32_t minimum;
--- a/nss_freq_log.c
+++ b/nss_freq_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_freq_log.h"
 
 /*
  * nss_freq_log_message_types_str
--- a/nss_freq_stats.c
+++ b/nss_freq_stats.c
@@ -24,6 +24,7 @@
 
 #include "nss_stats.h"
 #include "nss_tx_rx_common.h"
+#include "nss_freq_stats.h"
 
 /*
  * At any point, this object has the latest data about CPU utilization.
--- a/nss_n2h.c
+++ b/nss_n2h.c
@@ -357,7 +357,7 @@ static int nss_n2h_get_paged_payload_inf
  * nss_n2h_set_empty_buf_pool()
  *	Sets empty pool buffer
  */
-static int nss_n2h_set_empty_buf_pool(struct ctl_table *ctl, int write,
+static int nss_n2h_set_empty_buf_pool(const struct ctl_table *ctl, int write,
 				void __user *buffer,
 				size_t *lenp, loff_t *ppos,
 				nss_ptr_t core_num, int *new_val)
@@ -459,7 +459,7 @@ failure:
  * nss_n2h_set_empty_paged_pool_buf()
  *	Sets empty paged pool buffer
  */
-static int nss_n2h_set_empty_paged_pool_buf(struct ctl_table *ctl, int write,
+static int nss_n2h_set_empty_paged_pool_buf(const struct ctl_table *ctl, int write,
 				void __user *buffer,
 				size_t *lenp, loff_t *ppos,
 				nss_ptr_t core_num, int *new_val)
@@ -561,7 +561,7 @@ failure:
  * nss_n2h_set_water_mark()
  *	Sets water mark for N2H SOS
  */
-static int nss_n2h_set_water_mark(struct ctl_table *ctl, int write,
+static int nss_n2h_set_water_mark(const struct ctl_table *ctl, int write,
 					void __user *buffer,
 					size_t *lenp, loff_t *ppos,
 					uint32_t core_num, int *low, int *high)
@@ -675,7 +675,7 @@ failure:
  * nss_n2h_set_paged_water_mark()
  *	Sets water mark for paged pool N2H SOS
  */
-static int nss_n2h_set_paged_water_mark(struct ctl_table *ctl, int write,
+static int nss_n2h_set_paged_water_mark(const struct ctl_table *ctl, int write,
 					void __user *buffer,
 					size_t *lenp, loff_t *ppos,
 					uint32_t core_num, int *low, int *high)
@@ -789,7 +789,7 @@ failure:
  * nss_n2h_cfg_wifi_pool()
  *	Sets number of wifi payloads to adjust high water mark for N2H SoS
  */
-static int nss_n2h_cfg_wifi_pool(struct ctl_table *ctl, int write,
+static int nss_n2h_cfg_wifi_pool(const struct ctl_table *ctl, int write,
 					void __user *buffer,
 					size_t *lenp, loff_t *ppos,
 					int *payloads)
@@ -886,7 +886,7 @@ failure:
  * nss_n2h_empty_pool_buf_core1_handler()
  *	Sets the number of empty buffer for core 1
  */
-static int nss_n2h_empty_pool_buf_cfg_core1_handler(struct ctl_table *ctl,
+static int nss_n2h_empty_pool_buf_cfg_core1_handler(const struct ctl_table *ctl,
 				int write, void __user *buffer,
 				size_t *lenp, loff_t *ppos)
 {
@@ -898,7 +898,7 @@ static int nss_n2h_empty_pool_buf_cfg_co
  * nss_n2h_empty_pool_buf_core0_handler()
  *	Sets the number of empty buffer for core 0
  */
-static int nss_n2h_empty_pool_buf_cfg_core0_handler(struct ctl_table *ctl,
+static int nss_n2h_empty_pool_buf_cfg_core0_handler(const struct ctl_table *ctl,
 				int write, void __user *buffer,
 				size_t *lenp, loff_t *ppos)
 {
@@ -910,7 +910,7 @@ static int nss_n2h_empty_pool_buf_cfg_co
  * nss_n2h_empty_paged_pool_buf_cfg_core1_handler()
  *	Sets the number of empty paged buffer for core 1
  */
-static int nss_n2h_empty_paged_pool_buf_cfg_core1_handler(struct ctl_table *ctl,
+static int nss_n2h_empty_paged_pool_buf_cfg_core1_handler(const struct ctl_table *ctl,
 				int write, void __user *buffer,
 				size_t *lenp, loff_t *ppos)
 {
@@ -922,7 +922,7 @@ static int nss_n2h_empty_paged_pool_buf_
  * nss_n2h_empty_paged_pool_buf_cfg_core0_handler()
  *	Sets the number of empty paged buffer for core 0
  */
-static int nss_n2h_empty_paged_pool_buf_cfg_core0_handler(struct ctl_table *ctl,
+static int nss_n2h_empty_paged_pool_buf_cfg_core0_handler(const struct ctl_table *ctl,
 				int write, void __user *buffer,
 				size_t *lenp, loff_t *ppos)
 {
@@ -934,7 +934,7 @@ static int nss_n2h_empty_paged_pool_buf_
  * nss_n2h_water_mark_core1_handler()
  *	Sets water mark for core 1
  */
-static int nss_n2h_water_mark_core1_handler(struct ctl_table *ctl,
+static int nss_n2h_water_mark_core1_handler(const struct ctl_table *ctl,
 			int write, void __user *buffer,
 			size_t *lenp, loff_t *ppos)
 {
@@ -947,7 +947,7 @@ static int nss_n2h_water_mark_core1_hand
  * nss_n2h_water_mark_core0_handler()
  *	Sets water mark for core 0
  */
-static int nss_n2h_water_mark_core0_handler(struct ctl_table *ctl,
+static int nss_n2h_water_mark_core0_handler(const struct ctl_table *ctl,
 			int write, void __user *buffer,
 			size_t *lenp, loff_t *ppos)
 {
@@ -960,7 +960,7 @@ static int nss_n2h_water_mark_core0_hand
  * nss_n2h_paged_water_mark_core1_handler()
  *	Sets paged water mark for core 1
  */
-static int nss_n2h_paged_water_mark_core1_handler(struct ctl_table *ctl,
+static int nss_n2h_paged_water_mark_core1_handler(const struct ctl_table *ctl,
 			int write, void __user *buffer,
 			size_t *lenp, loff_t *ppos)
 {
@@ -973,7 +973,7 @@ static int nss_n2h_paged_water_mark_core
  * nss_n2h_paged_water_mark_core0_handler()
  *	Sets paged water mark for core 0
  */
-static int nss_n2h_paged_water_mark_core0_handler(struct ctl_table *ctl,
+static int nss_n2h_paged_water_mark_core0_handler(const struct ctl_table *ctl,
 			int write, void __user *buffer,
 			size_t *lenp, loff_t *ppos)
 {
@@ -986,7 +986,7 @@ static int nss_n2h_paged_water_mark_core
  * nss_n2h_wifi_payloads_handler()
  *	Sets number of wifi payloads
  */
-static int nss_n2h_wifi_payloads_handler(struct ctl_table *ctl,
+static int nss_n2h_wifi_payloads_handler(const struct ctl_table *ctl,
 			int write, void __user *buffer,
 			size_t *lenp, loff_t *ppos)
 {
@@ -998,7 +998,7 @@ static int nss_n2h_wifi_payloads_handler
  * nss_n2h_get_qos_mem_size_cfg_handler()
  *	Gets the QoS memory pool size
  */
-static int nss_n2h_get_qos_mem_size_cfg_handler(struct ctl_table *ctl,
+static int nss_n2h_get_qos_mem_size_cfg_handler(const struct ctl_table *ctl,
 			int write, void __user *buffer,
 			size_t *lenp, loff_t *ppos)
 {
@@ -1281,7 +1281,7 @@ failure:
  * nss_mitigation_handler()
  * Enable NSS MITIGATION
  */
-static int nss_n2h_mitigationcfg_core0_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_n2h_mitigationcfg_core0_handler(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[NSS_CORE_0];
@@ -1300,11 +1300,11 @@ static int nss_n2h_mitigationcfg_core0_h
 	}
 
 	if (!nss_n2h_core0_mitigation_cfg) {
-		printk(KERN_INFO "Disabling NSS MITIGATION\n");
+		dev_info(nss_ctx->dev, "Disabling NSS MITIGATION\n");
 		nss_n2h_mitigation_cfg(nss_ctx, 0, NSS_CORE_0);
 		return 0;
 	}
-	printk(KERN_INFO "Invalid input value.Valid value is 0, Runtime re-enabling not supported\n");
+	dev_info(nss_ctx->dev, "Invalid input value.Valid value is 0, Runtime re-enabling not supported\n");
 	return -EINVAL;
 }
 
@@ -1312,7 +1312,7 @@ static int nss_n2h_mitigationcfg_core0_h
  * nss_mitigation_handler()
  * Enable NSS MITIGATION
  */
-static int nss_n2h_mitigationcfg_core1_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_n2h_mitigationcfg_core1_handler(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[NSS_CORE_1];
@@ -1331,11 +1331,11 @@ static int nss_n2h_mitigationcfg_core1_h
 	}
 
 	if (!nss_n2h_core1_mitigation_cfg) {
-		printk(KERN_INFO "Disabling NSS MITIGATION\n");
+		dev_info(nss_ctx->dev, "Disabling NSS MITIGATION\n");
 		nss_n2h_mitigation_cfg(nss_ctx, 0, NSS_CORE_1);
 		return 0;
 	}
-	printk(KERN_INFO "Invalid input value.Valid value is 0, Runtime re-enabling not supported\n");
+	dev_info(nss_ctx->dev, "Invalid input value.Valid value is 0, Runtime re-enabling not supported\n");
 	return -EINVAL;
 }
 
@@ -1343,7 +1343,7 @@ static int nss_n2h_mitigationcfg_core1_h
  * nss_buf_handler()
  *	Add extra NSS bufs from host memory
  */
-static int nss_n2h_buf_cfg_core0_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_n2h_buf_cfg_core0_handler(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[NSS_CORE_0];
@@ -1367,14 +1367,14 @@ static int nss_n2h_buf_cfg_core0_handler
 	}
 
 	if ((nss_n2h_core0_add_buf_pool_size >= 1) && (nss_n2h_core0_add_buf_pool_size <= NSS_N2H_MAX_BUF_POOL_SIZE)) {
-		printk(KERN_INFO "configuring additional NSS pbufs\n");
+		dev_info(nss_ctx->dev, "Configuring additional NSS pbufs\n");
 		ret = nss_n2h_buf_pool_cfg(nss_ctx, nss_n2h_core0_add_buf_pool_size, NSS_CORE_0);
 		nss_n2h_core0_add_buf_pool_size = nss_ctx->buf_sz_allocated;
-		printk(KERN_INFO "additional pbufs of size %d got added to NSS\n", nss_ctx->buf_sz_allocated);
+		dev_info(nss_ctx->dev, "Additional pbufs of size %d got added to NSS\n", nss_ctx->buf_sz_allocated);
 		return ret;
 	}
 
-	printk(KERN_INFO "Invalid input value. should be greater than 1 and less than %d\n", NSS_N2H_MAX_BUF_POOL_SIZE);
+	dev_info(nss_ctx->dev, "Invalid input value. should be greater than 1 and less than %d\n", NSS_N2H_MAX_BUF_POOL_SIZE);
 	return -EINVAL;
 }
 
@@ -1382,7 +1382,7 @@ static int nss_n2h_buf_cfg_core0_handler
  * nss_n2h_buf_handler()
  *	Add extra NSS bufs from host memory
  */
-static int nss_n2h_buf_cfg_core1_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_n2h_buf_cfg_core1_handler(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[NSS_CORE_1];
@@ -1406,14 +1406,14 @@ static int nss_n2h_buf_cfg_core1_handler
 	}
 
 	if ((nss_n2h_core1_add_buf_pool_size >= 1) && (nss_n2h_core1_add_buf_pool_size <= NSS_N2H_MAX_BUF_POOL_SIZE)) {
-		printk(KERN_INFO "configuring additional NSS pbufs\n");
+		dev_info(nss_ctx->dev, "Configuring additional NSS pbufs\n");
 		ret = nss_n2h_buf_pool_cfg(nss_ctx, nss_n2h_core1_add_buf_pool_size, NSS_CORE_1);
 		nss_n2h_core1_add_buf_pool_size = nss_ctx->buf_sz_allocated;
-		printk(KERN_INFO "additional pbufs of size %d got added to NSS\n", nss_ctx->buf_sz_allocated);
+		dev_info(nss_ctx->dev, "Additional pbufs of size %d got added to NSS\n", nss_ctx->buf_sz_allocated);
 		return ret;
 	}
 
-	printk(KERN_INFO "Invalid input value. should be greater than 1 and less than %d\n", NSS_N2H_MAX_BUF_POOL_SIZE);
+	dev_info(nss_ctx->dev, "Invalid input value. should be greater than 1 and less than %d\n", NSS_N2H_MAX_BUF_POOL_SIZE);
 	return -EINVAL;
 }
 
@@ -1435,7 +1435,7 @@ static void nss_n2h_queue_limit_callback
  * nss_n2h_set_queue_limit_sync()
  *	Sets the n2h queue size limit synchronously.
  */
-static int nss_n2h_set_queue_limit_sync(struct ctl_table *ctl, int write, void __user *buffer,
+static int nss_n2h_set_queue_limit_sync(const struct ctl_table *ctl, int write, void __user *buffer,
 					size_t *lenp, loff_t *ppos, uint32_t core_id)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
@@ -1512,7 +1512,7 @@ static int nss_n2h_set_queue_limit_sync(
  * nss_n2h_queue_limit_core0_handler()
  *	Sets the n2h queue size limit for core0
  */
-static int nss_n2h_queue_limit_core0_handler(struct ctl_table *ctl,
+static int nss_n2h_queue_limit_core0_handler(const struct ctl_table *ctl,
 				int write, void __user *buffer,
 				size_t *lenp, loff_t *ppos)
 {
@@ -1524,7 +1524,7 @@ static int nss_n2h_queue_limit_core0_han
  * nss_n2h_queue_limit_core1_handler()
  *	Sets the n2h queue size limit for core1
  */
-static int nss_n2h_queue_limit_core1_handler(struct ctl_table *ctl,
+static int nss_n2h_queue_limit_core1_handler(const struct ctl_table *ctl,
 				int write, void __user *buffer,
 				size_t *lenp, loff_t *ppos)
 {
@@ -1602,7 +1602,7 @@ static nss_tx_status_t nss_n2h_host_bp_c
  * nss_n2h_host_bp_cfg_handler()
  *	Enable n2h back pressure.
  */
-static int nss_n2h_host_bp_cfg_handler(struct ctl_table *ctl, int write,
+static int nss_n2h_host_bp_cfg_handler(const struct ctl_table *ctl, int write,
 				void __user *buffer, size_t *lenp, loff_t *ppos, uint32_t core_id)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
@@ -1640,7 +1640,7 @@ static int nss_n2h_host_bp_cfg_handler(s
  * nss_n2h_host_bp_cfg_core0_handler()
  *	Enable n2h back pressure in core 0.
  */
-static int nss_n2h_host_bp_cfg_core0_handler(struct ctl_table *ctl, int write,
+static int nss_n2h_host_bp_cfg_core0_handler(const struct ctl_table *ctl, int write,
 				void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	return nss_n2h_host_bp_cfg_handler(ctl, write, buffer, lenp, ppos, NSS_CORE_0);
@@ -1650,7 +1650,7 @@ static int nss_n2h_host_bp_cfg_core0_han
  * nss_n2h_host_bp_cfg_core1_handler()
  *	Enable n2h back pressure in core 1.
  */
-static int nss_n2h_host_bp_cfg_core1_handler(struct ctl_table *ctl, int write,
+static int nss_n2h_host_bp_cfg_core1_handler(const struct ctl_table *ctl, int write,
 				void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	return nss_n2h_host_bp_cfg_handler(ctl, write, buffer, lenp, ppos, NSS_CORE_1);
@@ -1740,9 +1740,7 @@ static struct ctl_table nss_n2h_table_si
 		.maxlen			= sizeof(int),
 		.mode			= 0644,
 		.proc_handler	= &nss_n2h_get_qos_mem_size_cfg_handler,
-	},
-
-	{ }
+	}
 };
 
 static struct ctl_table nss_n2h_table_multi_core[] = {
@@ -1900,8 +1898,7 @@ static struct ctl_table nss_n2h_table_mu
 		.maxlen			= sizeof(int),
 		.mode			= 0644,
 		.proc_handler	= &nss_n2h_get_qos_mem_size_cfg_handler,
-	},
-	{ }
+	}
 };
 
 
--- a/nss_n2h_stats.c
+++ b/nss_n2h_stats.c
@@ -46,6 +46,7 @@ static ssize_t nss_n2h_stats_read(struct
 	size_t size_wr = 0;
 	ssize_t bytes_read = 0;
 	uint64_t *stats_shadow;
+	char n2h_tag[7];
 
 	char *lbuf = kzalloc(size_al, GFP_KERNEL);
 	if (unlikely(lbuf == NULL)) {
@@ -69,7 +70,8 @@ static ssize_t nss_n2h_stats_read(struct
 			stats_shadow[i] = nss_n2h_stats[core][i];
 		}
 		spin_unlock_bh(&nss_top_main.stats_lock);
-		size_wr += nss_stats_banner(lbuf, size_wr, size_al, "n2h", core);
+		snprintf(n2h_tag, 7, "N2H %d", core);
+		size_wr += nss_stats_banner(lbuf, size_wr, size_al, n2h_tag, NSS_STATS_SINGLE_CORE);
 		size_wr += nss_stats_print("n2h", NULL, NSS_STATS_SINGLE_INSTANCE
 						, nss_n2h_strings_stats
 						, stats_shadow
--- a/nss_n2h_strings.c
+++ b/nss_n2h_strings.c
@@ -20,6 +20,7 @@
 #include "nss_core.h"
 #include <nss_n2h.h>
 #include "nss_strings.h"
+#include "nss_n2h_strings.h"
 
 /*
  * nss_n2h_strings_stats
--- a/nss_ppe.c
+++ b/nss_ppe.c
@@ -357,7 +357,7 @@ void nss_ppe_init(void)
 	/*
 	 * Get the PPE base address
 	 */
-	ppe_pvt.ppe_base = ioremap_nocache(PPE_BASE_ADDR, PPE_REG_SIZE);
+	ppe_pvt.ppe_base = nss_ioremap(PPE_BASE_ADDR, PPE_REG_SIZE);
 	if (!ppe_pvt.ppe_base) {
 		nss_warning("DRV can't get PPE base address\n");
 		return;
--- a/nss_ppe_vp.c
+++ b/nss_ppe_vp.c
@@ -20,6 +20,8 @@
  * Header file for qca-ssdk APIs
  */
 #include <ref/ref_vsi.h>
+#include <linux/vmalloc.h>
+
 
 #include "nss_ppe_vp.h"
 #include "nss_ppe_vp_stats.h"
@@ -96,8 +98,8 @@ static void nss_ppe_vp_proc_help(void)
 {
 	nss_info_always("== for dynamic interface types read following file ==");
 	nss_info_always("/sys/kernel/debug/qca-nss-drv/stats/dynamic_if/type_names");
-	nss_info_always("NSS PPE VP create: echo <interface name> <dynamic interface type> > /proc/sys/nss/ppe_vp/create");
-	nss_info_always("NSS PPE VP destroy: echo <interface name> <dynamic interface type> > /proc/sys/nss/ppe_vp/destroy");
+	nss_info_always("NSS PPE VP create: echo <interface name> <dynamic interface type> > /proc/sys/dev/nss/ppe_vp/create");
+	nss_info_always("NSS PPE VP destroy: echo <interface name> <dynamic interface type> > /proc/sys/dev/nss/ppe_vp/destroy");
 }
 
 /*
@@ -210,20 +212,19 @@ static void nss_ppe_vp_callback(void *ap
  * nss_ppe_vp_parse_vp_cmd()
  *	Parse PPE VP create and destroy message and return the NSS interface number.
  * Command usage:
- * echo <interface name> <dynamic interface type> /proc/sys/nss/ppe_vp/create>
- * echo ath0 6 > /proc/sys/nss/ppe_vp/create
+ * echo <interface name> <dynamic interface type> /proc/sys/dev/nss/ppe_vp/create>
+ * echo ath0 6 > /proc/sys/dev/nss/ppe_vp/create
  * Since ath0 has only one type i.e. ath0 is NSS_DYNAMIC_INTERFACE_TYPE_VAP, the above command can be rewritten as
- * echo ath0 > /proc/sys/nss/ppe_vp/create => Here 6 can be ignored.
+ * echo ath0 > /proc/sys/dev/nss/ppe_vp/create => Here 6 can be ignored.
  */
-static nss_if_num_t nss_ppe_vp_parse_vp_cmd(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static nss_if_num_t nss_ppe_vp_parse_vp_cmd(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	int32_t if_num;
 	struct net_device *dev;
-	uint32_t  dynamic_if_type = (uint32_t)NSS_DYNAMIC_INTERFACE_TYPE_NONE;
+	uint32_t dynamic_if_type = (uint32_t)NSS_DYNAMIC_INTERFACE_TYPE_NONE;
 	struct nss_ctx_instance *nss_ctx = nss_ppe_vp_get_context();
 	char *pos;
 	char cmd_buf[NSS_PPE_VP_MAX_CMD_STR] = {0}, dev_name[NSS_PPE_VP_MAX_CMD_STR] = {0};
-	size_t count = *lenp;
 	int ret = proc_dostring(ctl, write, buffer, lenp, ppos);
 
 	if (!write) {
@@ -238,17 +239,7 @@ static nss_if_num_t nss_ppe_vp_parse_vp_
 
 	NSS_VERIFY_CTX_MAGIC(nss_ctx);
 
-	if (count >= NSS_PPE_VP_MAX_CMD_STR) {
-		nss_ppe_vp_proc_help();
-		nss_warning("%px: Input string too big", nss_ctx);
-		return -E2BIG;
-	}
-
-	if (copy_from_user(cmd_buf, buffer, count)) {
-		nss_warning("%px: Cannot copy user's entry to kernel memory", nss_ctx);
-		return -EFAULT;
-	}
-
+	/* proc_dostring has already copied the data to nss_ppe_vp_cmd */
 	if ((pos = strrchr(cmd_buf, '\n')) != NULL) {
 		*pos = '\0';
 	}
@@ -288,7 +279,7 @@ static nss_if_num_t nss_ppe_vp_parse_vp_
  * nss_ppe_vp_tx_msg()
  *	Transmit a ppe_vp message to NSS FW
  */
-nss_tx_status_t nss_ppe_vp_tx_msg(struct nss_ctx_instance *nss_ctx, struct nss_ppe_vp_msg *msg)
+static nss_tx_status_t nss_ppe_vp_tx_msg(struct nss_ctx_instance *nss_ctx, struct nss_ppe_vp_msg *msg)
 {
 	struct nss_cmn_msg *ncm = &msg->cm;
 	nss_if_num_t if_num = ncm->interface;
@@ -318,7 +309,7 @@ nss_tx_status_t nss_ppe_vp_tx_msg(struct
  * nss_ppe_vp_tx_msg_sync()
  *	Transmit a ppe_vp message to NSS firmware synchronously.
  */
-nss_tx_status_t nss_ppe_vp_tx_msg_sync(struct nss_ctx_instance *nss_ctx, struct nss_ppe_vp_msg *npvm)
+static nss_tx_status_t nss_ppe_vp_tx_msg_sync(struct nss_ctx_instance *nss_ctx, struct nss_ppe_vp_msg *npvm)
 {
 	nss_tx_status_t status;
 	int ret = 0;
@@ -693,7 +684,7 @@ static void nss_ppe_vp_handler(struct ns
  * nss_ppe_vp_destroy_handler()
  *	PPE VP destroy handler.
  */
-static int nss_ppe_vp_destroy_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_ppe_vp_destroy_handler(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_ctx_instance *nss_ctx = nss_ppe_vp_get_context();
 	int32_t if_num;
@@ -730,7 +721,7 @@ static int nss_ppe_vp_destroy_handler(st
  * nss_ppe_vp_create_handler()
  *	PPE VP create handler.
  */
-static int nss_ppe_vp_create_handler(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_ppe_vp_create_handler(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	int32_t if_num;
 	struct nss_ctx_instance *nss_ctx = nss_ppe_vp_get_context();
@@ -779,26 +770,7 @@ static struct ctl_table nss_ppe_vp_table
 		.maxlen         = sizeof(nss_ppe_vp_cmd),
 		.mode           = 0644,
 		.proc_handler   = &nss_ppe_vp_destroy_handler,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ppe_vp_dir[] = {
-	{
-		.procname	= "ppe_vp",
-		.mode		= 0555,
-		.child		= nss_ppe_vp_table,
-	},
-	{ }
-};
-
-static struct ctl_table nss_ppe_vp_root_dir[] = {
-	{
-		.procname	= "nss",
-		.mode		= 0555,
-		.child		= nss_ppe_vp_dir,
-	},
-	{ }
+	}
 };
 
 static struct ctl_table_header *nss_ppe_vp_procfs_header;
@@ -807,19 +779,19 @@ static struct ctl_table_header *nss_ppe_
  * nss_ppe_vp_procfs_register()
  *	Register sysctl specific to ppe_vp
  */
-void nss_ppe_vp_procfs_register(void)
+static void nss_ppe_vp_procfs_register(void)
 {
 	/*
 	 * Register sysctl table.
 	 */
-	nss_ppe_vp_procfs_header = register_sysctl_table(nss_ppe_vp_root_dir);
+	nss_ppe_vp_procfs_header = register_sysctl("dev/nss/ppe_vp", nss_ppe_vp_table);
 }
 
 /*
  * uss_ppe_vp_procfs_unregister()
  *	Unregister sysctl specific for ppe_vp
  */
-void nss_ppe_vp_procfs_unregister(void)
+static void nss_ppe_vp_procfs_unregister(void)
 {
 	/*
 	 * Unregister sysctl table.
--- a/nss_profiler.c
+++ b/nss_profiler.c
@@ -209,11 +209,12 @@ EXPORT_SYMBOL(nss_profile_dma_deregister
 struct nss_profile_sdma_ctrl *nss_profile_dma_get_ctrl(struct nss_ctx_instance *nss_ctx)
 {
 	struct nss_profile_sdma_ctrl *ctrl = nss_ctx->meminfo_ctx.sdma_ctrl;
+	int size = offsetof(struct nss_profile_sdma_ctrl, cidx);
 	if (!ctrl) {
 		return ctrl;
 	}
 
-	dmac_inv_range(ctrl, &ctrl->cidx);
+	dma_sync_single_for_cpu(nss_ctx->dev, (dma_addr_t) ctrl, size, DMA_FROM_DEVICE);
 	dsb(sy);
 	return ctrl;
 }
--- a/nss_qrfs_log.c
+++ b/nss_qrfs_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_qrfs_log.h"
 #define NSS_QRFS_LOG_IPV4 4
 #define NSS_QRFS_LOG_IPV6 6
 
--- a/nss_rps.c
+++ b/nss_rps.c
@@ -70,7 +70,7 @@ static inline void nss_rps_pri_map_usage
  * nss_rps_pri_map_print()
  *	Sysctl handler for printing rps/pri mapping.
  */
-static int nss_rps_pri_map_print(struct ctl_table *ctl, void __user *buffer,
+static int nss_rps_pri_map_print(void *buffer,
 				size_t *lenp, loff_t *ppos, int *pri_map)
 {
 	char *r_buf;
@@ -109,7 +109,7 @@ static int nss_rps_pri_map_print(struct
 	len = scnprintf(r_buf + cp_bytes, 4, "\n");
 	cp_bytes += len;
 
-	cp_bytes = simple_read_from_buffer(buffer, *lenp, ppos, r_buf, cp_bytes);
+	cp_bytes = memory_read_from_buffer(buffer, *lenp, ppos, r_buf, cp_bytes);
 	*lenp = cp_bytes;
 	kfree(r_buf);
 	return 0;
@@ -119,13 +119,10 @@ static int nss_rps_pri_map_print(struct
  * nss_rps_pri_map_parse()
  *	Sysctl handler for rps/pri mappings.
  */
-static int nss_rps_pri_map_parse(struct ctl_table *ctl, void __user *buffer,
-	size_t *lenp, loff_t *ppos, struct nss_rps_pri_map_parse_data *out)
+static int nss_rps_pri_map_parse(void *buffer,
+	size_t *lenp, struct nss_rps_pri_map_parse_data *out)
 {
-	size_t cp_bytes = 0;
 	char w_buf[5];
-	loff_t w_offset = 0;
-	char *str;
 	unsigned int pri;
 	int core, res;
 
@@ -140,14 +137,15 @@ static int nss_rps_pri_map_parse(struct
 	/*
 	 * It's a write operation
 	 */
-	cp_bytes = simple_write_to_buffer(w_buf, *lenp, &w_offset, buffer, 5);
-	if (cp_bytes != *lenp) {
-		nss_warning("failed to write to buffer\n");
-		return -EFAULT;
+	if (*lenp >= sizeof(w_buf)) {
+		nss_warning("Input too large: %zu\n", *lenp);
+		return -EINVAL;
 	}
 
-	str = w_buf;
-	res = sscanf(str, "%u %d", &pri, &core);
+	memcpy(w_buf, buffer, *lenp);
+	w_buf[*lenp] = '\0'; /* Ensure null termination */
+
+	res = sscanf(w_buf, "%u %d", &pri, &core);
 	if (res != NSS_RPS_PRI_MAP_PARAM_FIELD_COUNT) {
 		nss_warning("failed to read the buffer\n");
 		return -EFAULT;
@@ -406,8 +404,8 @@ static nss_tx_status_t nss_rps_pri_map_c
  * nss_rps_cfg_handler()
  *	Enable NSS RPS.
  */
-static int nss_rps_cfg_handler(struct ctl_table *ctl, int write,
-				void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_rps_cfg_handler(const struct ctl_table *ctl, int write,
+				void *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx;
@@ -457,8 +455,8 @@ static int nss_rps_cfg_handler(struct ct
  * nss_rps_hash_bitmap_cfg_handler()
  *	Configure NSS rps_hash_bitmap
  */
-static int nss_rps_hash_bitmap_cfg_handler(struct ctl_table *ctl, int write,
-				void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_rps_hash_bitmap_cfg_handler(const struct ctl_table *ctl, int write,
+				void *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx __attribute__((unused)) = &nss_top->nss[0];
@@ -520,8 +518,8 @@ static int nss_rps_hash_bitmap_cfg_handl
 /* nss_rps_pri_map_cfg_handler()
  *	Configure NSS rps_pri_map
  */
-static int nss_rps_pri_map_cfg_handler(struct ctl_table *ctl, int write,
-				void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_rps_pri_map_cfg_handler(const struct ctl_table *ctl, int write,
+				void *buffer, size_t *lenp, loff_t *ppos)
 {
 	struct nss_top_instance *nss_top = &nss_top_main;
 	struct nss_ctx_instance *nss_ctx = &nss_top->nss[0];
@@ -529,10 +527,10 @@ static int nss_rps_pri_map_cfg_handler(s
 	int ret, ret_pri_map;
 	struct nss_rps_pri_map_parse_data out, current_state;
 	if (!write) {
-		return nss_rps_pri_map_print(ctl, buffer, lenp, ppos, nss_rps_pri_map);
+		return nss_rps_pri_map_print(buffer, lenp, ppos, nss_rps_pri_map);
 	}
 
-	ret = nss_rps_pri_map_parse(ctl, buffer, lenp, ppos, &out);
+	ret = nss_rps_pri_map_parse(buffer, lenp, &out);
 
 	if (ret != NSS_SUCCESS) {
 		nss_rps_pri_map_usage();
@@ -573,8 +571,7 @@ static struct ctl_table nss_rps_table[]
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &nss_rps_pri_map_cfg_handler,
-	},
-	{ }
+	}
 };
 
 
@@ -613,7 +610,7 @@ static struct ctl_table_header *nss_rps_
  * nss_rps_pri_map_init_handler()
  *	Initialize pri_map for priority based rps selection.
  */
-void nss_rps_pri_map_init_handler(void)
+static void nss_rps_pri_map_init_handler(void)
 {
 	int i;
 
--- a/nss_virt_if.c
+++ b/nss_virt_if.c
@@ -450,6 +450,8 @@ EXPORT_SYMBOL(nss_virt_if_create_sync_ne
  */
 struct nss_virt_if_handle *nss_virt_if_create_sync(struct net_device *netdev)
 {
+	if (!nss_ctl_redirect)
+	  return NULL;
 	/*
 	 * NSS_N2H_INTERFACE is the nexthop of the dynamic interface which is created for handling the
 	 * n2h traffic.
--- a/nss_virt_if_stats.c
+++ b/nss_virt_if_stats.c
@@ -133,7 +133,7 @@ static int32_t nss_virt_if_interface_sta
  * nss_virt_if_stats_get()
  *	Get virt_if base node stats or interface stats by interface number.
  */
-bool nss_virt_if_stats_get(struct nss_ctx_instance *nss_ctx, uint32_t if_num, void *stats, bool is_base)
+static bool nss_virt_if_stats_get(struct nss_ctx_instance *nss_ctx, uint32_t if_num, void *stats, bool is_base)
 {
 	if (nss_virt_if_verify_if_num(if_num) == false) {
 		return false;
--- a/nss_vlan.c
+++ b/nss_vlan.c
@@ -246,7 +246,7 @@ EXPORT_SYMBOL(nss_vlan_tx_set_mtu_msg);
  * nss_vlan_tx_set_mac_addr_msg
  *	API to send change mac addr message to NSS FW
  */
-nss_tx_status_t nss_vlan_tx_set_mac_addr_msg(uint32_t vlan_if_num, uint8_t *addr)
+nss_tx_status_t nss_vlan_tx_set_mac_addr_msg(uint32_t vlan_if_num, const uint8_t *addr)
 {
 	struct nss_ctx_instance *nss_ctx = nss_vlan_get_context();
 	struct nss_vlan_msg nvm;
--- a/nss_vlan_log.c
+++ b/nss_vlan_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_vlan_log.h"
 
 /*
  * nss_vlan_log_message_types_str
