--- a/nss_crypto_cmn_log.c
+++ b/nss_crypto_cmn_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_crypto_cmn_log.h"
 
 /*
  * nss_crypto_cmn_log_message_types_str
--- a/nss_crypto_cmn_stats.c
+++ b/nss_crypto_cmn_stats.c
@@ -17,6 +17,8 @@
  **************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include "nss_core.h"
 #include "nss_crypto_cmn_stats.h"
 #include "nss_crypto_cmn_strings.h"
--- a/nss_crypto_log.c
+++ b/nss_crypto_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_crypto_log.h"
 
 /*
  * nss_crypto_log_message_types_str
--- a/nss_dtls_cmn_log.c
+++ b/nss_dtls_cmn_log.c
@@ -23,6 +23,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_dtls_cmn_log.h"
 
 /*
  * nss_dtls_cmn_log_message_types_str
--- a/nss_dtls_cmn_stats.c
+++ b/nss_dtls_cmn_stats.c
@@ -17,6 +17,8 @@
  ***************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include "nss_core.h"
 #include "nss_dtls_cmn.h"
 #include "nss_dtls_cmn_stats.h"
--- a/nss_dtls_log.c
+++ b/nss_dtls_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_dtls_log.h"
 
 /*
  * nss_dtls_log_message_types_str
--- a/nss_ipsec_cmn_stats.c
+++ b/nss_ipsec_cmn_stats.c
@@ -17,6 +17,8 @@
  ***************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include "nss_core.h"
 #include "nss_ipsec_cmn.h"
 #include "nss_ipsec_cmn_stats.h"
--- a/nss_sjack_log.c
+++ b/nss_sjack_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_sjack_log.h"
 
 /*
  * nss_sjack_log_message_types_str
--- a/nss_tls_log.c
+++ b/nss_tls_log.c
@@ -22,6 +22,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_tls_log.h"
 
 /*
  * nss_tls_log_message_types_str
--- a/nss_tls_stats.c
+++ b/nss_tls_stats.c
@@ -17,6 +17,8 @@
  ******************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include "nss_core.h"
 #include "nss_tls.h"
 #include "nss_tls_stats.h"
--- a/nss_trustsec_rx_log.c
+++ b/nss_trustsec_rx_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_trustsec_rx_log.h"
 
 /*
  * nss_trustsec_rx_log_message_types_str
--- a/nss_trustsec_tx_log.c
+++ b/nss_trustsec_tx_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_trustsec_tx_log.h"
 
 /*
  * nss_trustsec_tx_log_message_types_str
