--- a/nss_capwap_log.c
+++ b/nss_capwap_log.c
@@ -23,6 +23,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_capwap_log.h"
 
 /*
  * nss_capwap_log_message_types_str
--- a/nss_clmap_log.c
+++ b/nss_clmap_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_clmap_log.h"
 
 /*
  * nss_clmap_log_message_types_str
--- a/nss_dynamic_interface_log.c
+++ b/nss_dynamic_interface_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_dynamic_interface_log.h"
 
 /*
  * nss_dynamic_interface_log_message_types_str
--- a/nss_dynamic_interface_stats.c
+++ b/nss_dynamic_interface_stats.c
@@ -19,6 +19,7 @@
 
 #include "nss_tx_rx_common.h"
 #include "nss_dynamic_interface.h"
+#include "nss_dynamic_interface_stats.h"
 
 /*
  * nss_dynamic_interface_type_names
@@ -88,8 +89,15 @@ const char *nss_dynamic_interface_type_n
 	"NSS_DYNAMIC_INTERFACE_TYPE_RMNET_RX_H2N",
 	"NSS_DYNAMIC_INTERFACE_TYPE_WIFILI_EXTERNAL0",
 	"NSS_DYNAMIC_INTERFACE_TYPE_WIFILI_EXTERNAL1",
+	"NSS_DYNAMIC_INTERFACE_TYPE_TLS_INNER",
+	"NSS_DYNAMIC_INTERFACE_TYPE_TLS_OUTER",
+	"NSS_DYNAMIC_INTERFACE_TYPE_MIRROR",
+	"NSS_DYNAMIC_INTERFACE_TYPE_WIFI_EXT_VDEV_WDS",
 	"NSS_DYNAMIC_INTERFACE_TYPE_CAPWAP_HOST_INNER",
 	"NSS_DYNAMIC_INTERFACE_TYPE_CAPWAP_OUTER",
+	"NSS_DYNAMIC_INTERFACE_TYPE_WIFI_EXT_VDEV_VLAN",
+	"NSS_DYNAMIC_INTERFACE_TYPE_WIFI_MESH_INNER",
+	"NSS_DYNAMIC_INTERFACE_TYPE_WIFI_MESH_OUTER"
 };
 
 /*
--- a/nss_lso_rx.c
+++ b/nss_lso_rx.c
@@ -20,6 +20,7 @@
  */
 
 #include <nss_core.h>
+#include "nss_tx_rx_common.h"
 #include "nss_lso_rx_stats.h"
 #include "nss_lso_rx_strings.h"
 
--- a/nss_oam_log.c
+++ b/nss_oam_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_oam_log.h"
 
 /*
  * nss_oam_log_message_types_str
--- a/nss_portid_log.c
+++ b/nss_portid_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_portid_log.h"
 
 /*
  * nss_portid_log_message_types_str
--- a/nss_rmnet_rx.c
+++ b/nss_rmnet_rx.c
@@ -44,7 +44,7 @@ struct nss_rmnet_rx_handle *rmnet_rx_han
  * nss_rmnet_rx_verify_if_num()
  *	Verify if_num passed to us.
  */
-bool nss_rmnet_rx_verify_if_num(uint32_t if_num)
+static bool nss_rmnet_rx_verify_if_num(uint32_t if_num)
 {
 	enum nss_dynamic_interface_type type = nss_dynamic_interface_get_type(nss_rmnet_rx_get_context(), if_num);
 
--- a/nss_wifi_ext_vdev_log.c
+++ b/nss_wifi_ext_vdev_log.c
@@ -22,6 +22,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_wifi_ext_vdev_log.h"
 
 #define NSS_WIFI_EXT_VDEV_LOG_MESSAGE_TYPE_INDEX(type) ((type) - NSS_IF_MAX_MSG_TYPES)
 
--- a/nss_wifi_log.c
+++ b/nss_wifi_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_wifi_log.h"
 
 /*
  * nss_wifi_log_message_types_str
--- a/nss_wifi_mac_db.c
+++ b/nss_wifi_mac_db.c
@@ -17,6 +17,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_tx_rx_common.h"
 #include "nss_wifi_mac_db_if.h"
 
 /*
--- a/nss_wifi_mesh_log.c
+++ b/nss_wifi_mesh_log.c
@@ -23,6 +23,7 @@
 
 #include "nss_core.h"
 #include "nss_wifi_mesh.h"
+#include "nss_wifi_mesh_log.h"
 
 #define NSS_WIFI_MESH_LOG_MESSAGE_TYPE_INDEX(type) ((type) - NSS_IF_MAX_MSG_TYPES)
 
--- a/nss_wifi_mesh_strings.c
+++ b/nss_wifi_mesh_strings.c
@@ -80,8 +80,8 @@ struct nss_stats_info nss_wifi_mesh_stri
 	{"not_found",			NSS_STATS_TYPE_SPECIAL},
 	{"delete_success",		NSS_STATS_TYPE_SPECIAL},
 	{"update_success",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_expired",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_refresh_needed",	NSS_STATS_TYPE_SPECIAL},
+	{"path_expired",		NSS_STATS_TYPE_SPECIAL},
+	{"path_refresh_needed",	NSS_STATS_TYPE_SPECIAL},
 	{"add_requests",		NSS_STATS_TYPE_SPECIAL},
 	{"del_requests",		NSS_STATS_TYPE_SPECIAL},
 	{"update_requests",		NSS_STATS_TYPE_SPECIAL},
@@ -91,8 +91,8 @@ struct nss_stats_info nss_wifi_mesh_stri
 	{"metric_updations",		NSS_STATS_TYPE_SPECIAL},
 	{"block_mesh_fwd_updations",	NSS_STATS_TYPE_SPECIAL},
 	{"delete_failures",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_dummy_add_failures",NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_dummy_add_success",	NSS_STATS_TYPE_SPECIAL}
+	{"path_dummy_add_failures",	NSS_STATS_TYPE_SPECIAL},
+	{"path_dummy_add_success",	NSS_STATS_TYPE_SPECIAL}
 
 };
 
@@ -126,10 +126,10 @@ struct nss_stats_info nss_wifi_mesh_stri
 	{"update_requests",				NSS_STATS_TYPE_SPECIAL},
 	{"mda_updations",				NSS_STATS_TYPE_SPECIAL},
 	{"flag_updations",				NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_lookup_success",	NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_lookup_failures",	NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_add_failures",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_add_success",		NSS_STATS_TYPE_SPECIAL}
+	{"proxy_path_dummy_lookup_success",	NSS_STATS_TYPE_SPECIAL},
+	{"proxy_path_dummy_lookup_failures",	NSS_STATS_TYPE_SPECIAL},
+	{"proxy_path_dummy_add_failures",		NSS_STATS_TYPE_SPECIAL},
+	{"proxy_path_dummy_add_success",		NSS_STATS_TYPE_SPECIAL}
 };
 
 /*
--- a/nss_wifi_vdev.c
+++ b/nss_wifi_vdev.c
@@ -274,7 +274,7 @@ EXPORT_SYMBOL(nss_wifi_vdev_base_set_nex
 /*
  * nss_wifi_vdev_set_peer_next_hop()
  */
-nss_tx_status_t nss_wifi_vdev_set_peer_next_hop(struct nss_ctx_instance *ctx, uint32_t nss_if, uint8_t *addr, uint32_t next_hop_if)
+nss_tx_status_t nss_wifi_vdev_set_peer_next_hop(struct nss_ctx_instance *ctx, uint32_t nss_if, const uint8_t *addr, uint32_t next_hop_if)
 {
 	nss_tx_status_t status;
 	struct nss_wifi_vdev_msg *wifivdevmsg = kzalloc(sizeof(struct nss_wifi_vdev_msg), GFP_KERNEL);
--- a/nss_wifili_log.c
+++ b/nss_wifili_log.c
@@ -20,6 +20,7 @@
  */
 
 #include "nss_core.h"
+#include "nss_wifili_log.h"
 
 /*
  * nss_wifili_log_message_types_str
--- a/nss_wifili_stats.c
+++ b/nss_wifili_stats.c
@@ -43,7 +43,7 @@ struct nss_wifili_soc_stats soc_stats[NS
  * nss_wifili_target_type_string()
  *	Convert Target Type Integer to String
  */
-void nss_wifili_target_type_to_string(uint32_t target_type, char *target_type_str)
+static void nss_wifili_target_type_to_string(uint32_t target_type, char *target_type_str)
 {
 	switch (target_type) {
 
