--- a/nss_shaper.c
+++ b/nss_shaper.c
@@ -80,7 +80,7 @@ void *nss_shaper_register_shaper_bounce_
 	 */
 	reg = &nss_top->bounce_interface_registrants[if_num];
 	if (reg->registered) {
-		spin_unlock_bh(&nss_top->stats_lock);
+		spin_unlock_bh(&nss_top->lock);
 		module_put(owner);
 		nss_warning("Already registered: %u", if_num);
 		BUG_ON(false);
@@ -125,7 +125,7 @@ void nss_shaper_unregister_shaper_bounce
 	 */
 	reg = &nss_top->bounce_interface_registrants[if_num];
 	if (!reg->registered) {
-		spin_unlock_bh(&nss_top->stats_lock);
+		spin_unlock_bh(&nss_top->lock);
 		nss_warning("Already unregistered: %u", if_num);
 		BUG_ON(false);
 	}
@@ -185,7 +185,7 @@ void *nss_shaper_register_shaper_bounce_
 	 */
 	reg = &nss_top->bounce_bridge_registrants[if_num];
 	if (reg->registered) {
-		spin_unlock_bh(&nss_top->stats_lock);
+		spin_unlock_bh(&nss_top->lock);
 		module_put(owner);
 		nss_warning("Already registered: %u", if_num);
 		BUG_ON(false);
@@ -231,7 +231,7 @@ void nss_shaper_unregister_shaper_bounce
 	 */
 	reg = &nss_top->bounce_bridge_registrants[if_num];
 	if (!reg->registered) {
-		spin_unlock_bh(&nss_top->stats_lock);
+		spin_unlock_bh(&nss_top->lock);
 		nss_warning("Already unregistered: %u", if_num);
 		BUG_ON(false);
 	}
@@ -240,9 +240,9 @@ void nss_shaper_unregister_shaper_bounce
 	 * Wait until any bounce callback that is active is finished
 	 */
 	while (reg->callback_active) {
-		spin_unlock_bh(&nss_top->stats_lock);
+		spin_unlock_bh(&nss_top->lock);
 		yield();
-		spin_lock_bh(&nss_top->stats_lock);
+		spin_lock_bh(&nss_top->lock);
 	}
 
 	/*
@@ -283,7 +283,7 @@ nss_tx_status_t nss_shaper_bounce_interf
 	spin_lock_bh(&nss_top->lock);
 	reg = &nss_top->bounce_interface_registrants[if_num];
 	if (!reg->registered) {
-		spin_unlock_bh(&nss_top->stats_lock);
+		spin_unlock_bh(&nss_top->lock);
 		nss_warning("unregistered: %u", if_num);
 		return NSS_TX_FAILURE;
 	}
@@ -327,7 +327,7 @@ nss_tx_status_t nss_shaper_bounce_bridge
 	spin_lock_bh(&nss_top->lock);
 	reg = &nss_top->bounce_bridge_registrants[if_num];
 	if (!reg->registered) {
-		spin_unlock_bh(&nss_top->stats_lock);
+		spin_unlock_bh(&nss_top->lock);
 		nss_warning("unregistered: %u", if_num);
 		return NSS_TX_FAILURE;
 	}
