--- a/nss_dp_main.c
+++ b/nss_dp_main.c
@@ -956,7 +956,7 @@ static void nss_dp_phy_disconnect(struct
 {
 	struct device_node *np = dp_priv->pdev->dev.of_node;
 
-	if (!dp_priv->phydev)
+	if (IS_ERR_OR_NULL(dp_priv->phydev))
 		return;
 
 	phy_disconnect(dp_priv->phydev);
@@ -1066,6 +1066,14 @@ static int32_t nss_dp_probe(struct platf
 
 	dp_priv->drv_flags |= NSS_DP_PRIV_FLAG(INIT_DONE);
 
+	/* Register the network interface */
+	ret = register_netdev(netdev);
+	if (ret) {
+		netdev_dbg(netdev, "Error registering netdevice %s\n",
+				netdev->name);
+		goto netdev_register_fail;
+	}
+
 	if (of_find_property(np, "phy-handle", NULL) || of_phy_is_fixed_link(np)) {
 		dp_priv->phydev = of_phy_get_and_connect(netdev, np,
 				&nss_dp_adjust_link);
@@ -1091,6 +1099,7 @@ static int32_t nss_dp_probe(struct platf
 				&nss_dp_adjust_link,
 				dp_priv->phy_mii_type);
 		if (IS_ERR(dp_priv->phydev)) {
+			dp_priv->phydev = NULL;
 			netdev_dbg(netdev, "failed to connect to phy device\n");
 			goto phy_setup_fail;
 		}
@@ -1117,14 +1126,6 @@ static int32_t nss_dp_probe(struct platf
 
 	/* TODO: Features: CSUM, tx/rx offload... configure */
 
-	/* Register the network interface */
-	ret = register_netdev(netdev);
-	if (ret) {
-		netdev_dbg(netdev, "Error registering netdevice %s\n",
-								netdev->name);
-		goto phy_setup_fail;
-	}
-
 	dp_global_ctx.nss_dp[nss_dp_get_idx_from_macid(dp_priv->macid)] = dp_priv;
 	dp_global_ctx.slowproto_acl_bm = 0;
 
@@ -1158,6 +1159,8 @@ static int32_t nss_dp_probe(struct platf
 
 phy_setup_fail:
 	nss_dp_phy_disconnect(dp_priv);
+	unregister_netdev(netdev);
+netdev_register_fail:
 	dp_priv->data_plane_ops->deinit(dp_priv->dpc);
 data_plane_init_fail:
 	dp_priv->gmac_hal_ops->exit(dp_priv->gmac_hal_ctx);
