--- a/frontends/nss/ecm_nss_common.c
+++ b/frontends/nss/ecm_nss_common.c
@@ -67,6 +67,7 @@ bool ecm_nss_ipv6_is_conn_limit_reached(
 
 	return false;
 }
+EXPORT_SYMBOL(ecm_nss_ipv6_is_conn_limit_reached);
 #endif
 
 /*
@@ -116,3 +117,4 @@ bool ecm_nss_ipv4_is_conn_limit_reached(
 
 	return false;
 }
+EXPORT_SYMBOL(ecm_nss_ipv4_is_conn_limit_reached);
--- a/frontends/nss/ecm_nss_common.h
+++ b/frontends/nss/ecm_nss_common.h
@@ -278,7 +278,7 @@ static inline bool ecm_nss_common_igs_ac
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
 		if (likely(!(to_dev->ingress_cl_list))) {
 #else
-		if (likely(!(to_dev->miniq_ingress))) {
+		if (likely(!(to_dev->tcx_ingress))) {
 #endif
 			dev_put(to_dev);
 			continue;
--- a/frontends/nss/ecm_nss_ipv4.c
+++ b/frontends/nss/ecm_nss_ipv4.c
@@ -32,7 +32,7 @@
 #include <net/ip.h>
 #include <net/tcp.h>
 #include <net/addrconf.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
@@ -610,7 +610,8 @@ sync_conntrack:
 #else
 				timeouts = nf_ct_timeout_lookup(ct);
 				if (!timeouts) {
-					timeouts = udp_get_timeouts(nf_ct_net(ct));
+					struct nf_udp_net *un = nf_udp_pernet(nf_ct_net(ct));
+					timeouts = un->timeouts;
 				}
 
 				spin_lock_bh(&ct->lock);
@@ -700,7 +701,7 @@ static void ecm_nss_ipv4_stats_sync_req_
 	}
 	spin_unlock_bh(&ecm_nss_ipv4_lock);
 
-	usleep_range(ECM_NSS_IPV4_STATS_SYNC_UDELAY - 100, ECM_NSS_IPV4_STATS_SYNC_UDELAY);
+	msleep_interruptible(ECM_NSS_IPV4_STATS_SYNC_UDELAY / 1000);
 
 	/*
 	 * If index is 0, we are starting a new round, but if we still have time remain
@@ -714,7 +715,7 @@ static void ecm_nss_ipv4_stats_sync_req_
 		}
 
 		if (time_after(ecm_nss_ipv4_next_req_time, current_jiffies)) {
-			msleep(jiffies_to_msecs(ecm_nss_ipv4_next_req_time - current_jiffies));
+			msleep_interruptible(jiffies_to_msecs(ecm_nss_ipv4_next_req_time - current_jiffies));
 		}
 		ecm_nss_ipv4_roll_check_jiffies = jiffies;
 		ecm_nss_ipv4_next_req_time = ecm_nss_ipv4_roll_check_jiffies + ECM_NSS_IPV4_STATS_SYNC_PERIOD;
--- a/frontends/nss/ecm_nss_ipv6.c
+++ b/frontends/nss/ecm_nss_ipv6.c
@@ -33,7 +33,7 @@
 #include <net/addrconf.h>
 #include <net/ipv6.h>
 #include <net/tcp.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
@@ -587,7 +587,8 @@ sync_conntrack:
 #else
 				timeouts = nf_ct_timeout_lookup(ct);
 				if (!timeouts) {
-					timeouts = udp_get_timeouts(nf_ct_net(ct));
+					struct nf_udp_net *un = nf_udp_pernet(nf_ct_net(ct));
+					timeouts = un->timeouts;
 				}
 
 				spin_lock_bh(&ct->lock);
@@ -676,7 +677,7 @@ static void ecm_nss_ipv6_stats_sync_req_
 	}
 	spin_unlock_bh(&ecm_nss_ipv6_lock);
 
-	usleep_range(ECM_NSS_IPV6_STATS_SYNC_UDELAY - 100, ECM_NSS_IPV6_STATS_SYNC_UDELAY);
+	msleep_interruptible(ECM_NSS_IPV6_STATS_SYNC_UDELAY / 1000);
 
 	/*
 	 * If index is 0, we are starting a new round, but if we still have time remain
@@ -690,7 +691,7 @@ static void ecm_nss_ipv6_stats_sync_req_
 		}
 
 		if (time_after(ecm_nss_ipv6_next_req_time, current_jiffies)) {
-			msleep(jiffies_to_msecs(ecm_nss_ipv6_next_req_time - current_jiffies));
+			msleep_interruptible(jiffies_to_msecs(ecm_nss_ipv6_next_req_time - current_jiffies));
 		}
 		ecm_nss_ipv6_roll_check_jiffies = jiffies;
 		ecm_nss_ipv6_next_req_time = ecm_nss_ipv6_roll_check_jiffies + ECM_NSS_IPV6_STATS_SYNC_PERIOD;
--- a/frontends/nss/ecm_nss_multicast_ipv4.c
+++ b/frontends/nss/ecm_nss_multicast_ipv4.c
@@ -34,7 +34,7 @@
 #include <net/route.h>
 #include <net/ip.h>
 #include <net/addrconf.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
--- a/frontends/nss/ecm_nss_multicast_ipv6.c
+++ b/frontends/nss/ecm_nss_multicast_ipv6.c
@@ -33,7 +33,7 @@
 #include <net/addrconf.h>
 #include <net/ipv6.h>
 #include <net/tcp.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
@@ -1010,6 +1010,7 @@ static void ecm_nss_multicast_ipv6_conne
 			create->ingress_vlan_tag[interface_type_counts[ii_type]] = ((vlan_info.vlan_tpid << 16) | vlan_info.vlan_tag);
 			create->valid_flags |= NSS_IPV6_MC_RULE_CREATE_FLAG_INGRESS_VLAN_VALID;
 			DEBUG_TRACE("%px: vlan tag: %x\n", feci, create->ingress_vlan_tag[interface_type_counts[ii_type]]);
+			fallthrough;
 #else
 			rule_invalid = true;
 			DEBUG_TRACE("%px: VLAN - unsupported\n", feci);
--- a/frontends/nss/ecm_nss_non_ported_ipv4.c
+++ b/frontends/nss/ecm_nss_non_ported_ipv4.c
@@ -32,7 +32,7 @@
 #include <net/ip.h>
 #include <net/tcp.h>
 #include <net/addrconf.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
@@ -900,6 +900,7 @@ static void ecm_nss_non_ported_ipv4_conn
 			DEBUG_TRACE("%px: PPTP - unsupported\n", feci);
 #endif
 			break;
+#ifdef NSS_FIRMWARE_VERSION_12_5
 		case ECM_DB_IFACE_TYPE_RAWIP:
 #ifdef ECM_INTERFACE_RAWIP_ENABLE
 			nircm->valid_flags |= NSS_IPV4_RULE_CREATE_RAWIP_VALID;
@@ -908,6 +909,7 @@ static void ecm_nss_non_ported_ipv4_conn
 			DEBUG_TRACE("%px: RAWIP - unsupported\n", feci);
 #endif
 			break;
+#endif
 		default:
 			DEBUG_TRACE("%px: Ignoring: %d (%s)\n", feci, ii_type, ii_name);
 		}
@@ -1142,6 +1144,7 @@ static void ecm_nss_non_ported_ipv4_conn
 			DEBUG_TRACE("%px: IPSEC - unsupported\n", feci);
 #endif
 			break;
+#ifdef NSS_FIRMWARE_VERSION_12_5
 		case ECM_DB_IFACE_TYPE_RAWIP:
 #ifdef ECM_INTERFACE_RAWIP_ENABLE
 			nircm->valid_flags |= NSS_IPV4_RULE_CREATE_RAWIP_VALID;
@@ -1150,6 +1153,7 @@ static void ecm_nss_non_ported_ipv4_conn
 			DEBUG_TRACE("%px: RAWIP - unsupported\n", feci);
 #endif
 			break;
+#endif
 		default:
 			DEBUG_TRACE("%px: Ignoring: %d (%s)\n", feci, ii_type, ii_name);
 		}
@@ -1918,6 +1922,7 @@ struct ecm_front_end_connection_instance
 
 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_non_ported_ipv4_connection_instance_alloc);
 
 /*
  * ecm_nss_non_ported_ipv4_debugfs_init()
--- a/frontends/nss/ecm_nss_non_ported_ipv4.h
+++ b/frontends/nss/ecm_nss_non_ported_ipv4.h
@@ -19,7 +19,7 @@
 
 extern bool ecm_nss_non_ported_ipv4_debugfs_init(struct dentry *dentry);
 
-extern struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv4_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv4_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
--- a/frontends/nss/ecm_nss_non_ported_ipv6.c
+++ b/frontends/nss/ecm_nss_non_ported_ipv6.c
@@ -33,7 +33,7 @@
 #include <net/addrconf.h>
 #include <net/ipv6.h>
 #include <net/tcp.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
@@ -342,7 +342,7 @@ static void ecm_nss_non_ported_ipv6_conn
 	ip_addr_t src_ip;
 	ip_addr_t dest_ip;
 	ecm_front_end_acceleration_mode_t result_mode;
-#if defined(ECM_INTERFACE_GRE_TAP_ENABLE) || defined(ECM_INTERFACE_GRE_TUN_ENABLE)
+#if defined(ECM_INTERFACE_GRE_TAP_ENABLE) || defined(ECM_INTERFACE_GRE_TUN_ENABLE) || defined(ECM_INTERFACE_TUNIPIP6_ENABLE)
 	struct net_device *dev;
 #endif
 	int proto = ecm_db_connection_protocol_get(feci->ci);
@@ -1728,6 +1728,7 @@ struct ecm_front_end_connection_instance
 
 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_non_ported_ipv6_connection_instance_alloc);
 
 /*
  * ecm_nss_non_ported_ipv6_debugfs_init()
--- a/frontends/nss/ecm_nss_non_ported_ipv6.h
+++ b/frontends/nss/ecm_nss_non_ported_ipv6.h
@@ -19,7 +19,7 @@
 
 extern bool ecm_nss_non_ported_ipv6_debugfs_init(struct dentry *dentry);
 
-extern struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv6_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv6_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
--- a/frontends/nss/ecm_nss_ported_ipv4.c
+++ b/frontends/nss/ecm_nss_ported_ipv4.c
@@ -32,7 +32,7 @@
 #include <net/ip.h>
 #include <net/tcp.h>
 #include <net/addrconf.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
@@ -2044,6 +2044,7 @@ struct ecm_front_end_connection_instance
 
 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_ported_ipv4_connection_instance_alloc);
 
 /*
  * ecm_nss_ported_ipv4_debugfs_init()
--- a/frontends/nss/ecm_nss_ported_ipv4.h
+++ b/frontends/nss/ecm_nss_ported_ipv4.h
@@ -19,7 +19,7 @@
 
 extern bool ecm_nss_ported_ipv4_debugfs_init(struct dentry *dentry);
 
-extern struct ecm_front_end_connection_instance *ecm_nss_ported_ipv4_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_ported_ipv4_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
--- a/frontends/nss/ecm_nss_ported_ipv6.c
+++ b/frontends/nss/ecm_nss_ported_ipv6.c
@@ -33,7 +33,7 @@
 #include <net/addrconf.h>
 #include <net/ipv6.h>
 #include <net/tcp.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/uaccess.h>	/* for put_user */
 #include <net/ipv6.h>
 #include <linux/inet.h>
@@ -1951,6 +1951,7 @@ struct ecm_front_end_connection_instance
 
 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_ported_ipv6_connection_instance_alloc);
 
 /*
  * ecm_nss_ported_ipv6_debugfs_init()
--- a/frontends/nss/ecm_nss_ported_ipv6.h
+++ b/frontends/nss/ecm_nss_ported_ipv6.h
@@ -19,7 +19,7 @@
 
 extern bool ecm_nss_ported_ipv6_debugfs_init(struct dentry *dentry);
 
-extern struct ecm_front_end_connection_instance *ecm_nss_ported_ipv6_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_ported_ipv6_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
