--- a/frontends/cmn/ecm_ipv4.c
+++ b/frontends/cmn/ecm_ipv4.c
@@ -1732,7 +1732,7 @@ static unsigned int ecm_ipv4_post_routin
 	 * If operations have stopped then do not process packets
 	 */
 	spin_lock_bh(&ecm_ipv4_lock);
-	if (unlikely(ecm_front_end_ipv4_stopped) || unlikely(ecm_front_end_ipv4_stopped_temp)) {
+	if (unlikely(ecm_front_end_ipv4_stopped) || unlikely(atomic_read(&ecm_front_end_ipv4_stopped_temp))) {
 		spin_unlock_bh(&ecm_ipv4_lock);
 		DEBUG_TRACE("Front end stopped\n");
 		ecm_stats_v4_inc(ECM_STATS_V4_EXCEPTION_CMN, ECM_STATS_V4_EXCEPTION_FRONT_END_STOPPED);
@@ -1946,7 +1946,7 @@ static unsigned int ecm_ipv4_bridge_post
 	 * If operations have stopped then do not process packets
 	 */
 	spin_lock_bh(&ecm_ipv4_lock);
-	if (unlikely(ecm_front_end_ipv4_stopped) || unlikely(ecm_front_end_ipv4_stopped_temp)) {
+	if (unlikely(ecm_front_end_ipv4_stopped) || unlikely(atomic_read(&ecm_front_end_ipv4_stopped_temp))) {
 		spin_unlock_bh(&ecm_ipv4_lock);
 		DEBUG_TRACE("Front end stopped\n");
 		ecm_stats_v4_inc(ECM_STATS_V4_EXCEPTION_CMN, ECM_STATS_V4_EXCEPTION_BRIDGE_FRONT_END_STOPPED);
@@ -2196,7 +2196,7 @@ unsigned int ecm_ipv4_ovs_dp_process(str
 	 * If operations have stopped then do not process packets
 	 */
 	spin_lock_bh(&ecm_ipv4_lock);
-	if (unlikely(ecm_front_end_ipv4_stopped) || unlikely(ecm_front_end_ipv4_stopped_temp)) {
+	if (unlikely(ecm_front_end_ipv4_stopped) || unlikely(atomic_read(&ecm_front_end_ipv4_stopped_temp))) {
 		spin_unlock_bh(&ecm_ipv4_lock);
 		DEBUG_TRACE("Front end stopped\n");
 		return 1;
--- a/frontends/cmn/ecm_ipv6.c
+++ b/frontends/cmn/ecm_ipv6.c
@@ -1836,7 +1836,7 @@ static unsigned int ecm_ipv6_post_routin
 	 * If operations have stopped then do not process packets
 	 */
 	spin_lock_bh(&ecm_ipv6_lock);
-	if (unlikely(ecm_front_end_ipv6_stopped) || unlikely(ecm_front_end_ipv6_stopped_temp)) {
+	if (unlikely(ecm_front_end_ipv6_stopped) || unlikely(atomic_read(&ecm_front_end_ipv6_stopped_temp))) {
 		spin_unlock_bh(&ecm_ipv6_lock);
 		DEBUG_TRACE("Front end stopped\n");
 		ecm_stats_v6_inc(ECM_STATS_V6_EXCEPTION_CMN, ECM_STATS_V6_EXCEPTION_FRONT_END_STOPPED);
@@ -2063,7 +2063,7 @@ static unsigned int ecm_ipv6_bridge_post
 	 * If operations have stopped then do not process packets
 	 */
 	spin_lock_bh(&ecm_ipv6_lock);
-	if (unlikely(ecm_front_end_ipv6_stopped) || unlikely(ecm_front_end_ipv6_stopped_temp)) {
+	if (unlikely(ecm_front_end_ipv6_stopped) || unlikely(atomic_read(&ecm_front_end_ipv6_stopped_temp))) {
 		spin_unlock_bh(&ecm_ipv6_lock);
 		DEBUG_TRACE("Front end stopped\n");
 		ecm_stats_v6_inc(ECM_STATS_V6_EXCEPTION_CMN, ECM_STATS_V6_EXCEPTION_BRIDGE_FRONT_END_STOPPED);
@@ -2310,7 +2310,7 @@ unsigned int ecm_ipv6_ovs_dp_process(str
 	 * If operations have stopped then do not process packets
 	 */
 	spin_lock_bh(&ecm_ipv6_lock);
-	if (unlikely(ecm_front_end_ipv6_stopped) || unlikely(ecm_front_end_ipv6_stopped_temp)) {
+	if (unlikely(ecm_front_end_ipv6_stopped) || unlikely(atomic_read(&ecm_front_end_ipv6_stopped_temp))) {
 		spin_unlock_bh(&ecm_ipv6_lock);
 		DEBUG_TRACE("Front end stopped\n");
 		return 1;
--- a/frontends/ecm_front_end_ipv4.c
+++ b/frontends/ecm_front_end_ipv4.c
@@ -61,7 +61,7 @@ int ecm_front_end_ipv4_stopped = 0;	/* W
 /*
  * Temporary operational control
  */
-int ecm_front_end_ipv4_stopped_temp = 0;	/* When non-zero further traffic/process will not be processed where it is checked */
+atomic_t ecm_front_end_ipv4_stopped_temp = ATOMIC_INIT(0);	/* When non-zero further traffic/process will not be processed where it is checked */
 
 /*
  * ecm_front_end_ipv4_interface_construct_ip_addr_set()
@@ -381,7 +381,13 @@ bool ecm_front_end_ipv4_interface_constr
  */
 void ecm_front_end_ipv4_stop_temp(int num)
 {
-	ecm_front_end_ipv4_stopped_temp = num;
+	if (num) {
+		atomic_inc(&ecm_front_end_ipv4_stopped_temp);
+	} else {
+		if (atomic_read(&ecm_front_end_ipv4_stopped_temp) > 0) {
+			atomic_dec(&ecm_front_end_ipv4_stopped_temp);
+		}
+	}
 }
 
 /*
--- a/frontends/ecm_front_end_ipv6.c
+++ b/frontends/ecm_front_end_ipv6.c
@@ -62,7 +62,7 @@ int ecm_front_end_ipv6_stopped = 0;	/* W
 /*
  * Temporary operational control
  */
-int ecm_front_end_ipv6_stopped_temp = 0;	/* When non-zero further traffic/process will not be processed where it is checked */
+atomic_t ecm_front_end_ipv6_stopped_temp = ATOMIC_INIT(0);	/* When non-zero further traffic/process will not be processed where it is checked */
 
 /*
  * ecm_front_end_ipv6_interface_construct_ip_addr_set()
@@ -456,7 +456,13 @@ bool ecm_front_end_ipv6_interface_constr
  */
 void ecm_front_end_ipv6_stop_temp(int num)
 {
-	ecm_front_end_ipv6_stopped_temp = num;
+	if (num) {
+		atomic_inc(&ecm_front_end_ipv6_stopped_temp);
+	} else {
+		if (atomic_read(&ecm_front_end_ipv6_stopped_temp) > 0) {
+			atomic_dec(&ecm_front_end_ipv6_stopped_temp);
+		}
+	}
 }
 
 /*
--- a/frontends/include/ecm_front_end_ipv4.h
+++ b/frontends/include/ecm_front_end_ipv4.h
@@ -17,7 +17,7 @@
  */
 
 extern int ecm_front_end_ipv4_stopped;	/* When non-zero further traffic will not be processed */
-extern int ecm_front_end_ipv4_stopped_temp;	/* When non-zero further traffic/process will not be processed where it is checked */
+extern atomic_t ecm_front_end_ipv4_stopped_temp;	/* When non-zero further traffic/process will not be processed where it is checked */
 
 #ifdef ECM_MULTICAST_ENABLE
 extern int ecm_front_end_ipv4_mc_stopped;	/* When non-zero further traffic will not be processed */
--- a/frontends/include/ecm_front_end_ipv6.h
+++ b/frontends/include/ecm_front_end_ipv6.h
@@ -17,7 +17,7 @@
  */
 
 extern int ecm_front_end_ipv6_stopped;	/* When non-zero further traffic will not be processed */
-extern int ecm_front_end_ipv6_stopped_temp;	/* When non-zero further traffic/process will not be processed where it is checked */
+extern atomic_t ecm_front_end_ipv6_stopped_temp;	/* When non-zero further traffic/process will not be processed where it is checked */
 
 #ifdef ECM_MULTICAST_ENABLE
 extern int ecm_front_end_ipv6_mc_stopped;	/* When non-zero further traffic will not be processed */
