--- a/bridge/nss_bridge_mgr.c
+++ b/bridge/nss_bridge_mgr.c
@@ -1098,8 +1098,10 @@ int nss_bridge_mgr_register_br(struct ne
 	 */
 	b_pvt->ifnum = ifnum;
 	b_pvt->mtu = dev->mtu;
+#if defined(NSS_BRIDGE_MGR_PPE_SUPPORT)
 	b_pvt->wan_if_num = -1;
 	b_pvt->wan_if_enabled = false;
+#endif
 	ether_addr_copy(b_pvt->dev_addr, dev->dev_addr);
 	spin_lock(&br_mgr_ctx.lock);
 	list_add(&b_pvt->list, &br_mgr_ctx.list);
@@ -1165,6 +1167,7 @@ static int nss_bridge_mgr_bond_slave_cha
 		return NOTIFY_DONE;
 	}
 
+#if defined(NSS_BRIDGE_MGR_PPE_SUPPORT)
 	/*
 	 * Add or remove the slave based based on linking event
 	 */
@@ -1179,6 +1182,7 @@ static int nss_bridge_mgr_bond_slave_cha
 					cu_info->upper_dev->name, master->name);
 		}
 	}
+#endif
 
 	return NOTIFY_DONE;
 }
@@ -1252,14 +1256,13 @@ static int nss_bridge_mgr_changeaddr_eve
 
 /*
  * nss_bridge_mgr_changeupper_event()
- *	Bridge manager handles netdevice joining or leaving bridge notification.
+ *	Bridge manager handles upper device notifications not covered by
+ *	completed bridge join/leave events.
  */
 static int nss_bridge_mgr_changeupper_event(struct netdev_notifier_info *info)
 {
 	struct net_device *dev = netdev_notifier_info_to_dev(info);
-	struct net_device *master_dev;
 	struct netdev_notifier_changeupper_info *cu_info;
-	struct nss_bridge_pvt *b_pvt;
 
 	cu_info = (struct netdev_notifier_changeupper_info *)info;
 
@@ -1275,39 +1278,57 @@ static int nss_bridge_mgr_changeupper_ev
 	if (netif_is_bond_slave(dev))
 		return nss_bridge_mgr_bond_slave_changeupper(cu_info, dev);
 
-	master_dev = cu_info->upper_dev;
+	return NOTIFY_DONE;
+}
+
+/*
+ * nss_bridge_mgr_bridge_event()
+ *	Bridge manager handles completed bridge join/leave notifications.
+ */
+static int nss_bridge_mgr_bridge_event(struct netdev_notifier_info *info,
+					unsigned long event)
+{
+	struct net_device *dev = netdev_notifier_info_to_dev(info);
+	struct net_device *master_dev;
+	struct nss_bridge_pvt *b_pvt;
+
+	master_dev = netdev_master_upper_dev_get(dev);
+	if (!master_dev)
+		return NOTIFY_DONE;
 
-	/*
-	 * Check if upper_dev is a known bridge.
-	 */
 	b_pvt = nss_bridge_mgr_find_instance(master_dev);
 	if (!b_pvt)
 		return NOTIFY_DONE;
 
 	/*
-	 * Slave device is bond master and it is added/removed to/from bridge
+	 * Slave device is bond master and it is added/removed to/from bridge.
 	 */
 	if (netif_is_bond_master(dev)) {
 #if defined(NSS_BRIDGE_MGR_PPE_SUPPORT)
-		if (cu_info->linking)
+		if (event == NETDEV_BR_JOIN)
 			return nss_bridge_mgr_bond_master_join(dev, b_pvt);
 		else
 			return nss_bridge_mgr_bond_master_leave(dev, b_pvt);
 #endif
+		return NOTIFY_DONE;
 	}
 
-	if (cu_info->linking) {
-		nss_bridge_mgr_trace("%px: Interface %s joining bridge %s\n", b_pvt, dev->name, master_dev->name);
+	if (event == NETDEV_BR_JOIN) {
+		nss_bridge_mgr_trace("%px: Interface %s joining bridge %s\n",
+				b_pvt, dev->name, master_dev->name);
 		if (nss_bridge_mgr_join_bridge(dev, b_pvt)) {
-			nss_bridge_mgr_warn("%px: Interface %s failed to join bridge %s\n", b_pvt, dev->name, master_dev->name);
+			nss_bridge_mgr_warn("%px: Interface %s failed to join bridge %s\n",
+					b_pvt, dev->name, master_dev->name);
 		}
 
 		return NOTIFY_DONE;
 	}
 
-	nss_bridge_mgr_trace("%px: Interface %s leaving bridge %s\n", b_pvt, dev->name, master_dev->name);
+	nss_bridge_mgr_trace("%px: Interface %s leaving bridge %s\n",
+			b_pvt, dev->name, master_dev->name);
 	if (nss_bridge_mgr_leave_bridge(dev, b_pvt)) {
-		nss_bridge_mgr_warn("%px: Interface %s failed to leave bridge %s\n", b_pvt, dev->name, master_dev->name);
+		nss_bridge_mgr_warn("%px: Interface %s failed to leave bridge %s\n",
+				b_pvt, dev->name, master_dev->name);
 	}
 
 	return NOTIFY_DONE;
@@ -1345,6 +1366,9 @@ static int nss_bridge_mgr_netdevice_even
 	switch (event) {
 	case NETDEV_CHANGEUPPER:
 		return nss_bridge_mgr_changeupper_event(info);
+	case NETDEV_BR_JOIN:
+	case NETDEV_BR_LEAVE:
+		return nss_bridge_mgr_bridge_event(info, event);
 	case NETDEV_CHANGEADDR:
 		return nss_bridge_mgr_changeaddr_event(info);
 	case NETDEV_CHANGEMTU:
@@ -1365,7 +1389,6 @@ static struct notifier_block nss_bridge_
 	.notifier_call = nss_bridge_mgr_netdevice_event,
 };
 
-#if defined(NSS_BRIDGE_MGR_PPE_SUPPORT)
 /*
  * nss_bridge_mgr_is_physical_dev()
  *	Check if the device is on physical device.
@@ -1426,7 +1449,7 @@ static int nss_bridge_mgr_fdb_update_cal
 	struct net_device *br_dev = NULL;
 	fal_fdb_entry_t entry;
 
-	if (!event->br)
+	if (!event || !event->br || !event->orig_dev || !event->dev)
 		return NOTIFY_DONE;
 
 	br_dev = br_fdb_bridge_dev_get_and_hold(event->br);
@@ -1483,7 +1506,7 @@ static struct notifier_block nss_bridge_
  * nss_bridge_mgr_wan_inf_add_handler
  *	Marks an interface as a WAN interface for special handling by bridge.
  */
-static int nss_bridge_mgr_wan_intf_add_handler(struct ctl_table *table,
+static int nss_bridge_mgr_wan_intf_add_handler(const struct ctl_table *table,
 						int write, void __user *buffer,
 						size_t *lenp, loff_t *ppos)
 {
@@ -1532,7 +1555,7 @@ static int nss_bridge_mgr_wan_intf_add_h
  * nss_bridge_mgr_wan_inf_del_handler
  *	Un-marks an interface as a WAN interface.
  */
-static int nss_bridge_mgr_wan_intf_del_handler(struct ctl_table *table,
+static int nss_bridge_mgr_wan_intf_del_handler(const struct ctl_table *table,
 						int write, void __user *buffer,
 						size_t *lenp, loff_t *ppos)
 {
@@ -1590,28 +1613,8 @@ static struct ctl_table nss_bridge_mgr_t
 		.maxlen         = sizeof(char) * IFNAMSIZ,
 		.mode           = 0644,
 		.proc_handler   = &nss_bridge_mgr_wan_intf_del_handler,
-	},
-	{ }
-};
-
-static struct ctl_table nss_bridge_mgr_dir[] = {
-	{
-		.procname	= "bridge_mgr",
-		.mode		= 0555,
-		.child		= nss_bridge_mgr_table,
-	},
-	{ }
-};
-
-static struct ctl_table nss_bridge_mgr_root_dir[] = {
-	{
-		.procname	= "nss",
-		.mode		= 0555,
-		.child		= nss_bridge_mgr_dir,
-	},
-	{ }
+	}
 };
-#endif
 
 /*
  * nss_bridge_mgr_init_module()
@@ -1632,7 +1635,7 @@ int __init nss_bridge_mgr_init_module(vo
 #if defined(NSS_BRIDGE_MGR_PPE_SUPPORT)
 	br_mgr_ctx.wan_if_num = -1;
 	br_fdb_update_register_notify(&nss_bridge_mgr_fdb_update_notifier);
-	br_mgr_ctx.nss_bridge_mgr_header = register_sysctl_table(nss_bridge_mgr_root_dir);
+	br_mgr_ctx.nss_bridge_mgr_header = register_sysctl("nss/bridge_mgr", nss_bridge_mgr_table);
 
 	/*
 	 * Enable ACL rule to enable L2 exception. This is needed if PPE Virtual ports is added to bridge.
@@ -1679,6 +1682,7 @@ module_exit(nss_bridge_mgr_exit_module);
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS bridge manager");
+MODULE_INFO(description, "NSS bridge manager");
 
 module_param(ovs_enabled, bool, 0644);
 MODULE_PARM_DESC(ovs_enabled, "OVS bridge is enabled");
--- a/lag/nss_lag.c
+++ b/lag/nss_lag.c
@@ -378,3 +378,4 @@ module_exit(nss_lag_exit);
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS LAG Client");
+MODULE_INFO(description, "NSS LAG Client");
--- a/portifmgr/nss_portifmgr.c
+++ b/portifmgr/nss_portifmgr.c
@@ -187,16 +187,20 @@ drop:
 }
 
 /*
- * nss_portifmgr_get_stats()
+ * nss_portifmgr_get_stats64()
  *	Netdev get stats function to get port stats
  */
-static struct rtnl_link_stats64 *nss_portifmgr_get_stats(struct net_device *dev, struct rtnl_link_stats64 *stats)
+/*
+ * nss_nlgre_redir_cmn_dev_stats64
+ *	Report packet statistics to linux
+ */
+static void nss_portifmgr_get_stats64(struct net_device *dev,
+		struct rtnl_link_stats64 *stats)
 {
 	struct nss_portifmgr_priv *priv = (struct nss_portifmgr_priv *)netdev_priv(dev);
 	BUG_ON(priv == NULL);
 
 	nss_portid_get_stats(priv->if_num, stats);
-	return stats;
 }
 
 /*
@@ -225,7 +229,7 @@ static const struct net_device_ops nss_p
 	.ndo_start_xmit		= nss_portifmgr_start_xmit,
 	.ndo_set_mac_address	= eth_mac_addr,
 	.ndo_change_mtu		= nss_portifmgr_change_mtu,
-	.ndo_get_stats64	= nss_portifmgr_get_stats,
+	.ndo_get_stats64	= nss_portifmgr_get_stats64,
 };
 
 /*
@@ -262,13 +266,12 @@ struct net_device *nss_portifmgr_create_
 	ndev->vlan_features |= NSS_PORTIFMGR_SUPPORTED_FEATURES;
 	ndev->wanted_features |= NSS_PORTIFMGR_SUPPORTED_FEATURES;
 	ndev->mtu = real_dev->mtu - NSS_PORTIFMGR_EXTRA_HEADER_SIZE;
-	strlcpy(ndev->name, name, IFNAMSIZ);
+	strscpy(ndev->name, name, IFNAMSIZ);
 
 	/*
 	 * Setup temp mac address, this can be changed with ifconfig later
 	 */
-	memcpy(ndev->dev_addr, "\x00\x03\x7f\x11\x22\x00", ETH_ALEN);
-	ndev->dev_addr[5] = 0x10 + port_id;
+	eth_hw_addr_set(ndev, (const u8 []){0x00, 0x03, 0x7f, 0x11, 0x22, 0x10 + port_id});
 
 	err = register_netdev(ndev);
 	if (err) {
--- a/vlan/Makefile
+++ b/vlan/Makefile
@@ -8,7 +8,7 @@ ifeq ($(SoC),$(filter $(SoC),ipq807x ipq
 ccflags-y += -DNSS_VLAN_MGR_PPE_SUPPORT
 endif
 
-ccflags-y += -DNSS_VLAN_MGR_DEBUG_LEVEL=0
+ccflags-y += -DNSS_VLAN_MGR_DEBUG_LEVEL=4
 ccflags-y += -Wall -Werror
 
 ifneq (,$(filter $(CONFIG_BONDING),y m))
--- a/vlan/nss_vlan_mgr.c
+++ b/vlan/nss_vlan_mgr.c
@@ -448,8 +448,8 @@ static int nss_vlan_mgr_bond_configure_p
 	/*
 	 * Fields for match
 	 */
-	v->eg_xlt_rule.vsi_valid = true;	/* Use vsi as search key*/
-	v->eg_xlt_rule.vsi_enable = true;	/* Use vsi as search key*/
+	v->eg_xlt_rule.vsi_valid = 1;		/* Use vsi as search key*/
+	v->eg_xlt_rule.vsi_enable = 1;		/* Use vsi as search key*/
 	v->eg_xlt_rule.vsi = vsi;		/* Use vsi as search key*/
 	v->eg_xlt_rule.s_tagged = 0x7;		/* Accept tagged/untagged/priority tagged svlan */
 	v->eg_xlt_rule.c_tagged = 0x7;		/* Accept tagged/untagged/priority tagged cvlan */
@@ -612,8 +612,8 @@ static int nss_vlan_mgr_configure_ppe(st
 	/*
 	 * Fields for match
 	 */
-	v->eg_xlt_rule.vsi_valid = true;	/* Use vsi as search key*/
-	v->eg_xlt_rule.vsi_enable = true;	/* Use vsi as search key*/
+	v->eg_xlt_rule.vsi_valid = 1;		/* Use vsi as search key*/
+	v->eg_xlt_rule.vsi_enable = 1;		/* Use vsi as search key*/
 	v->eg_xlt_rule.vsi = vsi;		/* Use vsi as search key*/
 	v->eg_xlt_rule.s_tagged = 0x7;		/* Accept tagged/untagged/priority tagged svlan */
 	v->eg_xlt_rule.c_tagged = 0x7;		/* Accept tagged/untagged/priority tagged cvlan */
@@ -688,8 +688,10 @@ detach_vsi:
  */
 static void nss_vlan_mgr_instance_free(struct nss_vlan_pvt *v)
 {
+#ifdef NSS_VLAN_MGR_PPE_SUPPORT
 	int32_t i;
 	int ret = 0;
+#endif
 
 #ifdef NSS_VLAN_MGR_PPE_SUPPORT
 	if (v->ppe_vsi) {
@@ -999,8 +1001,11 @@ static int nss_vlan_mgr_register_event(s
 	int ret;
 #endif
 	uint32_t vlan_tag;
+#ifdef NSS_VLAN_MGR_PPE_SUPPORT
 	struct net_device *slave;
-	int32_t port, port_if;
+	int32_t port;
+#endif
+	int32_t port_if;
 	struct vlan_dev_priv *vlan;
 	struct net_device *real_dev;
 	bool is_bond_master = false;
@@ -1393,8 +1398,10 @@ return_with_error:
 int nss_vlan_mgr_join_bridge(struct net_device *dev, uint32_t bridge_vsi)
 {
 	struct nss_vlan_pvt *v = nss_vlan_mgr_instance_find_and_ref(dev);
+#ifdef NSS_VLAN_MGR_PPE_SUPPORT
 	struct net_device *real_dev;
 	int ret;
+#endif
 
 	if (!v)
 		return 0;
@@ -1454,8 +1461,10 @@ EXPORT_SYMBOL(nss_vlan_mgr_join_bridge);
 int nss_vlan_mgr_leave_bridge(struct net_device *dev, uint32_t bridge_vsi)
 {
 	struct nss_vlan_pvt *v = nss_vlan_mgr_instance_find_and_ref(dev);
+#ifdef NSS_VLAN_MGR_PPE_SUPPORT
 	struct net_device *real_dev;
 	int ret;
+#endif
 
 	if (!v)
 		return 0;
@@ -1540,7 +1549,7 @@ static int nss_vlan_mgr_update_ppe_tpid(
  * nss_vlan_mgr_tpid_proc_handler()
  *	Sets customer TPID and service TPID
  */
-static int nss_vlan_mgr_tpid_proc_handler(struct ctl_table *ctl,
+static int nss_vlan_mgr_tpid_proc_handler(const struct ctl_table *ctl,
 					  int write, void __user *buffer,
 					  size_t *lenp, loff_t *ppos)
 {
@@ -1568,32 +1577,7 @@ static struct ctl_table nss_vlan_table[]
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &nss_vlan_mgr_tpid_proc_handler,
-	},
-	{ }
-};
-
-/*
- * nss_vlan sysctl dir
- */
-static struct ctl_table nss_vlan_dir[] = {
-	{
-		.procname		= "vlan_client",
-		.mode			= 0555,
-		.child			= nss_vlan_table,
-	},
-	{ }
-};
-
-/*
- * nss_vlan systel root dir
- */
-static struct ctl_table nss_vlan_root_dir[] = {
-	{
-		.procname		= "nss",
-		.mode			= 0555,
-		.child			= nss_vlan_dir,
-	},
-	{ }
+	}
 };
 
 /*
@@ -1820,8 +1804,8 @@ void nss_vlan_mgr_add_vlan_rule(struct n
 	/*
 	 * Fields for match
 	 */
-	eg_xlt_rule.vsi_valid = true;	/* Use vsi as search key */
-	eg_xlt_rule.vsi_enable = true;	/* Use vsi as search key */
+	eg_xlt_rule.vsi_valid = 1;	/* Use vsi as search key */
+	eg_xlt_rule.vsi_enable = 1;	/* Use vsi as search key */
 	eg_xlt_rule.vsi = bridge_vsi;	/* Use vsi as search key */
 	eg_xlt_rule.s_tagged = 0x7;	/* Accept tagged/untagged/priority tagged svlan */
 	eg_xlt_rule.c_tagged = 0x7;	/* Accept tagged/untagged/priority tagged cvlan */
@@ -1888,8 +1872,8 @@ void nss_vlan_mgr_del_vlan_rule(struct n
 	/*
 	 * Fields for match
 	 */
-	eg_xlt_rule.vsi_valid = true;	/* Use vsi as search key */
-	eg_xlt_rule.vsi_enable = true;	/* Use vsi as search key */
+	eg_xlt_rule.vsi_valid = 1;	/* Use vsi as search key */
+	eg_xlt_rule.vsi_enable = 1;	/* Use vsi as search key */
 	eg_xlt_rule.vsi = bridge_vsi;	/* Use vsi as search key */
 	eg_xlt_rule.s_tagged = 0x7;	/* Accept tagged/untagged/priority tagged svlan */
 	eg_xlt_rule.c_tagged = 0x7;	/* Accept tagged/untagged/priority tagged cvlan */
@@ -1935,7 +1919,7 @@ int __init nss_vlan_mgr_init_module(void
 	vlan_mgr_ctx.stpid = ETH_P_8021Q;
 
 #ifdef NSS_VLAN_MGR_PPE_SUPPORT
-	vlan_mgr_ctx.sys_hdr = register_sysctl_table(nss_vlan_root_dir);
+	vlan_mgr_ctx.sys_hdr = register_sysctl("nss/vlan_client", nss_vlan_table);
 	if (!vlan_mgr_ctx.sys_hdr) {
 		nss_vlan_mgr_warn("Unabled to register sysctl table for vlan manager\n");
 		return -EFAULT;
@@ -1976,3 +1960,4 @@ module_exit(nss_vlan_mgr_exit_module);
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS vlan manager");
+MODULE_INFO(description, "NSS vlan manager");
