--- a/gre/nss_connmgr_gre.c
+++ b/gre/nss_connmgr_gre.c
@@ -186,7 +186,8 @@ static int nss_connmgr_gre_dev_init(stru
 	if ((dev->priv_flags_ext & IFF_EXT_GRE_V4_TAP) || (dev->type == ARPHRD_IPGRE)) {
 		dev->needed_headroom = sizeof(struct iphdr) + sizeof(struct ethhdr) + MAX_WIFI_HEADROOM + append;
 		dev->mtu = ETH_DATA_LEN - sizeof(struct iphdr) - append;
-		dev->features |= NETIF_F_NETNS_LOCAL | NETIF_F_SG | NETIF_F_FRAGLIST | NETIF_F_HIGHDMA;
+		dev->features |= NETIF_F_SG | NETIF_F_FRAGLIST | NETIF_F_HIGHDMA;
+		dev->netns_immutable = true;
 		dev->hw_features |= NETIF_F_SG | NETIF_F_FRAGLIST | NETIF_F_HIGHDMA;
 		return 0;
 	}
@@ -200,7 +201,7 @@ static int nss_connmgr_gre_dev_init(stru
 		dev->mtu = IPV6_MIN_MTU;
 	}
 
-	dev->features |= NETIF_F_NETNS_LOCAL;
+	dev->netns_immutable = true;
 	return 0;
 }
 
@@ -211,7 +212,6 @@ static int nss_connmgr_gre_dev_init(stru
 static void nss_connmgr_gre_dev_uninit(struct net_device *dev)
 {
 	free_percpu(dev->tstats);
-	return;
 }
 
 /*
@@ -277,16 +277,16 @@ static struct rtnl_link_stats64 *nss_con
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(3, 15, 0))
 			start = u64_stats_fetch_begin_bh(&tstats->syncp);
 #else
-			start = u64_stats_fetch_begin_irq(&tstats->syncp);
+			start = u64_stats_fetch_begin(&tstats->syncp);
 #endif
-			rx_packets = tstats->rx_packets;
-			tx_packets = tstats->tx_packets;
-			rx_bytes = tstats->rx_bytes;
-			tx_bytes = tstats->tx_bytes;
+			rx_packets = u64_stats_read(&tstats->rx_packets);
+			tx_packets = u64_stats_read(&tstats->tx_packets);
+			rx_bytes = u64_stats_read(&tstats->rx_bytes);
+			tx_bytes = u64_stats_read(&tstats->tx_bytes);
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(3, 15, 0))
 		} while (u64_stats_fetch_retry_bh(&tstats->syncp, start));
 #else
-		} while (u64_stats_fetch_retry_irq(&tstats->syncp, start));
+		} while (u64_stats_fetch_retry(&tstats->syncp, start));
 #endif
 
 		tot->rx_packets += rx_packets;
@@ -578,7 +578,6 @@ static void nss_connmgr_gre_tap_inner_ex
 	 */
 	skb->protocol = eth_type_trans(skb, dev);
 	netif_receive_skb(skb);
-	return;
 }
 
 /*
@@ -697,11 +696,11 @@ static void nss_connmgr_gre_event_receiv
 		tstats = this_cpu_ptr(dev->tstats);
 		u64_stats_update_begin(&tstats->syncp);
 		if (interface_type == NSS_DYNAMIC_INTERFACE_TYPE_GRE_INNER) {
-			tstats->tx_packets += stats->tx_packets;
-			tstats->tx_bytes += stats->tx_bytes;
+			u64_stats_add(&tstats->tx_packets, stats->tx_packets);
+			u64_stats_add(&tstats->tx_bytes, stats->tx_bytes);
 		} else if (interface_type == NSS_DYNAMIC_INTERFACE_TYPE_GRE_OUTER) {
-			tstats->rx_packets += stats->rx_packets;
-			tstats->rx_bytes += stats->rx_bytes;
+			u64_stats_add(&tstats->rx_packets, stats->rx_packets);
+			u64_stats_add(&tstats->rx_bytes, stats->rx_bytes);
 		}
 		u64_stats_update_end(&tstats->syncp);
 		dev->stats.rx_dropped += nss_cmn_rx_dropped_sum(stats);
@@ -724,10 +723,10 @@ static void nss_connmgr_gre_make_name(st
 {
 	switch (cfg->mode) {
 	case GRE_MODE_TUN:
-		strlcpy(name, "tun-%d", IFNAMSIZ);
+		strscpy(name, "tun-%d", IFNAMSIZ);
 		break;
 	case GRE_MODE_TAP:
-		strlcpy(name, "tap-%d", IFNAMSIZ);
+		strscpy(name, "tap-%d", IFNAMSIZ);
 		break;
 	default:
 		break;
@@ -757,7 +756,7 @@ static struct net_device *__nss_connmgr_
 	int ret = -1, retry, next_if_num_inner = 0, next_if_num_outer = 0;
 
 	if (cfg->name) {
-		strlcpy(name, cfg->name, IFNAMSIZ);
+		strscpy(name, cfg->name, IFNAMSIZ);
 	} else {
 		nss_connmgr_gre_make_name(cfg, name);
 	}
@@ -1672,7 +1671,7 @@ int nss_connmgr_gre_set_wifi_next_hop(st
 		return GRE_ERR_NEXT_NODE_UNREG_IN_AE;
 	}
 
-	ctx = nss_wifi_get_context();
+	ctx = nss_wifili_get_context();
 	status = nss_wifi_vdev_set_next_hop(ctx, ifnumber, NSS_GRE_INTERFACE);
 	if (status != NSS_TX_SUCCESS) {
 		nss_connmgr_gre_info("%px: wifi drv api failed to set next hop\n", wifi_vdev);
@@ -1710,31 +1709,31 @@ uint32_t nss_connmgr_gre_get_hlen(struct
  *	Map User configured flags to tunnel specific flag.
  */
 void nss_connmgr_gre_set_gre_flags(struct nss_connmgr_gre_cfg *cfg,
-					   uint16_t *o_flags, uint16_t *i_flags)
+					   unsigned long *o_flags, unsigned long *i_flags)
 
 {
 	if (cfg->ikey_valid) {
-		*i_flags |= TUNNEL_KEY;
+		set_bit(IP_TUNNEL_KEY_BIT, i_flags);
 	}
 
 	if (cfg->okey_valid) {
-		*o_flags |= TUNNEL_KEY;
+		set_bit(IP_TUNNEL_KEY_BIT, o_flags);
 	}
 
 	if (cfg->iseq_valid) {
-		*i_flags |= TUNNEL_SEQ;
+		set_bit(IP_TUNNEL_SEQ_BIT, i_flags);
 	}
 
 	if (cfg->oseq_valid) {
-		*o_flags |= TUNNEL_SEQ;
+		set_bit(IP_TUNNEL_SEQ_BIT, o_flags);
 	}
 
 	if (cfg->icsum_valid) {
-		*i_flags |= TUNNEL_CSUM;
+		set_bit(IP_TUNNEL_CSUM_BIT, i_flags);
 	}
 
 	if (cfg->ocsum_valid) {
-		*o_flags |= TUNNEL_CSUM;
+		set_bit(IP_TUNNEL_CSUM_BIT, o_flags);
 	}
 }
 
@@ -1742,33 +1741,33 @@ void nss_connmgr_gre_set_gre_flags(struc
  * nss_connmgr_gre_get_nss_config_flags()
  *	Map tunnel specific flags to NSS flags.
  */
-uint32_t nss_connmgr_gre_get_nss_config_flags(uint16_t o_flags, uint16_t i_flags,
+uint32_t nss_connmgr_gre_get_nss_config_flags(const unsigned long *o_flags, const unsigned long *i_flags,
 						  uint8_t tos, uint8_t ttl,
 						  uint16_t frag_off)
 {
 	uint32_t gre_flags = 0;
 
-	if (i_flags & TUNNEL_KEY) {
+	if (test_bit(IP_TUNNEL_KEY_BIT, i_flags)) {
 		gre_flags |= NSS_GRE_CONFIG_IKEY_VALID;
 	}
 
-	if (o_flags & TUNNEL_KEY) {
+	if (test_bit(IP_TUNNEL_KEY_BIT, o_flags)) {
 		gre_flags |= NSS_GRE_CONFIG_OKEY_VALID;
 	}
 
-	if (i_flags & TUNNEL_SEQ) {
+	if (test_bit(IP_TUNNEL_SEQ_BIT, i_flags)) {
 		gre_flags |= NSS_GRE_CONFIG_ISEQ_VALID;
 	}
 
-	if (o_flags & TUNNEL_SEQ) {
+	if (test_bit(IP_TUNNEL_SEQ_BIT, o_flags)) {
 		gre_flags |= NSS_GRE_CONFIG_OSEQ_VALID;
 	}
 
-	if (i_flags & TUNNEL_CSUM) {
+	if (test_bit(IP_TUNNEL_CSUM_BIT, i_flags)) {
 		gre_flags |= NSS_GRE_CONFIG_ICSUM_VALID;
 	}
 
-	if (o_flags & TUNNEL_CSUM) {
+	if (test_bit(IP_TUNNEL_CSUM_BIT, o_flags)) {
 		gre_flags |= NSS_GRE_CONFIG_OCSUM_VALID;
 	}
 
@@ -1866,3 +1865,4 @@ module_exit(nss_connmgr_gre_exit_module)
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS gre offload manager");
+MODULE_INFO(description, "NSS gre offload manager");
--- a/gre/nss_connmgr_gre.h
+++ b/gre/nss_connmgr_gre.h
@@ -23,6 +23,7 @@
 #ifndef _NSS_CONNMGR_GRE_H_
 #define _NSS_CONNMGR_GRE_H_
 
+#include <net/ip_tunnels.h>
 #include "nss_connmgr_gre_public.h"
 
 /*
@@ -81,9 +82,9 @@ typedef struct nss_gre_info nss_connmgr_
 
 uint32_t nss_connmgr_gre_get_hlen(struct nss_connmgr_gre_cfg *cfg);
 void nss_connmgr_gre_set_gre_flags(struct nss_connmgr_gre_cfg *cfg,
-					   uint16_t *o_flags, uint16_t *i_flags);
+					   unsigned long *o_flags, unsigned long *i_flags);
 
-uint32_t nss_connmgr_gre_get_nss_config_flags(uint16_t o_flags, uint16_t i_flags,
+uint32_t nss_connmgr_gre_get_nss_config_flags(const unsigned long *o_flags, const unsigned long *i_flags,
 						  uint8_t tos, uint8_t ttl,
 						  uint16_t frag_off);
 int nss_connmgr_gre_set_wifi_next_hop(struct net_device *wifi_vdev);
--- a/gre/nss_connmgr_gre_v4.c
+++ b/gre/nss_connmgr_gre_v4.c
@@ -45,7 +45,7 @@ static struct net_device *nss_connmgr_gr
 	struct net_device *dev;
 	uint32_t ip_addr __attribute__ ((unused)) = ntohl(dest_ip);
 
-	rt = ip_route_output(&init_net, htonl(dest_ip), 0, 0, 0);
+	rt = ip_route_output(&init_net, htonl(dest_ip), 0, 0, 0, 0);
 	if (IS_ERR(rt)) {
 		nss_connmgr_gre_warning("Unable to lookup route for %pI4\n", &ip_addr);
 		return NULL;
@@ -87,7 +87,7 @@ static int nss_connmgr_gre_v4_get_mac_ad
 	dev_put(local_dev);
 	nss_connmgr_gre_info("Src MAC address for %pI4 is %pM\n", &laddr, src_mac);
 
-	rt = ip_route_output(&init_net, raddr, 0, 0, 0);
+	rt = ip_route_output(&init_net, raddr, 0, 0, 0, 0);
 	if (IS_ERR(rt)) {
 		nss_connmgr_gre_warning("route look up failed for %pI4\n", &raddr);
 		return GRE_ERR_RADDR_ROUTE_LOOKUP;
@@ -216,9 +216,9 @@ int nss_connmgr_gre_v4_set_config(struct
 		t->parms.o_key = cfg->okey;
 	}
 
-	nss_connmgr_gre_set_gre_flags(cfg, &t->parms.o_flags, &t->parms.i_flags);
+	nss_connmgr_gre_set_gre_flags(cfg, t->parms.o_flags, t->parms.i_flags);
 
-	strlcpy(t->parms.name, dev->name, IFNAMSIZ);
+	strscpy(t->parms.name, dev->name, IFNAMSIZ);
 	t->dev = dev;
 	return GRE_SUCCESS;
 }
--- a/gre/nss_connmgr_gre_v6.c
+++ b/gre/nss_connmgr_gre_v6.c
@@ -95,7 +95,8 @@ static int nss_connmgr_gre_v6_get_mac_ad
 	/*
 	 * Find src MAC address
 	 */
-	local_dev = (struct net_device *)ipv6_dev_find(&init_net, &src_addr, 1);
+	local_dev = NULL;
+	local_dev = (struct net_device *)ipv6_dev_find(&init_net, &src_addr, local_dev);
 	if (!local_dev) {
 		nss_connmgr_gre_warning("Unable to find local dev for %pI6", src_ip);
 		return GRE_ERR_NO_LOCAL_NETDEV;
@@ -321,9 +322,9 @@ int nss_connmgr_gre_v6_set_config(struct
 		t->parms.o_key = cfg->okey;
 	}
 
-	nss_connmgr_gre_set_gre_flags(cfg, &t->parms.o_flags, &t->parms.i_flags);
+	nss_connmgr_gre_set_gre_flags(cfg, t->parms.o_flags, t->parms.i_flags);
 
-	strlcpy(t->parms.name, dev->name, IFNAMSIZ);
+	strscpy(t->parms.name, dev->name, IFNAMSIZ);
 	t->dev = dev;
 	return GRE_SUCCESS;
 }
--- a/gre/test/nss_connmgr_gre_test.c
+++ b/gre/test/nss_connmgr_gre_test.c
@@ -143,7 +143,7 @@ static ssize_t nss_connmgr_gre_test_writ
 		 * parameter parsing for delete command
 		 */
 		if (!strcmp(param, "dev")) {
-			strlcpy(dev_name, value, IFNAMSIZ);
+			strscpy(dev_name, value, IFNAMSIZ);
 			dev_name_valid = true;
 			break;
 		}
@@ -153,19 +153,19 @@ static ssize_t nss_connmgr_gre_test_writ
 		 * tap create command
 		 */
 		if (!strcmp(param, "next_dev")) {
-			strlcpy(dev_name, value, IFNAMSIZ);
+			strscpy(dev_name, value, IFNAMSIZ);
 			dev_name_valid = true;
 			continue;
 		}
 
 		if (!strcmp(param, "saddr")) {
-			strlcpy(saddr, value, 20);
+			strscpy(saddr, value, 20);
 			saddr_valid = true;
 			continue;
 		}
 
 		if (!strcmp(param, "daddr")) {
-			strlcpy(daddr, value, 20);
+			strscpy(daddr, value, 20);
 			daddr_valid = true;
 			continue;
 		}
@@ -223,16 +223,18 @@ static int nss_connmgr_gre_test_show_pro
  */
 static int nss_connmgr_gre_test_open_proc(struct inode *inode, struct file *filp)
 {
-	return single_open(filp, nss_connmgr_gre_test_show_proc, PDE_DATA(inode));
+	return single_open(filp, nss_connmgr_gre_test_show_proc, pde_data(inode));
 }
 
 /*
  * Proc ops
  */
-static const struct file_operations nss_connmgr_gre_test_proc_ops =  {
-	.open =  nss_connmgr_gre_test_open_proc,
-	.write = nss_connmgr_gre_test_write_proc,
-	.read = seq_read,
+static const struct proc_ops nss_connmgr_gre_test_proc_ops =  {
+	.proc_open	= nss_connmgr_gre_test_open_proc,
+	.proc_read	= seq_read,
+	.proc_lseek	= seq_lseek,
+	.proc_release	= single_release,
+	.proc_write	= nss_connmgr_gre_test_write_proc,
 };
 
 /*
@@ -268,3 +270,4 @@ module_exit(nss_connmgr_gre_test_exit_mo
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS gre test module");
+MODULE_INFO(description, "NSS gre test module");
--- a/nss_connmgr_tun6rd.c
+++ b/nss_connmgr_tun6rd.c
@@ -101,10 +101,10 @@ static void nss_tun6rd_update_dev_stats(
 
 	u64_stats_init(&stats.syncp);
 	u64_stats_update_begin(&stats.syncp);
-	stats.rx_packets = sync_stats->node_stats.rx_packets;
-	stats.rx_bytes = sync_stats->node_stats.rx_bytes;
-	stats.tx_packets = sync_stats->node_stats.tx_packets;
-	stats.tx_bytes = sync_stats->node_stats.tx_bytes;
+	u64_stats_set(&stats.rx_packets, sync_stats->node_stats.rx_packets);
+	u64_stats_set(&stats.rx_bytes, sync_stats->node_stats.rx_bytes);
+	u64_stats_set(&stats.tx_packets, sync_stats->node_stats.tx_packets);
+	u64_stats_set(&stats.tx_bytes, sync_stats->node_stats.tx_bytes);
 	u64_stats_update_end(&stats.syncp);
 #else
 	struct nss_tun6rd_stats stats;
@@ -499,3 +499,4 @@ module_exit(nss_tun6rd_exit_module);
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS tun6rd offload manager");
+MODULE_INFO(description, "NSS tun6rd offload manager");
--- a/pvxlanmgr/nss_pvxlanmgr.c
+++ b/pvxlanmgr/nss_pvxlanmgr.c
@@ -305,7 +305,7 @@ static void nss_pvxlanmgr_dummy_netdev_s
 	dev->priv_destructor = NULL;
 #endif
 
-	memcpy(dev->dev_addr, "\x00\x00\x00\x00\x00\x00", dev->addr_len);
+	eth_hw_addr_set(dev, (const u8 *)"\x00\x00\x00\x00\x00\x00");
 	memset(dev->broadcast, 0xff, dev->addr_len);
 	memcpy(dev->perm_addr, dev->dev_addr, dev->addr_len);
 }
--- a/tunipip6/nss_connmgr_tunipip6.c
+++ b/tunipip6/nss_connmgr_tunipip6.c
@@ -16,6 +16,7 @@
  **************************************************************************
  */
 
+#include <linux/vmalloc.h>
 #include <linux/version.h>
 #include <linux/types.h>
 #include <linux/ip.h>
@@ -174,7 +175,7 @@ static void nss_tunipip6_encap_exception
 	nss_tunipip6_info("%px: received - %d bytes name %s ver %x\n",
 			skb, skb->len, dev->name, iph->version);
 
-	rt = ip_route_output(&init_net, iph->daddr, 0, 0, 0);
+	rt = ip_route_output(&init_net, iph->daddr, 0, 0, 0, 0);
 	if (unlikely(IS_ERR(rt))) {
 		nss_tunipip6_info("%px: Failed to find IPv4 route for dest %pI4 src %pI4\n", skb, &iph->daddr, &iph->saddr);
 		dev_kfree_skb_any(skb);
@@ -258,7 +259,7 @@ static void nss_tunipip6_decap_exception
 	struct iphdr *iph;
 	struct rtable *rt;
 	int cpu;
-	int8_t ver = skb->data[0] >> 4;
+	__attribute__((unused)) int8_t ver = skb->data[0] >> 4;
 
 	nss_tunipip6_trace("%px: received - %d bytes name %s ver %x\n",
 			dev, skb->len, dev->name, ver);
@@ -284,7 +285,7 @@ static void nss_tunipip6_decap_exception
 	iph = ip_hdr(skb);
 	nss_tunipip6_assert(iph->version == 4);
 
-	rt = ip_route_output(&init_net, iph->daddr, 0, 0, 0);
+	rt = ip_route_output(&init_net, iph->daddr, 0, 0, 0, 0);
 	if (unlikely(IS_ERR(rt))) {
 		nss_tunipip6_info("%px: Failed to find IPv4 route for %pI4\n", skb, &iph->daddr);
 		dev_kfree_skb_any(skb);
@@ -354,11 +355,11 @@ static void nss_tunipip6_update_dev_stat
 
 	memset(&stats, 0, sizeof(stats));
 	if (interface_type == NSS_DYNAMIC_INTERFACE_TYPE_TUNIPIP6_INNER) {
-		stats.tx_packets = sync_stats->node_stats.tx_packets;
-		stats.tx_bytes = sync_stats->node_stats.tx_bytes;
+		u64_stats_set(&stats.tx_packets, sync_stats->node_stats.tx_packets);
+		u64_stats_set(&stats.tx_bytes, sync_stats->node_stats.tx_bytes);
 	} else if (interface_type == NSS_DYNAMIC_INTERFACE_TYPE_TUNIPIP6_OUTER) {
-		stats.rx_packets = sync_stats->node_stats.rx_packets;
-		stats.rx_bytes = sync_stats->node_stats.rx_bytes;
+		u64_stats_set(&stats.rx_packets, sync_stats->node_stats.rx_packets);
+		u64_stats_set(&stats.rx_bytes, sync_stats->node_stats.rx_bytes);
 	} else {
 		nss_tunipip6_warning("%px: Invalid interface type received from NSS\n", dev);
 		return;
@@ -1181,3 +1182,4 @@ module_exit(nss_tunipip6_exit_module);
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS tunipip6 offload manager");
+MODULE_INFO(description, "NSS tunipip6 offload manager");
--- a/tunipip6/nss_connmgr_tunipip6_stats.c
+++ b/tunipip6/nss_connmgr_tunipip6_stats.c
@@ -16,6 +16,8 @@
  **************************************************************************
  */
 
+#include <linux/vmalloc.h>
+
 #include <nss_api_if.h>
 #include "nss_connmgr_tunipip6_priv.h"
 
--- a/tunipip6/nss_connmgr_tunipip6_sysctl.c
+++ b/tunipip6/nss_connmgr_tunipip6_sysctl.c
@@ -52,7 +52,7 @@ enum nss_tunipip6_sysctl_mode {
 };
 
 
-static int nss_tunipip6_data_parser(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos, enum nss_tunipip6_sysctl_mode mode)
+static int nss_tunipip6_data_parser(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos, enum nss_tunipip6_sysctl_mode mode)
 {
 	char dev_name[NETDEV_STR_LEN] = {0}, ipv6_prefix_str[PREFIX_STR_LEN] = {0}, ipv6_suffix_str[PREFIX_STR_LEN] = {0}, ipv4_prefix_str[PREFIX_STR_LEN] = {0};
 	uint32_t ipv6_prefix[4], ipv6_prefix_len, ipv6_suffix[4], ipv6_suffix_len, ipv4_prefix, ipv4_prefix_len, ea_len, psid_offset;
@@ -108,7 +108,7 @@ static int nss_tunipip6_data_parser(stru
 		 */
 
 		if (!strcmp(param, "netdev")) {
-			strlcpy(dev_name, value, 30);
+			strscpy(dev_name, value, 30);
 			dev = dev_get_by_name(&init_net, dev_name);
 			if (!dev) {
 				kfree(pfree);
@@ -147,7 +147,7 @@ static int nss_tunipip6_data_parser(stru
 		}
 
 		if (!strcmp(param, "ipv4_prefix")) {
-			strlcpy(ipv4_prefix_str, value, 30);
+			strscpy(ipv4_prefix_str, value, 30);
 			ret = in4_pton(ipv4_prefix_str, -1, (uint8_t *)&ipv4_prefix, -1, NULL);
 			if (ret != 1) {
 				kfree(pfree);
@@ -173,7 +173,7 @@ static int nss_tunipip6_data_parser(stru
 		}
 
 		if (!strcmp(param, "ipv6_prefix")) {
-			strlcpy(ipv6_prefix_str, value, 100);
+			strscpy(ipv6_prefix_str, value, 100);
 			ret = in6_pton(ipv6_prefix_str, -1, (uint8_t *)&ipv6_prefix, -1, NULL);
 			if (ret != 1) {
 				kfree(pfree);
@@ -199,7 +199,7 @@ static int nss_tunipip6_data_parser(stru
 		}
 
 		if (!strcmp(param, "ipv6_suffix")) {
-			strlcpy(ipv6_suffix_str, value, 100);
+			strscpy(ipv6_suffix_str, value, 100);
 			ret = in6_pton(ipv6_suffix_str, -1, (uint8_t *)&ipv6_suffix, -1, NULL);
 			if (ret != 1) {
 				kfree(pfree);
@@ -368,27 +368,27 @@ fail:
 	return 0;
 }
 
-static int nss_tunipip6_cmd_procfs_add_maprule(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_tunipip6_cmd_procfs_add_maprule(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	return nss_tunipip6_data_parser(ctl, write, buffer, lenp, ppos, NSS_TUNIPIP6_SYSCTL_ADD_MAPRULE);
 }
 
-static int nss_tunipip6_cmd_procfs_del_maprule(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_tunipip6_cmd_procfs_del_maprule(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	return nss_tunipip6_data_parser(ctl, write, buffer, lenp, ppos, NSS_TUNIPIP6_SYSCTL_DEL_MAPRULE);
 }
 
-static int nss_tunipip6_cmd_procfs_flush_fmr_rule(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_tunipip6_cmd_procfs_flush_fmr_rule(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	return nss_tunipip6_data_parser(ctl, write, buffer, lenp, ppos, NSS_TUNIPIP6_SYSCTL_FLUSH_FMR_RULE);
 }
 
-static int nss_tunipip6_cmd_procfs_enable_frag_id(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_tunipip6_cmd_procfs_enable_frag_id(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	return nss_tunipip6_data_parser(ctl, write, buffer, lenp, ppos, NSS_TUNIPIP6_SYSCTL_FRAG_ID);
 }
 
-static int nss_tunipip6_cmd_procfs_read_help(struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
+static int nss_tunipip6_cmd_procfs_read_help(const struct ctl_table *ctl, int write, void __user *buffer, size_t *lenp, loff_t *ppos)
 {
 	int ret = proc_dointvec(ctl, write, buffer, lenp, ppos);
 
@@ -445,35 +445,7 @@ static struct ctl_table nss_tunipip6_tab
 		.maxlen			= sizeof(nss_tunipip6_data),
 		.mode			= 0400,
 		.proc_handler		= &nss_tunipip6_cmd_procfs_read_help,
-	},
-	{ }
-};
-
-static struct ctl_table nss_tunipip6_root_dir[] = {
-	{
-		.procname		= "ipip6",
-		.mode			= 0555,
-		.child			= nss_tunipip6_table,
-	},
-	{ }
-};
-
-static struct ctl_table nss_tunipip6_nss_root_dir[] = {
-	{
-		.procname		= "nss",
-		.mode			= 0555,
-		.child			= nss_tunipip6_root_dir,
-	},
-	{ }
-};
-
-static struct ctl_table nss_tunipip6_root[] = {
-	{
-		.procname		= "dev",
-		.mode			= 0555,
-		.child			= nss_tunipip6_nss_root_dir,
-	},
-	{ }
+	}
 };
 
 static struct ctl_table_header *nss_tunipip6_ctl_header;
@@ -483,7 +455,7 @@ static struct ctl_table_header *nss_tuni
  * 	Register command line interface for tunipip6.
  */
 bool nss_tunipip6_sysctl_register(void) {
-	nss_tunipip6_ctl_header = register_sysctl_table(nss_tunipip6_root);
+	nss_tunipip6_ctl_header = register_sysctl("drv/nss/ipip6", nss_tunipip6_table);
 	if (!nss_tunipip6_ctl_header) {
 		return false;
 	}
--- a/vxlanmgr/nss_vxlanmgr.c
+++ b/vxlanmgr/nss_vxlanmgr.c
@@ -84,7 +84,7 @@ int32_t nss_vxlanmgr_bind_ipsec_by_ip(un
 {
 	int32_t ipsec_if_num;
 	nss_vxlanmgr_get_ipsec_if_num_by_ip_callback_t ipsec_cb;
-	struct nss_ctx_instance *nss_ctx = nss_vxlan_get_ctx();
+	__attribute__((unused)) struct nss_ctx_instance *nss_ctx = nss_vxlan_get_ctx();
 
 	/*
 	 * Check if the VxLAN interface is applied over an IPsec interface by querying the IPsec.
@@ -225,3 +225,4 @@ module_exit(nss_vxlanmgr_exit_module);
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS VxLAN manager");
+MODULE_INFO(description, "NSS VxLAN manager");
--- a/vxlanmgr/nss_vxlanmgr_tunnel.c
+++ b/vxlanmgr/nss_vxlanmgr_tunnel.c
@@ -489,8 +489,8 @@ static void nss_vxlanmgr_tunnel_inner_st
 
 	tstats = this_cpu_ptr(dev->tstats);
 	u64_stats_update_begin(&tstats->syncp);
-	tstats->tx_packets += stats->node_stats.tx_packets;
-	tstats->tx_bytes += stats->node_stats.tx_bytes;
+	u64_stats_add(&tstats->tx_packets, stats->node_stats.tx_packets);
+	u64_stats_add(&tstats->tx_bytes, stats->node_stats.tx_bytes);
 	u64_stats_update_end(&tstats->syncp);
 	netdev_stats->tx_dropped += dropped;
 }
@@ -526,8 +526,8 @@ static void nss_vxlanmgr_tunnel_outer_st
 
 	tstats = this_cpu_ptr(dev->tstats);
 	u64_stats_update_begin(&tstats->syncp);
-	tstats->rx_packets += stats->node_stats.tx_packets;
-	tstats->rx_bytes += stats->node_stats.tx_bytes;
+	u64_stats_add(&tstats->rx_packets, stats->node_stats.tx_packets);
+	u64_stats_add(&tstats->rx_bytes, stats->node_stats.tx_bytes);
 	u64_stats_update_end(&tstats->syncp);
 	netdev_stats->rx_dropped += dropped;
 	dev_put(dev);
