--- a/l2tp/l2tpv2/nss_connmgr_l2tpv2.c
+++ b/l2tp/l2tpv2/nss_connmgr_l2tpv2.c
@@ -319,7 +319,7 @@ static struct nss_connmgr_l2tpv2_session
 	 */
 	dev_hold(dev);
 	l2tpv2_session_data->dev = dev;
-	strlcpy(session->ifname, dev->name, IFNAMSIZ);
+	strscpy(session->ifname, dev->name, IFNAMSIZ);
 
 	/*
 	 * There is no need for protecting simultaneous addition &
@@ -417,7 +417,7 @@ static void nss_connmgr_l2tpv2_exception
 				/*
 				 * set skb_iif
 				 */
-				rt = ip_route_output(&init_net, iph_inner->saddr, 0, 0, 0);
+				rt = ip_route_output(&init_net, iph_inner->saddr, 0, 0, 0, 0);
 				if (unlikely(IS_ERR(rt))) {
 					nss_connmgr_l2tpv2_warning("Martian packets !!!");
 				} else {
@@ -859,7 +859,7 @@ EXPORT_SYMBOL(l2tpmgr_unregister_ipsecmg
  * nss_connmgr_l2tpv2_proc_handler()
  *	Read and write handler for sysctl.
  */
-static int nss_connmgr_l2tpv2_proc_handler(struct ctl_table *ctl,
+static int nss_connmgr_l2tpv2_proc_handler(const struct ctl_table *ctl,
 					  int write, void __user *buffer,
 					  size_t *lenp, loff_t *ppos)
 {
@@ -979,33 +979,9 @@ static struct ctl_table nss_connmgr_l2tp
 		.maxlen		= L2TP_SYSCTL_STR_LEN_MAX,
 		.mode		= 0644,
 		.proc_handler	= &nss_connmgr_l2tpv2_proc_handler,
-	},
-	{ }
-};
-
-/*
- * nss_connmgr_l2tpv2_dir
- */
-static struct ctl_table nss_connmgr_l2tpv2_dir[] = {
-	{
-		.procname		= "l2tpv2",
-		.mode			= 0555,
-		.child			= nss_connmgr_l2tpv2_table,
-	},
-	{ }
+	}
 };
 
-/*
- * nss_connmgr_l2tpv2_sysroot sysctl root dir
- */
-static struct ctl_table nss_connmgr_l2tpv2_sysroot[] = {
-	{
-		.procname		= "nss",
-		.mode			= 0555,
-		.child			= nss_connmgr_l2tpv2_dir,
-	},
-	{ }
-};
 #endif
 
 /*
@@ -1030,7 +1006,7 @@ int __init nss_connmgr_l2tpv2_init_modul
 	}
 #endif
 #if defined(NSS_L2TP_IPSEC_BIND_BY_NETDEV)
-	ctl_tbl_hdr = register_sysctl_table(nss_connmgr_l2tpv2_sysroot);
+	ctl_tbl_hdr = register_sysctl("nss/l2tpv2", nss_connmgr_l2tpv2_table);
 	if (!ctl_tbl_hdr) {
 		nss_connmgr_l2tpv2_info("Unable to register sysctl table for L2TP conn mgr\n");
 		return -EFAULT;
@@ -1069,3 +1045,4 @@ module_exit(nss_connmgr_l2tpv2_exit_modu
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS l2tpv2 over ppp offload manager");
+MODULE_INFO(description, "NSS l2tpv2 over ppp offload manager");
--- a/l2tp/l2tpv2/nss_l2tpv2_stats.c
+++ b/l2tp/l2tpv2/nss_l2tpv2_stats.c
@@ -129,7 +129,7 @@ void nss_l2tpv2_update_dev_stats(struct
 		dev_put(dev);
 		return;
 	}
-	session = l2tp_tunnel_get_session(tunnel, data.l2tpv2.session.session_id);
+	session = l2tp_v2_session_get(dev_net(dev), data.l2tpv2.tunnel.tunnel_id, data.l2tpv2.session.session_id);
 	if (!session) {
 		tunnel_put(tunnel);
 		dev_put(dev);
--- a/map/map-t/nss_connmgr_map_t.c
+++ b/map/map-t/nss_connmgr_map_t.c
@@ -1101,3 +1101,4 @@ module_exit(nss_connmgr_map_t_exit_modul
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS map_t offload manager");
+MODULE_INFO(description, "NSS map_t offload manager");
--- a/pppoe/nss_connmgr_pppoe.c
+++ b/pppoe/nss_connmgr_pppoe.c
@@ -439,3 +439,4 @@ module_exit(nss_connmgr_pppoe_exit_modul
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS pppoe offload manager");
+MODULE_INFO(description, "NSS pppoe offload manager");
--- a/pptp/nss_connmgr_pptp.c
+++ b/pptp/nss_connmgr_pptp.c
@@ -889,3 +889,4 @@ module_exit(nss_connmgr_pptp_exit_module
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS pptp offload manager");
+MODULE_INFO(description, "NSS pptp offload manager");
