--- a/nss_qdisc/igs/nss_ifb.c
+++ b/nss_qdisc/igs/nss_ifb.c
@@ -544,8 +544,10 @@ static void nss_ifb_update_dev_stats(str
 	 * post shaping. Therefore IFB interface's stats should be updated
 	 * with NSS firmware's IFB TX stats only.
 	 */
-	stats.rx_packets = stats.tx_packets = node_stats->tx_packets;
-	stats.rx_bytes = stats.tx_bytes = node_stats->tx_bytes;
+	u64_stats_set(&stats.rx_packets, node_stats->tx_packets);
+	u64_stats_set(&stats.tx_packets, node_stats->tx_packets);
+	u64_stats_set(&stats.rx_bytes, node_stats->tx_bytes);
+	u64_stats_set(&stats.tx_bytes, node_stats->tx_bytes);
 	dev->stats.rx_dropped = dev->stats.tx_dropped += sync_stats->igs_stats.tx_dropped;
 	u64_stats_update_end(&stats.syncp);
 
--- a/nss_qdisc/igs/nss_mirred.c
+++ b/nss_qdisc/igs/nss_mirred.c
@@ -89,10 +89,10 @@ static int nss_mirred_init(struct net *n
 {
 #else
 static int nss_mirred_init(struct net *net, struct nlattr *nla,
-			   struct nlattr *est, struct tc_action **tc_act, int ovr,
-			   int bind, bool rtnl_held, struct tcf_proto *tp,
-			   struct netlink_ext_ack *extack)
+			   struct nlattr *est, struct tc_action **tc_act,
+			   struct tcf_proto *tp, u32 flags, struct netlink_ext_ack *extack)
 {
+	bool bind = flags & TCA_ACT_FLAGS_BIND;
 	struct tc_action_net *tn = net_generic(net, nss_mirred_net_id);
 	u32 index;
 #endif
@@ -240,7 +240,7 @@ static int nss_mirred_init(struct net *n
 
 	if (!ret) {
 		ret = tcf_idr_create(tn, index, est, tc_act, &nss_mirred_act_ops,
-				bind, true);
+				bind, true, 0);
 		if (ret) {
 			tcf_idr_cleanup(tn, index);
 			return ret;
@@ -308,7 +308,7 @@ static int nss_mirred_act(struct sk_buff
 	 * Update the last use of action.
 	 */
 	tcf_lastuse_update(&act->tcf_tm);
-	bstats_cpu_update(this_cpu_ptr(act->common.cpu_bstats), skb);
+	bstats_update(this_cpu_ptr(act->common.cpu_bstats), skb);
 
 	rcu_read_lock();
 	retval = READ_ONCE(act->tcf_action);
@@ -767,3 +767,4 @@ module_init(nss_mirred_init_module);
 module_exit(nss_mirred_cleanup_module);
 
 MODULE_LICENSE("Dual BSD/GPL");
+MODULE_INFO(description, "NSS mirred action client");
--- a/nss_qdisc/nss_bf.c
+++ b/nss_qdisc/nss_bf.c
@@ -290,7 +290,7 @@ static void nss_bf_destroy_class(struct
  * nss_bf_delete_class()
  *	Detaches a class from operation, but does not destroy it.
  */
-static int nss_bf_delete_class(struct Qdisc *sch, unsigned long arg)
+static int nss_bf_delete_class(struct Qdisc *sch, unsigned long arg, struct netlink_ext_ack *extack)
 {
 	struct nss_bf_sched_data *q = qdisc_priv(sch);
 	struct nss_bf_class_data *cl = (struct nss_bf_class_data *)arg;
@@ -377,7 +377,7 @@ static int nss_bf_graft_class(struct Qdi
 	 */
 	nss_qdisc_info("Grafting old: %px with new: %px\n", *old, new);
 	if (*old != &noop_qdisc) {
-		struct nss_qdisc *nq_old = (struct nss_qdisc *)qdisc_priv(*old);
+		struct nss_qdisc *nq_old = (struct nss_qdisc *)qdisc_priv(((struct Qdisc *)(*old)));
 		nss_qdisc_info("Detaching old: %px\n", *old);
 		nim_detach.msg.shaper_configure.config.msg.shaper_node_config.qos_tag = cl->nq.qos_tag;
 		if (nss_qdisc_node_detach(&cl->nq, nq_old, &nim_detach,
--- a/nss_qdisc/nss_htb.c
+++ b/nss_qdisc/nss_htb.c
@@ -516,7 +516,7 @@ static void nss_htb_destroy_class(struct
  * nss_htb_delete_class()
  *	Detaches a class from operation, but does not destroy it.
  */
-static int nss_htb_delete_class(struct Qdisc *sch, unsigned long arg)
+static int nss_htb_delete_class(struct Qdisc *sch, unsigned long arg, struct netlink_ext_ack *extack)
 {
 	struct nss_htb_sched_data *q = qdisc_priv(sch);
 	struct nss_htb_class_data *cl = (struct nss_htb_class_data *)arg;
@@ -638,7 +638,7 @@ static int nss_htb_graft_class(struct Qd
 	nss_qdisc_info("grafting old: %x with new: %x\n", (*old)->handle, new->handle);
 	if (*old != &noop_qdisc) {
 		nss_qdisc_trace("detaching old: %x\n", (*old)->handle);
-		nq_old = qdisc_priv(*old);
+		nq_old = qdisc_priv(((struct Qdisc *)(*old)));
 		nim_detach.msg.shaper_configure.config.msg.shaper_node_config.qos_tag = cl->nq.qos_tag;
 		nim_detach.msg.shaper_configure.config.msg.shaper_node_config.snc.htb_group_detach.child_qos_tag = nq_old->qos_tag;
 		if (nss_qdisc_node_detach(&cl->nq, nq_old, &nim_detach,
--- a/nss_qdisc/nss_ppe.c
+++ b/nss_qdisc/nss_ppe.c
@@ -348,8 +348,8 @@ static void nss_ppe_queue_disable(struct
 	/*
 	 * Disable queue enqueue, dequeue and flush the queue.
 	 */
-	fal_qm_enqueue_ctrl_set(0, npq->q.ucast_qid, false);
-	fal_scheduler_dequeue_ctrl_set(0, npq->q.ucast_qid, false);
+	fal_qm_enqueue_ctrl_set(0, npq->q.ucast_qid, 0);
+	fal_scheduler_dequeue_ctrl_set(0, npq->q.ucast_qid, 0);
 	fal_queue_flush(0, port_num, npq->q.ucast_qid);
 
 	nss_qdisc_info("Disable SSDK level0 queue scheduler successful\n");
@@ -370,8 +370,8 @@ static void nss_ppe_queue_enable(struct
 	/*
 	 * Enable queue enqueue and dequeue.
 	 */
-	fal_qm_enqueue_ctrl_set(0, npq->q.ucast_qid, true);
-	fal_scheduler_dequeue_ctrl_set(0, npq->q.ucast_qid, true);
+	fal_qm_enqueue_ctrl_set(0, npq->q.ucast_qid, 1);
+	fal_scheduler_dequeue_ctrl_set(0, npq->q.ucast_qid, 1);
 
 	nss_qdisc_info("Enable SSDK level0 queue scheduler successful\n");
 }
@@ -535,14 +535,14 @@ static void nss_ppe_all_queue_disable(ui
 	 * Disable queue enqueue, dequeue and flush the queue.
 	 */
 	for (offset = 0; offset < nss_ppe_max_get(port_num, NSS_PPE_UCAST_QUEUE); offset++) {
-		fal_qm_enqueue_ctrl_set(0, qid + offset, false);
-		fal_scheduler_dequeue_ctrl_set(0, qid + offset, false);
+		fal_qm_enqueue_ctrl_set(0, qid + offset, 0);
+		fal_scheduler_dequeue_ctrl_set(0, qid + offset, 0);
 		fal_queue_flush(0, port_num, qid + offset);
 	}
 
 	for (offset = 0; offset < nss_ppe_max_get(port_num, NSS_PPE_MCAST_QUEUE); offset++) {
-		fal_qm_enqueue_ctrl_set(0, mcast_qid + offset, false);
-		fal_scheduler_dequeue_ctrl_set(0, mcast_qid + offset, false);
+		fal_qm_enqueue_ctrl_set(0, mcast_qid + offset, 0);
+		fal_scheduler_dequeue_ctrl_set(0, mcast_qid + offset, 0);
 		fal_queue_flush(0, port_num, mcast_qid + offset);
 	}
 
@@ -563,13 +563,13 @@ static void nss_ppe_all_queue_enable(uin
 	 * Enable queue enqueue and dequeue.
 	 */
 	for (offset = 0; offset < nss_ppe_max_get(port_num, NSS_PPE_UCAST_QUEUE); offset++) {
-		fal_qm_enqueue_ctrl_set(0, qid + offset, true);
-		fal_scheduler_dequeue_ctrl_set(0, qid + offset, true);
+		fal_qm_enqueue_ctrl_set(0, qid + offset, 1);
+		fal_scheduler_dequeue_ctrl_set(0, qid + offset, 1);
 	}
 
 	for (offset = 0; offset < nss_ppe_max_get(port_num, NSS_PPE_MCAST_QUEUE); offset++) {
-		fal_qm_enqueue_ctrl_set(0, mcast_qid + offset, true);
-		fal_scheduler_dequeue_ctrl_set(0, mcast_qid + offset, true);
+		fal_qm_enqueue_ctrl_set(0, mcast_qid + offset, 1);
+		fal_scheduler_dequeue_ctrl_set(0, mcast_qid + offset, 1);
 	}
 
 	nss_qdisc_info("Enable SSDK level0 queue scheduler successful\n");
@@ -589,15 +589,15 @@ static void nss_ppe_assigned_queue_enabl
 	spin_lock_bh(&ppe_port->lock);
 	res = ppe_port->res_used[NSS_PPE_UCAST_QUEUE];
 	while (res) {
-		fal_qm_enqueue_ctrl_set(0, qid + res->offset, true);
-		fal_scheduler_dequeue_ctrl_set(0, qid + res->offset, true);
+		fal_qm_enqueue_ctrl_set(0, qid + res->offset, (a_bool_t)true);
+		fal_scheduler_dequeue_ctrl_set(0, qid + res->offset, (a_bool_t)true);
 		res = res->next;
 	}
 
 	res = ppe_port->res_used[NSS_PPE_MCAST_QUEUE];
 	while (res) {
-		fal_qm_enqueue_ctrl_set(0, mcast_qid + res->offset, true);
-		fal_scheduler_dequeue_ctrl_set(0, mcast_qid + res->offset, true);
+		fal_qm_enqueue_ctrl_set(0, mcast_qid + res->offset, (a_bool_t)true);
+		fal_scheduler_dequeue_ctrl_set(0, mcast_qid + res->offset, (a_bool_t)true);
 		res = res->next;
 	}
 
@@ -642,7 +642,7 @@ static int nss_ppe_l1_queue_scheduler_co
 	l1cfg.e_pri = NSS_PPE_PRIORITY_MAX - npq->scheduler.priority;
 	l1cfg.c_drr_id = npq->l1c_drrid;
 	l1cfg.e_drr_id = npq->l1e_drrid;
-	l1cfg.drr_frame_mode = NSS_PPE_FRAME_CRC;
+	l1cfg.drr_frame_mode = (fal_qos_drr_frame_mode_t)NSS_PPE_FRAME_CRC;
 
 	nss_qdisc_trace("SSDK level1 configuration: Port:%d, l0spid:%d, c_drrid:%d, c_pri:%d, c_drr_wt:%d, e_drrid:%d, e_pri:%d, e_drr_wt:%d, l1spid:%d\n",
 			port_num, npq->l0spid, l1cfg.c_drr_id, l1cfg.c_pri, l1cfg.c_drr_wt, l1cfg.e_drr_id, l1cfg.e_pri, l1cfg.e_drr_wt, l1cfg.sp_id);
@@ -935,7 +935,7 @@ static int nss_ppe_l0_queue_scheduler_co
 	l0cfg.e_pri = NSS_PPE_PRIORITY_MAX - npq->scheduler.priority;
 	l0cfg.c_drr_id = npq->l0c_drrid;
 	l0cfg.e_drr_id = npq->l0e_drrid;
-	l0cfg.drr_frame_mode = NSS_PPE_FRAME_CRC;
+	l0cfg.drr_frame_mode = (fal_qos_drr_frame_mode_t)NSS_PPE_FRAME_CRC;
 
 	nss_qdisc_trace("SSDK level0 configuration: Port:%d, ucast_qid:%d, c_drrid:%d, c_pri:%d, c_drr_wt:%d, e_drrid:%d, e_pri:%d, e_drr_wt:%d, l0spid:%d\n",
 			port_num, npq->q.ucast_qid, l0cfg.c_drr_id, l0cfg.c_pri, l0cfg.c_drr_wt, l0cfg.e_drr_id, l0cfg.e_pri, l0cfg.e_drr_wt, l0cfg.sp_id);
@@ -1059,7 +1059,7 @@ static int nss_ppe_port_shaper_set(struc
 	cfg.c_shaper_en = 1;
 	cfg.cbs = npq->shaper.cburst;
 	cfg.cir = (npq->shaper.crate / 1000) * 8;
-	cfg.shaper_frame_mode = NSS_PPE_FRAME_CRC;
+	cfg.shaper_frame_mode = (fal_shaper_frame_mode_t)NSS_PPE_FRAME_CRC;
 
 	/*
 	 * Take HW scaling into consideration
@@ -1141,7 +1141,7 @@ static int nss_ppe_flow_shaper_set(struc
 	cfg.e_shaper_en = 1;
 	cfg.ebs = npq->shaper.cburst;
 	cfg.eir = ((npq->shaper.crate / 1000) * 8) - cfg.cir;
-	cfg.shaper_frame_mode = NSS_PPE_FRAME_CRC;
+	cfg.shaper_frame_mode = (fal_shaper_frame_mode_t)NSS_PPE_FRAME_CRC;
 
 	/*
 	 * Take HW scaling into consideration
@@ -1225,7 +1225,7 @@ static int nss_ppe_queue_shaper_set(stru
 	cfg.e_shaper_en = 1;
 	cfg.ebs = npq->shaper.cburst;
 	cfg.eir = ((npq->shaper.crate / 1000) * 8) - cfg.cir;
-	cfg.shaper_frame_mode = NSS_PPE_FRAME_CRC;
+	cfg.shaper_frame_mode = (fal_shaper_frame_mode_t)NSS_PPE_FRAME_CRC;
 
 	/*
 	 * Take HW scaling into consideration
@@ -1293,7 +1293,6 @@ static void nss_ppe_attach_free(uint32_t
 	spin_unlock_bh(&ppe_port->lock);
 
 	nss_qdisc_info("port:%d, type:%d, res:%px\n", port, res->type, res);
-	return;
 }
 
 /*
@@ -2379,7 +2378,7 @@ int nss_ppe_init(struct Qdisc *sch, stru
 		 * nothing useful and thus we don't allocate any resource".
 		 */
 		nss_qdisc_trace("Qdisc parent = %px, handle=%x\n", nq->parent,  nq->parent->qos_tag);
-		if ((nq->parent->npq.sub_type == NSS_SHAPER_CONFIG_PPE_SN_TYPE_HTB)) {
+		if (nq->parent->npq.sub_type == NSS_SHAPER_CONFIG_PPE_SN_TYPE_HTB) {
 			nq->npq.level = nq->parent->npq.level;
 		} else {
 			nq->npq.level = nq->parent->npq.level - 1;
--- a/nss_qdisc/nss_ppe_mc.c
+++ b/nss_qdisc/nss_ppe_mc.c
@@ -32,8 +32,8 @@ int nss_ppe_mcast_queue_reset(struct nss
 		return 0;
 	}
 
-	fal_qm_enqueue_ctrl_set(0, npq->q.mcast_qid, false);
-	fal_scheduler_dequeue_ctrl_set(0, npq->q.mcast_qid, false);
+	fal_qm_enqueue_ctrl_set(0, npq->q.mcast_qid, 0);
+	fal_scheduler_dequeue_ctrl_set(0, npq->q.mcast_qid, 0);
 	fal_queue_flush(0, port_num, npq->q.mcast_qid);
 
 	/*
@@ -153,8 +153,8 @@ int nss_ppe_mcast_queue_set(struct nss_q
 		}
 	}
 
-	fal_qm_enqueue_ctrl_set(0, npq->q.mcast_qid, true);
-	fal_scheduler_dequeue_ctrl_set(0, npq->q.mcast_qid, true);
+	fal_qm_enqueue_ctrl_set(0, npq->q.mcast_qid, 1);
+	fal_scheduler_dequeue_ctrl_set(0, npq->q.mcast_qid, 1);
 
 	nss_qdisc_info("SSDK multicast queue configuration successful for port:%d\n", port_num);
 	return 0;
@@ -164,4 +164,4 @@ fail:
 	npq->q.mcast_qid = 0;
 	npq->q.mcast_valid = 0;
 	return -EINVAL;
-}
\ No newline at end of file
+}
--- a/nss_qdisc/nss_prio.c
+++ b/nss_qdisc/nss_prio.c
@@ -341,7 +341,7 @@ static int nss_prio_graft(struct Qdisc *
 
 	nss_qdisc_info("Grafting old: %px with new: %px\n", *old, new);
 	if (*old != &noop_qdisc) {
-		struct nss_qdisc *nq_old = qdisc_priv(*old);
+		struct nss_qdisc *nq_old = qdisc_priv(((struct Qdisc *)(*old)));
 		nss_qdisc_info("Detaching old: %px\n", *old);
 		nim_detach.msg.shaper_configure.config.msg.shaper_node_config.qos_tag = q->nq.qos_tag;
 
--- a/nss_qdisc/nss_qdisc.c
+++ b/nss_qdisc/nss_qdisc.c
@@ -14,6 +14,7 @@
  **************************************************************************
  */
 
+#include <linux/timer.h>
 #include <nss_api_if.h>
 #include "nss_qdisc.h"
 #include "nss_fifo.h"
@@ -374,7 +375,6 @@ static int nss_qdisc_refresh_bshaper_ass
 	br_update.port_list_count = 0;
 	br_update.unassign_count = 0;
 
-	read_lock(&dev_base_lock);
 	dev = first_net_device(&init_net);
 
 	while(dev) {
@@ -421,7 +421,6 @@ static int nss_qdisc_refresh_bshaper_ass
 nextdev:
 		dev = next_net_device(dev);
 	}
-	read_unlock(&dev_base_lock);
 
 	nss_qdisc_info("List count %d\n", br_update.port_list_count);
 
@@ -1140,15 +1139,16 @@ unsigned int nss_qdisc_drop(struct Qdisc
 {
 	struct nss_qdisc *nq = qdisc_priv(sch);
 	unsigned int ret;
+	struct sk_buff *to_free = qdisc_peek_head(sch);
 
 	if (!nq->is_virtual) {
-		ret = __qdisc_queue_drop_head(sch, &sch->q);
+		ret = __qdisc_queue_drop_head(sch, &sch->q, &to_free);
 	} else {
 		spin_lock_bh(&nq->bounce_protection_lock);
 		/*
 		 * This function is safe to call within locks
 		 */
-		ret = __qdisc_queue_drop_head(sch, &sch->q);
+		ret = __qdisc_queue_drop_head(sch, &sch->q, &to_free);
 		spin_unlock_bh(&nq->bounce_protection_lock);
 	}
 
@@ -1212,7 +1212,7 @@ static bool nss_qdisc_iterate_fl(struct
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0))
 	status = tc_classify(skb, tcf, &res, false);
 #else
-	status = tcf_classify(skb, tcf, &res, false);
+	status = tcf_classify(skb, NULL, tcf, &res, false);
 #endif
 	if ((status == TC_ACT_STOLEN) || (status == TC_ACT_QUEUED)) {
 		return 1;
@@ -2188,7 +2188,7 @@ int __nss_qdisc_init(struct Qdisc *sch,
 	 * This is to prevent mixing NSS and PPE qdisc with linux qdisc.
 	 */
 	if ((parent != TC_H_ROOT) && (root->ops->owner != THIS_MODULE)) {
-		nss_qdisc_warning("NSS qdisc %px (type %d) used along with non-nss qdiscs,"
+		nss_qdisc_info("NSS qdisc %px (type %d) used along with non-nss qdiscs,"
 			" or the interface is currently down", nq->qdisc, nq->type);
 	}
 
@@ -2605,14 +2605,14 @@ int nss_qdisc_init(struct Qdisc *sch, st
  *  Wrapper around gnet_stats_copy_basic()
  */
 int nss_qdisc_gnet_stats_copy_basic(struct Qdisc *sch, struct gnet_dump *d,
-				struct gnet_stats_basic_packed *b)
+				struct gnet_stats_basic_sync *b)
 {
 #if (LINUX_VERSION_CODE <= KERNEL_VERSION(3, 18, 0))
 	return gnet_stats_copy_basic(d, b);
 #elif (LINUX_VERSION_CODE < KERNEL_VERSION(4, 8, 0))
 	return gnet_stats_copy_basic(d, NULL, b);
 #else
-	return gnet_stats_copy_basic(qdisc_root_sleeping_running(sch), d, NULL, b);
+	return gnet_stats_copy_basic(d, NULL, b, true);
 #endif
 }
 
@@ -2952,3 +2952,4 @@ module_init(nss_qdisc_module_init)
 module_exit(nss_qdisc_module_exit)
 
 MODULE_LICENSE("Dual BSD/GPL");
+MODULE_INFO(description, "NSS qdisc client");
--- a/nss_qdisc/nss_qdisc.h
+++ b/nss_qdisc/nss_qdisc.h
@@ -14,6 +14,7 @@
  **************************************************************************
  */
 
+#include <linux/timer.h>
 #include <linux/module.h>
 #include <linux/of.h>
 #include <linux/types.h>
@@ -217,7 +218,7 @@ struct nss_qdisc {
 						/* Shaper configure callback for reading shaper specific
 						 * responses (e.g. memory size).
 						 */
-	struct gnet_stats_basic_packed bstats;	/* Basic class statistics */
+	struct gnet_stats_basic_sync bstats;	/* Basic class statistics */
 	struct gnet_stats_queue qstats;		/* Qstats for use by classes */
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 13, 0))
 	atomic_t refcnt;			/* Reference count for class use */
@@ -464,7 +465,7 @@ extern int nss_qdisc_init(struct Qdisc *
  *  Wrapper around gnet_stats_copy_basic()
  */
 extern int nss_qdisc_gnet_stats_copy_basic(struct Qdisc *sch,
-				struct gnet_dump *d, struct gnet_stats_basic_packed *b);
+				struct gnet_dump *d, struct gnet_stats_basic_sync *b);
 
 /*
  * nss_qdisc_gnet_stats_copy_queue()
--- a/nss_qdisc/nss_qdisc_stats.c
+++ b/nss_qdisc/nss_qdisc_stats.c
@@ -17,6 +17,7 @@
  **************************************************************************
  */
 
+#include <linux/timer.h>
 #include "nss_qdisc.h"
 
 static LIST_HEAD(nss_qdisc_stats_list);			/* List of stats request nodes */
@@ -82,7 +83,7 @@ static void nss_qdisc_stats_sync_restart
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
 	struct nss_qdisc *nq = (struct nss_qdisc *)data;
 #else
-	struct nss_qdisc_stats_wq *nqsw = from_timer(nqsw, tm, stats_get_timer);
+	struct nss_qdisc_stats_wq *nqsw = timer_container_of(nqsw, tm, stats_get_timer);
 	struct nss_qdisc *nq = nqsw->nq;
 #endif
 
@@ -161,7 +162,7 @@ static void nss_qdisc_stats_process_node
 {
 	struct Qdisc *qdisc;
 	struct nss_qdisc *nq;
-	struct gnet_stats_basic_packed *bstats;
+	struct gnet_stats_basic_sync *bstats;
 	struct gnet_stats_queue *qstats;
 	uint32_t qos_tag = response->qos_tag;
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 13, 0))
@@ -215,8 +216,8 @@ static void nss_qdisc_stats_process_node
 	 * Update qdisc->bstats
 	 */
 	spin_lock_bh(&nq->lock);
-	bstats->bytes += (__u64)response->sn_stats.delta.dequeued_bytes;
-	bstats->packets += response->sn_stats.delta.dequeued_packets;
+	u64_stats_add(&bstats->bytes, (__u64)response->sn_stats.delta.dequeued_bytes);
+	u64_stats_add(&bstats->packets, response->sn_stats.delta.dequeued_packets);
 
 	/*
 	 * Update qdisc->qstats
@@ -490,7 +491,7 @@ void nss_qdisc_stats_sync_many_exit(stru
 	 * We will have to wait until we have received the pending response.
 	 */
 cleanup:
-	if (del_timer(&nqsw->stats_get_timer) > 0) {
+	if (timer_delete(&nqsw->stats_get_timer) > 0) {
 		/*
 		 * The timer was still active (counting down) when it was deleted.
 		 * Therefore we are sure that there are no pending stats request
--- a/nss_qdisc/nss_tbl.c
+++ b/nss_qdisc/nss_tbl.c
@@ -350,7 +350,7 @@ static int nss_tbl_graft(struct Qdisc *s
 
 	nss_qdisc_info("Grafting old: %px with new: %px\n", *old, new);
 	if (*old != &noop_qdisc) {
-		struct nss_qdisc *nq_old = (struct nss_qdisc *)qdisc_priv(*old);
+		struct nss_qdisc *nq_old = (struct nss_qdisc *)qdisc_priv(((struct Qdisc *)(*old)));
 		nss_qdisc_info("Detaching old: %px\n", *old);
 		nim_detach.msg.shaper_configure.config.msg.shaper_node_config.qos_tag = q->nq.qos_tag;
 		if (nss_qdisc_node_detach(&q->nq, nq_old, &nim_detach,
--- a/nss_qdisc/nss_wred.c
+++ b/nss_qdisc/nss_wred.c
@@ -291,7 +291,7 @@ static int nss_wred_change(struct Qdisc
 
 	nim.msg.shaper_configure.config.msg.shaper_node_config.qos_tag = q->nq.qos_tag;
 	nim.msg.shaper_configure.config.msg.shaper_node_config.snc.wred_param.limit = qopt->limit;
-	nim.msg.shaper_configure.config.msg.shaper_node_config.snc.wred_param.weight_mode = qopt->weight_mode;
+	nim.msg.shaper_configure.config.msg.shaper_node_config.snc.wred_param.weight_mode = (nss_shaper_config_wred_weight_mode_t)qopt->weight_mode;
 	nim.msg.shaper_configure.config.msg.shaper_node_config.snc.wred_param.weight_mode_value = qopt->weight_mode_value;
 	nim.msg.shaper_configure.config.msg.shaper_node_config.snc.wred_param.rap.min = qopt->rap.min;
 	nim.msg.shaper_configure.config.msg.shaper_node_config.snc.wred_param.rap.max = qopt->rap.max;
--- a/nss_qdisc/nss_wrr.c
+++ b/nss_qdisc/nss_wrr.c
@@ -400,7 +400,7 @@ failure:
 	return -EINVAL;
 }
 
-static int nss_wrr_delete_class(struct Qdisc *sch, unsigned long arg)
+static int nss_wrr_delete_class(struct Qdisc *sch, unsigned long arg, struct netlink_ext_ack *extack)
 {
 	struct nss_wrr_sched_data *q = qdisc_priv(sch);
 	struct nss_wrr_class_data *cl = (struct nss_wrr_class_data *)arg;
@@ -484,7 +484,7 @@ static int nss_wrr_graft_class(struct Qd
 	 */
 	nss_qdisc_info("Grafting old: %px with new: %px\n", *old, new);
 	if (*old != &noop_qdisc) {
-		struct nss_qdisc *nq_child = qdisc_priv(*old);
+		struct nss_qdisc *nq_child = qdisc_priv(((struct Qdisc *)(*old)));
 		nss_qdisc_info("Detaching old: %px\n", *old);
 		nim_detach.msg.shaper_configure.config.msg.shaper_node_config.qos_tag = cl->nq.qos_tag;
 		if (nss_qdisc_node_detach(&cl->nq, nq_child, &nim_detach,
