--- a/mc_forward.c
+++ b/mc_forward.c
@@ -31,6 +31,9 @@
 #include "mc_snooping.h"
 #include "mc_osdep.h"
 
+int mc_forward_init(void) __maybe_unused;
+void mc_forward_exit(void) __maybe_unused;
+
 static void mc_retag(void *iph, __be16 etype, __be32 dscp)
 {
 	__be32 _dscp = MC_DSCP(dscp);
--- a/mc_main.c
+++ b/mc_main.c
@@ -77,5 +77,6 @@ static void __exit mc_exit(void)
 
 module_init(mc_init);
 module_exit(mc_exit);
+MODULE_INFO(description, "QCA multicast snooping");
 MODULE_LICENSE("GPL v2");
 
--- a/mc_osdep.h
+++ b/mc_osdep.h
@@ -25,7 +25,7 @@
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 0, 0))
 static inline int os_br_pass_frame_up(struct sk_buff *skb)
 {
-	return br_pass_frame_up(skb);
+	return br_pass_frame_up(skb, false);
 }
 #else
 static inline int os_br_pass_frame_up(struct sk_buff *skb)
--- a/mc_snooping.c
+++ b/mc_snooping.c
@@ -33,7 +33,7 @@
 #include <linux/etherdevice.h>
 #include <linux/rculist.h>
 #include <linux/timer.h>
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 #include <asm/atomic.h>
 #include <linux/random.h>
 #include <net/ip.h>
@@ -645,7 +645,7 @@ static void mc_mdb_destroy(struct mc_mdb
 	}
 
 	hlist_del_rcu(&mdb->hlist);
-	del_timer_sync(&mdb->etimer);
+	timer_delete_sync(&mdb->etimer);
 
 	if (mdb->flood_ifcnt || mdb->encap_dev_cnt) {
 		mdb->flood_ifcnt = 0;
@@ -677,7 +677,7 @@ static void mc_agingtimer_reset(struct m
 
 	if (timer_pending(&mc->agingtimer) ?
 			time_after(mc->agingtimer.expires, jiffies + expires) :
-				try_to_del_timer_sync(&mc->agingtimer) >= 0) {
+				timer_delete_sync_try(&mc->agingtimer) >= 0) {
 		mod_timer(&mc->agingtimer, jiffies + expires);
 		MC_PRINT("Reset Group Membership Interval ageing timer, expires = %u\n",
 				jiffies_to_msecs(expires) / 1000);
@@ -705,7 +705,7 @@ static void mc_rtimer_reset(struct mc_st
 
 	if (timer_pending(&mc->rtimer) ?
 			time_after(mc->rtimer.expires, jiffies + expires) :
-				try_to_del_timer_sync(&mc->rtimer) >= 0) {
+				timer_delete_sync_try(&mc->rtimer) >= 0) {
 		mod_timer(&mc->rtimer, jiffies + expires);
 		MC_PRINT("Reset Querier Interval ageing timer, expires = %u\n",
 				jiffies_to_msecs(expires) / 1000);
@@ -722,7 +722,7 @@ static void mc_mdb_expired(unsigned long
 #endif
 {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0))
-	struct mc_mdb_entry *mdb = from_timer(mdb, tl, etimer);
+	struct mc_mdb_entry *mdb = timer_container_of(mdb, tl, etimer);
 #else
 	struct mc_mdb_entry *mdb = (struct mc_mdb_entry *)data;
 #endif
@@ -1215,7 +1215,7 @@ static void mc_leave_group(struct mc_str
 	fg->ageing_timer = now - expire_time + expire_delay;
 	if (timer_pending(&mc->agingtimer) ?
 		time_after(mc->agingtimer.expires, jiffies + expire_delay) :
-		try_to_del_timer_sync(&mc->agingtimer) >= 0) {
+		timer_delete_sync_try(&mc->agingtimer) >= 0) {
 		mod_timer(&mc->agingtimer, jiffies + expire_delay);
 		MC_PRINT("Reset Group Membership Interval ageing timer, expires = %u ms\n",
 			leave_delay_msecond);
@@ -2082,7 +2082,7 @@ static void mc_query_cycle_start(struct
 			goto out;
 		if (timer_pending(&mdb->etimer) ?
 				time_after(mdb->etimer.expires, now + root_qe->max_resp_time) :
-				try_to_del_timer_sync(&mdb->etimer) >= 0) {
+				timer_delete_sync_try(&mdb->etimer) >= 0) {
 			mdb->timer_base = now;
 			mod_timer(&mdb->etimer, now + root_qe->max_resp_time);
 			mc_query_timer_dump(mdb, root_qe->max_resp_time, 1);
@@ -2097,7 +2097,7 @@ static void mc_query_cycle_start(struct
 					continue;
 				if (timer_pending(&mdb->etimer) ?
 						time_after(mdb->etimer.expires, now + root_qe->max_resp_time) :
-						try_to_del_timer_sync(&mdb->etimer) >= 0) {
+						timer_delete_sync_try(&mdb->etimer) >= 0) {
 					mdb->timer_base = now;
 					mod_timer(&mdb->etimer, now + root_qe->max_resp_time);
 					mc_query_timer_dump(mdb, root_qe->max_resp_time, 0);
@@ -2948,10 +2948,10 @@ static int mc_stop(struct mc_struct *mc)
 	mc_mdb_flush(mc);
 	mc_rlist_flush(mc);
 
-	del_timer_sync(&mc->qtimer);
-	del_timer_sync(&mc->agingtimer);
-	del_timer_sync(&mc->rtimer);
-	del_timer_sync(&mc->evtimer);
+	timer_delete_sync(&mc->qtimer);
+	timer_delete_sync(&mc->agingtimer);
+	timer_delete_sync(&mc->rtimer);
+	timer_delete_sync(&mc->evtimer);
 
 	return 0;
 }
@@ -2977,7 +2977,7 @@ static void mc_router_cleanup(unsigned l
 {
 	int delay_reset = 0;
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0))
-	struct mc_struct *mc = from_timer(mc, tl, rtimer);
+	struct mc_struct *mc = timer_container_of(mc, tl, rtimer);
 #else
 	struct mc_struct *mc = (struct mc_struct *)data;
 #endif
@@ -3032,7 +3032,7 @@ static void mc_mdb_query(unsigned long d
 {
 	int i;
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0))
-	struct mc_struct *mc = from_timer(mc, tl, qtimer);
+	struct mc_struct *mc = timer_container_of(mc, tl, qtimer);
 #else
 	struct mc_struct *mc = (struct mc_struct *)data;
 #endif
@@ -3087,7 +3087,7 @@ static void mc_mdb_cleanup(unsigned long
 {
 	int i;
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0))
-	struct mc_struct *mc = from_timer(mc, tl, agingtimer);
+	struct mc_struct *mc = timer_container_of(mc, tl, agingtimer);
 #else
 	struct mc_struct *mc = (struct mc_struct *)data;
 #endif
@@ -3118,7 +3118,6 @@ static void mc_mdb_cleanup(unsigned long
 		os_hlist_for_each_entry_rcu(mdb, mdbh, &mc->hash[i], hlist) {
 			struct mc_port_group *pg;
 			struct hlist_node *pgh;
-			struct net_device *port;
 			unsigned long expire_time = mc->membership_interval;
 
 			if (hlist_empty(&mdb->pslist)) {
@@ -3153,7 +3152,6 @@ static void mc_mdb_cleanup(unsigned long
 					else if (time_before(this_timer, next_timer))
 						next_timer = this_timer;
 				}
-				port = pg->port;
 			}
 		}
 	}
@@ -3254,7 +3252,6 @@ static void mc_acl_table_init(struct mc_
 	mc->mld_acl.pattern_count = 4;
 #endif
 
-	return;
 }
 
 /* mc_event_delay
@@ -3267,7 +3264,7 @@ static void mc_event_delay(unsigned long
 #endif
 {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0))
-	struct mc_struct *mc = from_timer(mc, tl, evtimer);
+	struct mc_struct *mc = timer_container_of(mc, tl, evtimer);
 
 	mc_netlink_event_send(mc, MC_EVENT_MDB_UPDATED,	0, NULL);
 #else
--- a/mc_snooping.h
+++ b/mc_snooping.h
@@ -44,6 +44,8 @@ void mc_detach(struct net_device *dev);
 
 int mc_has_more_instance(void);
 
+int mc_proc_create_snooper_entry(void);
+
 extern void mc_nbp_change(struct mc_struct *mc, struct net_device *dev, int event);
 
 extern void mc_fdb_change(__u8 *mac, int change);
