--- a/mc_netfilter.c
+++ b/mc_netfilter.c
@@ -26,6 +26,7 @@
 #include "mc_private.h"
 #include "mc_api.h"
 #include "mc_osdep.h"
+#include "mc_netfilter.h"
 
 /* mc_br_port_get
  *	get bridge port by ifindex
@@ -49,7 +50,7 @@ static struct net_bridge_port *mc_br_por
 /* mc_pre_routing_hook
  *	prerouting hook
  */
-static unsigned int mc_pre_routing_hook(void *priv,
+unsigned int mc_pre_routing_hook(void *priv,
 				    struct sk_buff *skb,
 				    const struct nf_hook_state *state)
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 13, 0))
@@ -233,7 +234,7 @@ static bool mc_is_report_or_leave(struct
 /* mc_forward_hook
  *	forward hook to the linux kernel
  */
-static unsigned int mc_forward_hook(void *priv,
+unsigned int mc_forward_hook(void *priv,
 				    struct sk_buff *skb,
 				    const struct nf_hook_state *state)
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 13, 0))
--- a/mc_netfilter.h
+++ b/mc_netfilter.h
@@ -24,13 +24,13 @@ void mc_netfilter_exit(void);
 #include <linux/netdevice.h>
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 13, 0))
-unsigned int mc_pre_routing_hook(const struct nf_hook_ops *ops, struct sk_buff *skb,
-				 const struct net_device *in, const struct net_device *out,
-				 int(*okfn)(struct sk_buff *));
+unsigned int mc_pre_routing_hook(void *priv,
+				    struct sk_buff *skb,
+				    const struct nf_hook_state *state);
 
-unsigned int mc_forward_hook(const struct nf_hook_ops *ops, struct sk_buff *skb,
-			     const struct net_device *in, const struct net_device *out,
-			     int(*okfn)(struct sk_buff *));
+unsigned int mc_forward_hook(void *priv,
+				    struct sk_buff *skb,
+				    const struct nf_hook_state *state);
 #else
 unsigned int mc_pre_routing_hook(unsigned int hooknum, struct sk_buff *skb,
 				 const struct net_device *in, const struct net_device *out,
--- a/mc_netlink.c
+++ b/mc_netlink.c
@@ -776,7 +776,7 @@ static void __mc_netlink_receive(struct
 					mc->local_query_interval = i->val * HZ;
 					if (timer_pending(&mc->qtimer) ?
 						time_after(mc->qtimer.expires, jiffies + mc->local_query_interval) :
-							try_to_del_timer_sync(&mc->qtimer) >= 0) {
+							timer_delete_sync_try(&mc->qtimer) >= 0) {
 						if (mc->started)
 							mod_timer(&mc->qtimer, jiffies + mc->local_query_interval);
 					}
--- a/mc_wifi_events.c
+++ b/mc_wifi_events.c
@@ -24,6 +24,7 @@
 #include "mc_netlink.h"
 #include "mc_snooping.h"
 #include "mc_forward.h"
+#include "mc_wifi_events.h"
 #include <linux/in6.h>
 
 struct sta_leave_event {
@@ -81,7 +82,7 @@ static int mc_event_recv(struct socket *
  * mc_sta_leave_event()
  *	Delete the mac entry of sta from bridge snooping table
  */
-void mc_sta_leave_event(struct net_device *dev,__u8 *from_mac)
+static void mc_sta_leave_event(struct net_device *dev,__u8 *from_mac)
 {
 
 	struct mc_struct *mc;
@@ -154,7 +155,7 @@ void mc_sta_leave_event(struct net_devic
  * mc_process_sta_leave_event()
  * 	Parse and process sta_leave event
  */
-void mc_process_sta_leave_event(struct nlmsghdr *nlh, int cmd)
+static void mc_process_sta_leave_event(struct nlmsghdr *nlh, int cmd)
 {
 	struct nlattr **tb = NULL;
 	uint8_t mac[ETH_ALEN] = {0};
@@ -438,7 +439,7 @@ void mc_handle_vbss_event(struct net_dev
 /* mc_process_vendor_event()
  *	Process the vendor events
  */
-void mc_process_vendor_event(struct nlmsghdr *nlh, int cmd)
+static void mc_process_vendor_event(struct nlmsghdr *nlh, int cmd)
 {
 	struct net_device *dev;
 	u_int32_t ifidx;
@@ -557,7 +558,7 @@ static int mc_event_handler(void *buf, i
  * 	Parse the generic control family response message and return
  * 	the multicast id of MLME and VENDOR multicast groups of nl80211 family.
  */
-int mc_get_event_id(struct nlmsghdr *nlh, const char *event_name, int *event_id)
+static int mc_get_event_id(struct nlmsghdr *nlh, const char *event_name, int *event_id)
 {
     struct nlattr *tb[CTRL_ATTR_MAX+1];
     struct nlattr *mcgrp;
@@ -608,7 +609,7 @@ int mc_get_event_id(struct nlmsghdr *nlh
  *	Get the multicast id of MLME and VENDOR multicast groups of nl80211 family.
  */
 
-int mc_parse_genl_ctrl_response(struct nlmsghdr *nlh, int *mlme_id, int *vendor_id)
+static int mc_parse_genl_ctrl_response(struct nlmsghdr *nlh, int *mlme_id, int *vendor_id)
 {
 	int res_mlme, res_vendor;
 
@@ -642,7 +643,7 @@ int mc_parse_genl_ctrl_response(struct n
  *	Construct a message to generic control family to resolve the
  *	nl80211 multicast groups.
  */
-struct nlmsghdr *mc_construct_nl_message(void)
+static struct nlmsghdr *mc_construct_nl_message(void)
 {
 	int flags = GFP_ATOMIC;
 	struct sk_buff *skb;
@@ -705,7 +706,7 @@ err:
  *	Resolve the nl80211 family multicast groups to
  *	receive Wi-Fi specific events.
  */
-int mc_resolve_nl80211_family(struct socket *sock, struct sockaddr_nl *addr)
+static int mc_resolve_nl80211_family(struct socket *sock, struct sockaddr_nl *addr)
 {
 	struct nlmsghdr *nlh;
 	struct kvec iov = {0};
