--- a/mc_snooping.c
+++ b/mc_snooping.c
@@ -1596,7 +1596,6 @@ static int mc_ipv4_igmp3_report(struct m
 
 		filter_num = 0;
 		switch (type) {
-		case IGMPV3_MODE_IS_INCLUDE:
 		case IGMPV3_CHANGE_TO_INCLUDE:
 			if (!grec->grec_nsrcs) {
 				mc_ipv4_leave_group(mc, group, skb, from_mac, port);
@@ -1670,6 +1669,7 @@ static int mc_ipv4_igmp3_report(struct m
 			/* CHECK X */
 			filter_num = mc_ipv4_source_list_filter(mdb, grec, (__be32 *)mdb->x.srcs, mdb->x.nsrcs);
 			break;
+		case IGMPV3_MODE_IS_INCLUDE:
 		case IGMPV3_ALLOW_NEW_SOURCES:
 			mc_source_list_update(fg, (__u8 *)grec->grec_src, MC_GREC_NSRCS(ntohs(grec->grec_nsrcs)),
 					sizeof(__be32), 0, IGMPV3_ALLOW_NEW_SOURCES);
@@ -1896,7 +1896,6 @@ static int mc_ipv6_mld2_report(struct mc
 
 		filter_num = 0;
 		switch (grec->grec_type) {
-		case MLD2_MODE_IS_INCLUDE:
 		case MLD2_CHANGE_TO_INCLUDE:
 			if (!grec->grec_nsrcs) {
 				mc_ipv6_leave_group(mc, &grec->grec_mca, skb, from_mac, port);
@@ -1968,6 +1967,7 @@ static int mc_ipv6_mld2_report(struct mc
 			/* CHECK X */
 			filter_num = mc_ipv6_source_list_filter(mdb, grec, (struct in6_addr *)mdb->x.srcs, mdb->x.nsrcs);
 			break;
+		case MLD2_MODE_IS_INCLUDE:
 		case MLD2_ALLOW_NEW_SOURCES:
 			mc_source_list_update(fg, (__u8 *)grec->grec_src, MC_GREC_NSRCS(ntohs(grec->grec_nsrcs)),
 					sizeof(struct in6_addr), 0, MLD2_ALLOW_NEW_SOURCES);
