--- a/include/hsl/phy/hsl_phy.h
+++ b/include/hsl/phy/hsl_phy.h
@@ -576,6 +576,8 @@ typedef struct {
 	struct net_device *netdev[SW_MAX_NR_PORT];
 } phy_info_t;
 /*qca808x_end*/
+#define MALIBU5PORT_PHY         0x004DD0B1
+#define MALIBU2PORT_PHY         0x004DD0B2
 #define QCA8030_PHY             0x004DD076
 #define QCA8033_PHY             0x004DD074
 #define QCA8035_PHY             0x004DD072
@@ -1156,7 +1158,8 @@ struct hsl_phy_api *hsl_phy_api_get(a_ui
 			struct nss_phy_ptp_ops *ptp_ops = NULL;					\
 			struct nss_phy_device nss_phydev;					\
 			hsl_port_nss_phy_ops_get(dev_id, port_id, &nss_phydev, &nss_phy_ops);	\
-			ptp_ops = nss_phy_ops->ptp_ops;						\
+			if (nss_phy_ops)							\
+				ptp_ops = nss_phy_ops->ptp_ops;				\
 			if (ptp_ops && ptp_ops->func)						\
 				rv = ptp_ops->func(&nss_phydev, ##__VA_ARGS__);			\
 		}										\
--- a/src/hsl/phy/hsl_phy.c
+++ b/src/hsl/phy/hsl_phy.c
@@ -249,8 +249,10 @@ int ssdk_phy_driver_init(a_uint32_t dev_
 			}
 /*qca808x_start*/
 			phy_id = hsl_phyid_get(dev_id, i);
+#ifdef MHT
 			if (phy_id == QCA8084_PHY)
 				mht_interface_ops_init(dev_id, i);
+#endif
 			phytype = hsl_phytype_get_by_phyid(dev_id, phy_id);
 			if (MAX_PHY_CHIP != phytype) {
 				phy_info[dev_id]->phy_type[i] = phytype;
@@ -1776,7 +1778,7 @@ hsl_phydev_eee_update(a_uint32_t dev_id,
 	rv = hsl_phy_phydev_get(dev_id, phy_addr, &phydev);
 	SW_RTN_ON_ERROR(rv);
 
-	phydev->eee_enabled = (adv == 0 ? A_FALSE : A_TRUE);
+	phydev->eee_cfg.eee_enabled = (adv == 0 ? A_FALSE : A_TRUE);
 
 	linkmode_mod_bit(ETHTOOL_LINK_MODE_100baseT_Full_BIT,
 		phydev->advertising_eee, adv & FAL_PHY_EEE_100BASE_T);
