--- a/src/hsl/mht/mht_interface_ctrl.c
+++ b/src/hsl/mht/mht_interface_ctrl.c
@@ -24,6 +24,7 @@
 #include "ssdk_mht_clk.h"
 #include "ssdk_clk.h"
 
+#include "phylib.h"
 static a_uint16_t
 mht_uniphy_xpcs_mmd_read(a_uint32_t dev_id, a_uint16_t mmd_num,
 	a_uint16_t mmd_reg)
@@ -827,13 +828,12 @@ mht_interface_ops_init(a_uint32_t dev_id
 		return SW_BAD_PTR;
 	}
 
-	if (!phydev->shared->priv) {
-		SSDK_ERROR("phydev->shared->priv is null\n");
+	shared_priv = phy_package_get_priv(phydev);
+	if (!shared_priv) {
+		SSDK_ERROR("phy package private data is null\n");
 		return SW_BAD_PTR;
 	}
 
-	shared_priv = phydev->shared->priv;
-
 	shared_priv->phy_clk_init = ssdk_mht_gcc_clock_init;
 	shared_priv->phy_qusgmii_mode_set = mht_interface_uqxgmii_mode_set;
 	shared_priv->phy_clk_en_set = ssdk_mht_port_clk_en_set;
