--- a/src/ref/ref_mib.c
+++ b/src/ref/ref_mib.c
@@ -47,6 +47,10 @@
 #include "ssdk_plat.h"
 #include "ref_vlan.h"
 
+static int _qca_ar8327_sw_capture_port_tx_counter(a_uint32_t dev_id, int port);
+static int _qca_ar8327_sw_capture_port_rx_counter(a_uint32_t dev_id, int port);
+void qca_ar8327_sw_mib_task(struct qca_phy_priv *priv);
+
 #if defined(IN_SWCONFIG)
 int
 _qca_ar8327_sw_capture_port_counter(struct switch_dev *dev, int port)
@@ -482,7 +486,7 @@ qca_ar8327_sw_get_port_mib(struct switch
 }
 #endif
 
-int
+static int
 _qca_ar8327_sw_capture_port_tx_counter(a_uint32_t dev_id, int port)
 {
     fal_mib_info_t  mib_Info;
@@ -493,7 +497,7 @@ _qca_ar8327_sw_capture_port_tx_counter(a
     return 0;
 }
 
-int
+static int
 _qca_ar8327_sw_capture_port_rx_counter(a_uint32_t dev_id, int port)
 {
     fal_mib_info_t  mib_Info;
--- a/src/ref/ref_misc.c
+++ b/src/ref/ref_misc.c
@@ -54,6 +54,7 @@
 #include "hsl_phy.h"
 #endif
 
+#if defined(IN_SWCONFIG)
 int
 qca_ar8327_sw_set_max_frame_size(struct switch_dev *dev,
 										const struct switch_attr *attr,
@@ -201,3 +202,4 @@ qca_ar8327_sw_reset_switch(struct switch
 	return rv;
 }
 
+#endif
--- a/src/shell_lib/shell_io.c
+++ b/src/shell_lib/shell_io.c
@@ -979,16 +979,16 @@ cmd_sscanf(const char *buf, const char *
 	if(buf[0] == '0' && (buf[1] == 'x' || buf[1] == 'X'))
 	{
 		if(!fmt)
-			strlcpy(fmt_tmp, "%x", sizeof(fmt_tmp));
+			strscpy(fmt_tmp, "%x", sizeof(fmt_tmp));
 		else
 		{
 			if(strspn(fmt, "%lLxXhH") != strlen(fmt))
 				return SW_BAD_VALUE;
 			if(fmt[0] == '%' && ((fmt[1] == 'l' || fmt[1] == 'L') &&
 				(fmt[2] == 'l' || fmt[2] == 'L')))
-				strlcpy(fmt_tmp, "%llx", sizeof(fmt_tmp));
+				strscpy(fmt_tmp, "%llx", sizeof(fmt_tmp));
 			else
-				strlcpy(fmt_tmp, fmt, sizeof(fmt_tmp));
+				strscpy(fmt_tmp, fmt, sizeof(fmt_tmp));
 		}
 	}
 	else
@@ -996,16 +996,16 @@ cmd_sscanf(const char *buf, const char *
 		if(strspn(buf, "1234567890") != strlen(buf))
 			return SW_BAD_VALUE;
 		if(!fmt)
-			strlcpy(fmt_tmp, "%d", sizeof(fmt_tmp));
+			strscpy(fmt_tmp, "%d", sizeof(fmt_tmp));
 		else
 		{
 			if(strspn(fmt, "%lLdD") != strlen(fmt))
 				return SW_BAD_VALUE;
 			if(fmt[0] == '%' && ((fmt[1] == 'l' || fmt[1] == 'L') &&
 				(fmt[2] == 'l' || fmt[2] == 'L')))
-				strlcpy(fmt_tmp, "%lld", sizeof(fmt_tmp));
+				strscpy(fmt_tmp, "%lld", sizeof(fmt_tmp));
 			else
-				strlcpy(fmt_tmp, fmt, sizeof(fmt_tmp));
+				strscpy(fmt_tmp, fmt, sizeof(fmt_tmp));
 		}
 	}
 	if(sscanf(buf, fmt_tmp, arg_val) != 1)
@@ -2735,7 +2735,7 @@ cmd_data_check_portmap(char *cmdstr, fal
         return SW_OK;
     }
 
-    strlcpy(tmp_str, cmdstr, sizeof(tmp_str));
+    strscpy(tmp_str, cmdstr, sizeof(tmp_str));
     tmp = (void *) strsep(&cmdstr, ",");
     while (tmp)
     {
@@ -11338,7 +11338,7 @@ cmd_data_check_toeplitz_hash_secret_key(
 						break;
 					}
 					/* copy 8 chars from cmd */
-					strlcpy(cmd_word, cmd, sizeof(cmd_word));
+					strscpy(cmd_word, cmd, sizeof(cmd_word));
 					sscanf(cmd_word, "%x", &entry.key[words]);
 					cmd += 8;
 				}
@@ -13042,7 +13042,7 @@ cmd_data_check_tunnel_encap_entry(char *
 						break;
 					}
 					/* copy 2 chars from cmd */
-					strlcpy(cmd_byte, cmd, sizeof(cmd_byte));
+					strscpy(cmd_byte, cmd, sizeof(cmd_byte));
 					sscanf(cmd_byte, "%hhx",
 						&(entry.pkt_header.pkt_header_data[bytes]));
 					cmd += 2;
