--- a/src/init/ssdk_dts.c
+++ b/src/init/ssdk_dts.c
@@ -26,6 +26,7 @@
 #endif
 #include <linux/etherdevice.h>
 #include <linux/clk.h>
+#include <linux/platform_device.h>
 
 #include "ssdk_init.h"
 #include "ssdk_dts.h"
--- a/src/init/ssdk_mht_pinctrl.c
+++ b/src/init/ssdk_mht_pinctrl.c
@@ -287,7 +287,7 @@ static sw_error_t mht_gpio_pin_cfg_set(a
                 rv = mht_gpio_pin_cfg_set_drvs(dev_id, pin, arg);
                 break;
 
-            case PIN_CONFIG_OUTPUT:
+            case PIN_CONFIG_LEVEL:
                 rv = mht_gpio_pin_cfg_set_oe(dev_id, pin, A_TRUE);
                 rv = mht_gpio_set_bit(dev_id, pin, arg);
                 break;
--- a/src/init/ssdk_plat.c
+++ b/src/init/ssdk_plat.c
@@ -307,8 +307,8 @@ sw_error_t qca_bitbang_mii_raw_read_try(
 	if (strncmp(bus->id, "gpio", strlen("gpio")))
 		return SW_NOT_FOUND;
 
-	if (gpio_priv && gpio_priv->sw_read) {
-		*val = gpio_priv->sw_read(bus, reg);
+	if (gpio_priv) {
+		*val = mdiobb_read_c22(bus, reg >> 5, reg & 0x1f);
 		return SW_OK;
 	}
 
@@ -323,8 +323,7 @@ sw_error_t qca_bitbang_mii_raw_write_try
 	if (strncmp(bus->id, "gpio", strlen("gpio")))
 		return SW_NOT_FOUND;
 
-	if (gpio_priv && gpio_priv->sw_write) {
-		gpio_priv->sw_write(bus, reg, val);
+	if (gpio_priv && !mdiobb_write_c22(bus, reg >> 5, reg & 0x1f, val)) {
 		return SW_OK;
 	}
 
@@ -339,13 +338,13 @@ sw_error_t qca_bitbang_mii_raw_update_tr
 	if (strncmp(bus->id, "gpio", strlen("gpio")))
 		return SW_NOT_FOUND;
 
-	if (gpio_priv && gpio_priv->sw_read && gpio_priv->sw_write) {
+	if (gpio_priv) {
 		a_uint32_t val;
 
-		val = gpio_priv->sw_read(bus, reg);
+		val = mdiobb_read_c22(bus, reg >> 5, reg & 0x1f);
 		val &= ~clear;
 		val |= set;
-		gpio_priv->sw_write(bus, reg, val);
+		mdiobb_write_c22(bus, reg >> 5, reg & 0x1f, val);
 
 		return SW_OK;
 	}
