--- a/src/init/ssdk_init.c
+++ b/src/init/ssdk_init.c
@@ -1006,11 +1006,9 @@ qca_mac_sw_sync_work_task(struct work_st
 
 	mutex_lock(&priv->mac_sw_sync_lock);
 
-	 if((p_adpt_api = adpt_api_ptr_get(priv->device_id)) != NULL) {
-		if (NULL == p_adpt_api->adpt_port_polling_sw_sync_set)
-			return;
+	p_adpt_api = adpt_api_ptr_get(priv->device_id);
+	if (p_adpt_api && p_adpt_api->adpt_port_polling_sw_sync_set)
 		p_adpt_api->adpt_port_polling_sw_sync_set(priv);
-	}
 
 	mutex_unlock(&priv->mac_sw_sync_lock);
 
