--- a/src/adpt/hppe/adpt_hppe_portctrl.c
+++ b/src/adpt/hppe/adpt_hppe_portctrl.c
@@ -102,6 +102,31 @@
  */
 #define GMAC_TX_THD    0x1
 
+static a_bool_t
+_adpt_hppe_port_phy_connected(a_uint32_t dev_id, fal_port_t port_id)
+{
+	a_bool_t force_port = 0;
+
+	if (dev_id >= SW_MAX_NR_DEV || port_id >= SW_MAX_NR_PORT)
+		return A_FALSE;
+
+	force_port = hsl_port_feature_get(dev_id, port_id, PHY_F_FORCE);
+	if (force_port == A_TRUE) {
+		SSDK_DEBUG("port_id %d is a force port!\n", port_id);
+		return A_FALSE;
+	}
+
+	if (A_TRUE == hsl_port_is_sfp(dev_id, port_id)) {
+		SSDK_DEBUG("port_id %d is a SFP port!\n", port_id);
+		return A_FALSE;
+	}
+
+	if ((SSDK_PHYSICAL_PORT0 == port_id) || (SSDK_PHYSICAL_PORT7 == port_id))
+		return A_FALSE;
+
+	return hsl_port_prop_check(dev_id, port_id, HSL_PP_PHY);
+}
+
 static sw_error_t
 _adpt_phy_status_get_from_ppe(a_uint32_t dev_id, a_uint32_t port_id,
 		struct port_phy_status *phy_status)
@@ -734,7 +759,7 @@ adpt_hppe_port_link_status_get(a_uint32_
 	}
 
 	/* for those ports without PHY device should be sfp port */
-	if (A_FALSE == hsl_port_phy_connected (dev_id, port_id))
+	if (A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id))
 	{
 		rv = _adpt_phy_status_get_from_ppe(dev_id,
 			port_id, &phy_status);
@@ -1013,7 +1038,7 @@ adpt_hppe_ports_link_status_get(a_uint32
 	for (port_id = 0; port_id < SW_MAX_NR_PORT; port_id++)
 	{
 		/* for those ports without PHY device should be sfp port */
-		if (A_FALSE == hsl_port_phy_connected(dev_id, port_id))
+		if (A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id))
 		{
 			if (hsl_port_prop_check(dev_id, port_id, HSL_PP_CPU) ||
 				hsl_port_prop_check(dev_id, port_id, HSL_PP_INNER))
@@ -1163,7 +1188,7 @@ adpt_hppe_port_duplex_get(a_uint32_t dev
 	}
 
 	/* for those ports without PHY device should be sfp port */
-	if (A_FALSE == hsl_port_phy_connected (dev_id, port_id))
+	if (A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id))
 	{
 
 		rv = _adpt_phy_status_get_from_ppe(dev_id,
@@ -1215,7 +1240,7 @@ adpt_hppe_port_combo_prefer_medium_get(a
 		return SW_BAD_PARAM;
 	}
 
-	if (A_FALSE == hsl_port_phy_connected (dev_id, port_id) &&
+	if (A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id) &&
 		A_FALSE == hsl_port_is_sfp(dev_id, port_id))
 	{
 		return SW_NOT_SUPPORTED;
@@ -1287,7 +1312,7 @@ adpt_hppe_port_flowctrl_forcemode_set(a_
 
 	if ((port_id < SSDK_PHYSICAL_PORT1) || (port_id > SSDK_PHYSICAL_PORT6))
 		return SW_BAD_VALUE;
-	if(!hsl_port_phy_connected(dev_id, port_id) && !enable)
+	if(!_adpt_hppe_port_phy_connected(dev_id, port_id) && !enable)
 	{
 		return SW_NOT_SUPPORTED;
 	}
@@ -1367,7 +1392,7 @@ adpt_hppe_port_txfc_status_set(a_uint32_
 	sw_error_t rv = SW_OK;
 	a_bool_t force_mode = A_FALSE;
 
-	if(A_FALSE == hsl_port_phy_connected(dev_id, port_id))
+	if(A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id))
 	{
 		rv = _adpt_hppe_port_txfc_status_set(dev_id, port_id, enable);
 		SW_RTN_ON_ERROR(rv);
@@ -1451,7 +1476,7 @@ adpt_hppe_port_rxfc_status_set(a_uint32_
 	a_bool_t force_mode = A_FALSE;
 	a_uint32_t phy_addr = 0;
 
-	if(A_FALSE == hsl_port_phy_connected(dev_id, port_id))
+	if(A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id))
 	{
 		rv = _adpt_hppe_port_rxfc_status_set(dev_id, port_id, enable);
 		SW_RTN_ON_ERROR(rv);
@@ -1522,7 +1547,7 @@ _adpt_hppe_port_combo_prefer_medium_set(
 	if (rv == SW_OK)
 	{
 		/*init port status to triger polling*/
-		if(hsl_port_phy_connected(dev_id, port_id) == A_TRUE)
+		if(_adpt_hppe_port_phy_connected(dev_id, port_id) == A_TRUE)
 		{
 			_adpt_hppe_port_txfc_status_set(dev_id, port_id, A_FALSE);
 			_adpt_hppe_port_rxfc_status_set(dev_id, port_id, A_FALSE);
@@ -1549,7 +1574,7 @@ adpt_hppe_port_combo_prefer_medium_set(a
 		return SW_BAD_PARAM;
 	}
 
-	if (A_FALSE == hsl_port_phy_connected (dev_id, port_id) &&
+	if (A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id) &&
 		A_FALSE == hsl_port_is_sfp(dev_id, port_id))
 	{
 		return SW_NOT_SUPPORTED;
@@ -2865,7 +2890,7 @@ adpt_hppe_port_speed_get(a_uint32_t dev_
 	}
 
 	/* for those ports without PHY device should be sfp port */
-	if (A_FALSE == hsl_port_phy_connected (dev_id, port_id))
+	if (A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id))
 	{
 		if (port_id == SSDK_PHYSICAL_PORT0)
 			*pspeed = FAL_SPEED_10000;
@@ -3692,7 +3717,7 @@ adpt_hppe_port_phy_status_get(a_uint32_t
 	ADPT_NULL_POINT_CHECK(phy_status);
 
 	/* for those ports without PHY device should be sfp port or a internal port*/
-	if (A_FALSE == hsl_port_phy_connected (dev_id, port_id)) {
+	if (A_FALSE == _adpt_hppe_port_phy_connected(dev_id, port_id)) {
 		if (port_id != SSDK_PHYSICAL_PORT0) {
 			fal_port_interface_mode_t mode = PORT_INTERFACE_MODE_MAX;
 			rv = adpt_hppe_port_interface_mode_get(dev_id, port_id,
@@ -4351,7 +4376,7 @@ adpt_hppe_phy_interface_mode_switch(a_ui
 {
 	sw_error_t rv = SW_OK;
 
-	if (A_TRUE == hsl_port_phy_connected(dev_id, port_id)) {
+	if (A_TRUE == _adpt_hppe_port_phy_connected(dev_id, port_id)) {
 		SSDK_DEBUG("phy port %d change interface mode!\n", port_id);
 		rv = adpt_hppe_port_interface_mode_switch(dev_id, port_id);
 		SW_RTN_ON_ERROR(rv);
