--- a/src/hsl/phy/hsl_phy.c
+++ b/src/hsl/phy/hsl_phy.c
@@ -962,9 +962,6 @@ hsl_port_phydev_get_status(a_uint32_t de
 	}
 
 	mutex_lock(&phydev->lock);
-	/* make sure that ssdk can get the phy realtime status so */
-	/* need to read phy status manually */
-	phy_read_status(phydev);
 	phy_status->link_status = phydev->link;
 	phy_status->speed = phydev->speed;
 	phy_status->duplex = phydev->duplex;
