--- a/src/init/ssdk_dts.c
+++ b/src/init/ssdk_dts.c
@@ -310,7 +310,8 @@ ssdk_netdev_switch_t *ssdk_dts_netdev_sw
 	a_uint32_t index = 0;
 
 	for(index = 0; index < SSDK_NETDEV_SWITCH_NUM; index++) {
-		if(ssdk_dt_global.netdev_switch[index].switch_netdev_port == port_id)
+		if(ssdk_dt_global.netdev_switch[index].switch_connected == A_TRUE &&
+			ssdk_dt_global.netdev_switch[index].switch_netdev_port == port_id)
 			return &ssdk_dt_global.netdev_switch[index];
 	}
 
