--- a/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
@@ -714,7 +714,7 @@ static void edma_configure_rings(struct
  * edma_hw_reset()
  *	Reset EDMA Hardware during initialization
  */
-int edma_hw_reset(struct edma_hw *ehw)
+static int edma_hw_reset(struct edma_hw *ehw)
 {
 	struct reset_control *rst;
 	struct platform_device *pdev = ehw->pdev;
--- a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
@@ -279,7 +279,7 @@ int edma_init(void)
 	/*
 	 * Remap register resource
 	 */
-	edma_hw.reg_base = ioremap_nocache((edma_hw.reg_resource)->start,
+	edma_hw.reg_base = ioremap((edma_hw.reg_resource)->start,
 				resource_size(edma_hw.reg_resource));
 	if (!edma_hw.reg_base) {
 		pr_warn("Unable to remap EDMA register memory.\n");
@@ -844,8 +844,7 @@ static int edma_register_netdevice(struc
 	 * NAPI add
 	 */
 	if (!edma_hw.napi_added) {
-		netif_napi_add(netdev, &edma_hw.napi, edma_napi,
-				EDMA_NAPI_WORK);
+		netif_napi_add(netdev, &edma_hw.napi, edma_napi);
 		/*
 		 * Register the interrupt handlers and enable interrupts
 		 */
--- a/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
@@ -227,9 +227,8 @@ void nss_phy_tstamp_rx_buf(__attribute__
 	 * set to the correct PTP class value by calling ptp_classify_raw
 	 * in drv->rxtstamp function.
 	 */
-	if (ndev && ndev->phydev && ndev->phydev->drv &&
-			ndev->phydev->drv->rxtstamp)
-		if(ndev->phydev->drv->rxtstamp(ndev->phydev, skb, 0))
+	if (ndev && ndev->phydev && phy_has_rxtstamp(ndev->phydev))
+		if (phy_rxtstamp(ndev->phydev, skb, 0))
 			return;
 
 	netif_receive_skb(skb);
@@ -248,9 +247,8 @@ void nss_phy_tstamp_tx_buf(struct net_de
 	 * set to the correct PTP class value by calling ptp_classify_raw
 	 * in the drv->txtstamp function.
 	 */
-	if (ndev && ndev->phydev && ndev->phydev->drv &&
-			ndev->phydev->drv->txtstamp)
-		ndev->phydev->drv->txtstamp(ndev->phydev, skb, 0);
+	if (ndev && ndev->phydev && phy_has_txtstamp(ndev->phydev))
+		phy_txtstamp(ndev->phydev, skb, 0);
 }
 EXPORT_SYMBOL(nss_phy_tstamp_tx_buf);
 
