--- a/nss_dp_ethtools.c
+++ b/nss_dp_ethtools.c
@@ -248,12 +248,15 @@ static inline void nss_dp_fal_to_ethtool
  * nss_dp_get_eee()
  *	Get EEE settings.
  */
-static int32_t nss_dp_get_eee(struct net_device *netdev, struct ethtool_eee *eee)
+static int32_t nss_dp_get_eee(struct net_device *netdev, struct ethtool_keee *eee)
 {
 	struct nss_dp_dev *dp_priv = (struct nss_dp_dev *)netdev_priv(netdev);
 	fal_port_eee_cfg_t port_eee_cfg;
 	uint32_t port_id;
 	sw_error_t ret;
+	u32 supported = 0;
+	u32 advertised = 0;
+	u32 lp_advertised = 0;
 
 	memset(&port_eee_cfg, 0, sizeof(fal_port_eee_cfg_t));
 	port_id = dp_priv->macid;
@@ -266,9 +269,12 @@ static int32_t nss_dp_get_eee(struct net
 	/*
 	 * Translate the FAL linkmode types to ethtool linkmode types.
 	 */
-	nss_dp_fal_to_ethtool_linkmode_xlate(&eee->supported, &port_eee_cfg.capability);
-	nss_dp_fal_to_ethtool_linkmode_xlate(&eee->advertised, &port_eee_cfg.advertisement);
-	nss_dp_fal_to_ethtool_linkmode_xlate(&eee->lp_advertised, &port_eee_cfg.link_partner_advertisement);
+	nss_dp_fal_to_ethtool_linkmode_xlate(&supported, &port_eee_cfg.capability);
+	nss_dp_fal_to_ethtool_linkmode_xlate(&advertised, &port_eee_cfg.advertisement);
+	nss_dp_fal_to_ethtool_linkmode_xlate(&lp_advertised, &port_eee_cfg.link_partner_advertisement);
+	ethtool_convert_legacy_u32_to_link_mode(eee->supported, supported);
+	ethtool_convert_legacy_u32_to_link_mode(eee->advertised, advertised);
+	ethtool_convert_legacy_u32_to_link_mode(eee->lp_advertised, lp_advertised);
 	eee->eee_enabled = port_eee_cfg.enable;
 	eee->eee_active = port_eee_cfg.eee_status;
 	eee->tx_lpi_enabled = port_eee_cfg.lpi_tx_enable;
@@ -281,12 +287,13 @@ static int32_t nss_dp_get_eee(struct net
  * nss_dp_set_eee()
  *	Set EEE settings.
  */
-static int32_t nss_dp_set_eee(struct net_device *netdev, struct ethtool_eee *eee)
+static int32_t nss_dp_set_eee(struct net_device *netdev, struct ethtool_keee *eee)
 {
 	struct nss_dp_dev *dp_priv = (struct nss_dp_dev *)netdev_priv(netdev);
 	fal_port_eee_cfg_t port_eee_cfg, port_eee_cur_cfg;
 	uint32_t port_id, pos;
 	sw_error_t ret;
+	u32 advertised = 0;
 
 	memset(&port_eee_cfg, 0, sizeof(fal_port_eee_cfg_t));
 	memset(&port_eee_cur_cfg, 0, sizeof(fal_port_eee_cfg_t));
@@ -306,8 +313,10 @@ static int32_t nss_dp_set_eee(struct net
 	/*
 	 * Translate the ethtool speed types to FAL speed types.
 	 */
-	while (eee->advertised) {
-		pos = ffs(eee->advertised);
+	if (!ethtool_convert_link_mode_to_legacy_u32(&advertised, eee->advertised))
+		return -EIO;
+	while (advertised) {
+		pos = ffs(advertised);
 		switch (1 << (pos - 1)) {
 		case ADVERTISED_10baseT_Full:
 			if (port_eee_cur_cfg.capability & FAL_PHY_EEE_10BASE_T) {
@@ -359,7 +368,7 @@ static int32_t nss_dp_set_eee(struct net
 			return -EIO;
 		}
 
-		eee->advertised &= (~(1 << (pos - 1)));
+		advertised &= (~(1 << (pos - 1)));
 	}
 
 	port_eee_cfg.lpi_tx_enable = eee->tx_lpi_enabled;
