--- a/nss_dp_main.c
+++ b/nss_dp_main.c
@@ -35,13 +35,17 @@
 #include <fal/fal_port_ctrl.h>
 #endif
 #include <net/switchdev.h>
-#if (LINUX_VERSION_CODE > KERNEL_VERSION(6, 6, 0))
 #include <net/netdev_rx_queue.h>
+#ifdef CONFIG_RFS_ACCEL
+#include <linux/hash.h>
+#include <net/hotdata.h>
+#include <net/rps.h>
 #endif
 #if defined(NSS_DP_MAC_POLL_SUPPORT)
 #include <init/ssdk_init.h>
 #endif
 #include "nss_dp_hal.h"
+#include "nss_dp_ppeds.h"
 
 #define JUMBO_MRU_3K 3072
 
@@ -550,17 +554,17 @@ static int nss_dp_rx_flow_steer(struct n
 	}
 
 	hash = skb_get_hash(skb);
-	rxflow = &flow_table->flows[hash & flow_table->mask];
+	rxflow = &flow_table->flows[hash_32(hash, flow_table->log)];
 	rxcpu = (uint32_t)rxflow->cpu;
 
-	sock_flow_table = rcu_dereference(rps_sock_flow_table);
+	sock_flow_table = rcu_dereference(net_hotdata.rps_sock_flow_table);
 	if (!sock_flow_table) {
 		netdev_dbg(netdev, "Global RPS flow table not found\n");
 		return -EINVAL;
 	}
 
 	rfscpu = sock_flow_table->ents[hash & sock_flow_table->mask];
-	rfscpu &= rps_cpu_mask;
+	rfscpu &= net_hotdata.rps_cpu_mask;
 
 	if (rxcpu == rfscpu)
 		return 0;
@@ -911,7 +915,7 @@ bool nss_dp_is_phy_dev(struct net_device
 /*
  * nss_dp_adjust_link()
  */
-void nss_dp_adjust_link(struct net_device *netdev)
+static void nss_dp_adjust_link(struct net_device *netdev)
 {
 	struct nss_dp_dev *dp_priv = netdev_priv(netdev);
 	int current_state = dp_priv->link_state;
@@ -1130,7 +1134,7 @@ fail:
  * Note: We only remove the physical ports here. Virtual
  * port devices are removed explicitly by the VP module.
  */
-static int nss_dp_remove(struct platform_device *pdev)
+static void nss_dp_remove(struct platform_device *pdev)
 {
 	uint32_t i;
 	struct nss_dp_dev *dp_priv;
@@ -1176,8 +1180,6 @@ static int nss_dp_remove(struct platform
 		free_netdev(dp_priv->netdev);
 		dp_global_ctx.nss_dp[i] = NULL;
 	}
-
-	return 0;
 }
 
 static struct of_device_id nss_dp_dt_ids[] = {
@@ -1260,7 +1262,7 @@ EXPORT_SYMBOL(nss_dp_nsm_sawf_sc_stats_r
 /*
  * nss_dp_init()
  */
-int __init nss_dp_init(void)
+static int __init nss_dp_init(void)
 {
 	int ret, i;
 
@@ -1348,7 +1350,7 @@ int __init nss_dp_init(void)
 /*
  * nss_dp_exit()
  */
-void __exit nss_dp_exit(void)
+static void __exit nss_dp_exit(void)
 {
 	/*
 	 * Ensure netdev remove is done before HAL cleanup.
@@ -1366,3 +1368,4 @@ module_exit(nss_dp_exit);
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_DESCRIPTION("NSS Data Plane Network Driver");
+MODULE_INFO(description, "NSS Data Plane Network Driver");
