--- a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
@@ -279,7 +279,7 @@ int edma_init(void)
 	/*
 	 * Remap register resource
 	 */
-	edma_hw.reg_base = ioremap_nocache((edma_hw.reg_resource)->start,
+	edma_hw.reg_base = ioremap((edma_hw.reg_resource)->start,
 				resource_size(edma_hw.reg_resource));
 	if (!edma_hw.reg_base) {
 		pr_warn("Unable to remap EDMA register memory.\n");
@@ -844,7 +844,7 @@ static int edma_register_netdevice(struc
 	 * NAPI add
 	 */
 	if (!edma_hw.napi_added) {
-		netif_napi_add(netdev, &edma_hw.napi, edma_napi,
+		netif_napi_add_weight(netdev, &edma_hw.napi, edma_napi,
 				EDMA_NAPI_WORK);
 		/*
 		 * Register the interrupt handlers and enable interrupts
