--- a/nss_dp_main.c
+++ b/nss_dp_main.c
@@ -17,6 +17,7 @@
  */
 
 #include <linux/kernel.h>
+#include <linux/err.h>
 #include <linux/module.h>
 #include <linux/version.h>
 #include <linux/of.h>
@@ -499,7 +500,7 @@ static int nss_dp_open(struct net_device
 
 	netif_start_queue(netdev);
 
-	if (!dp_priv->link_poll) {
+	if (!dp_priv->phydev) {
 		/* Notify data plane link is up */
 		if (dp_priv->data_plane_ops->link_state(dp_priv->dpc, 1)) {
 			netdev_dbg(netdev, "Data plane set link failed\n");
@@ -951,6 +952,20 @@ static void nss_dp_adjust_link(struct ne
 	}
 }
 
+static void nss_dp_phy_disconnect(struct nss_dp_dev *dp_priv)
+{
+	struct device_node *np = dp_priv->pdev->dev.of_node;
+
+	if (!dp_priv->phydev)
+		return;
+
+	phy_disconnect(dp_priv->phydev);
+	dp_priv->phydev = NULL;
+
+	if (of_phy_is_fixed_link(np))
+		of_phy_deregister_fixed_link(np);
+}
+
 /*
  * nss_dp_probe()
  */
@@ -960,19 +975,28 @@ static int32_t nss_dp_probe(struct platf
 	struct nss_dp_dev *dp_priv;
 	struct device_node *np = pdev->dev.of_node;
 	struct nss_gmac_hal_platform_data gmac_hal_pdata;
+	const char *name = of_get_property(np, "label", NULL);
 	int32_t ret = 0;
 	uint8_t phy_id[MII_BUS_ID_SIZE + 3];
+	int assign_type;
 #if defined(NSS_DP_PPE_SUPPORT)
 	uint32_t vsi_id;
 	fal_port_t port_id;
 #endif
 
+	if (name) {
+		assign_type = NET_NAME_PREDICTABLE;
+	} else {
+		name = "eth%d";
+		assign_type = NET_NAME_ENUM;
+	}
+
 	/* TODO: See if we need to do some SoC level common init */
 
-	netdev = alloc_etherdev_mqs(sizeof(struct nss_dp_dev),
-			NSS_DP_NETDEV_TX_QUEUE_NUM, NSS_DP_NETDEV_RX_QUEUE_NUM);
+	netdev = alloc_netdev_mqs(sizeof(struct nss_dp_dev), name, assign_type,
+			ether_setup, NSS_DP_NETDEV_TX_QUEUE_NUM, NSS_DP_NETDEV_RX_QUEUE_NUM);
 	if (!netdev) {
-		pr_info("alloc_etherdev() failed\n");
+		dev_err(&pdev->dev, "alloc_netdev_mqs() failed\n");
 		return -ENOMEM;
 	}
 
@@ -1042,7 +1066,19 @@ static int32_t nss_dp_probe(struct platf
 
 	dp_priv->drv_flags |= NSS_DP_PRIV_FLAG(INIT_DONE);
 
-	if (dp_priv->link_poll) {
+	if (of_find_property(np, "phy-handle", NULL) || of_phy_is_fixed_link(np)) {
+		dp_priv->phydev = of_phy_get_and_connect(netdev, np,
+				&nss_dp_adjust_link);
+		if (IS_ERR_OR_NULL(dp_priv->phydev)) {
+			dp_priv->phydev = NULL;
+			netdev_err(netdev, "failed to connect to phy device\n");
+			if (of_phy_is_fixed_link(np))
+				of_phy_deregister_fixed_link(np);
+			goto phy_setup_fail;
+		}
+
+		phy_attached_info(dp_priv->phydev);
+	} else if (dp_priv->link_poll) {
 		dp_priv->miibus = nss_dp_mdio_attach(pdev);
 		if (!dp_priv->miibus) {
 			netdev_dbg(netdev, "failed to find miibus\n");
@@ -1121,6 +1157,7 @@ static int32_t nss_dp_probe(struct platf
 	return 0;
 
 phy_setup_fail:
+	nss_dp_phy_disconnect(dp_priv);
 	dp_priv->data_plane_ops->deinit(dp_priv->dpc);
 data_plane_init_fail:
 	dp_priv->gmac_hal_ops->exit(dp_priv->gmac_hal_ctx);
@@ -1151,10 +1188,7 @@ static void nss_dp_remove(struct platfor
 		dp_ops = dp_priv->data_plane_ops;
 		hal_ops = dp_priv->gmac_hal_ops;
 
-		if (dp_priv->phydev) {
-			phy_disconnect(dp_priv->phydev);
-			dp_priv->phydev = NULL;
-		}
+		nss_dp_phy_disconnect(dp_priv);
 
 #if defined(NSS_DP_PPE_SUPPORT)
 		/*
