--- a/arch/arm64/boot/dts/qcom/ipq6018-cp01-c1.dts
+++ b/arch/arm64/boot/dts/qcom/ipq6018-cp01-c1.dts
@@ -7,7 +7,7 @@
 
 /dts-v1/;
 
-#include "ipq6018-mp5496.dtsi"
+#include "ipq6018.dtsi"
 
 / {
 	model = "Qualcomm Technologies, Inc. IPQ6018/AP-CP01-C1";
--- a/arch/arm64/boot/dts/qcom/ipq6018-mp5496.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq6018-mp5496.dtsi
@@ -4,8 +4,6 @@
  * apply to most devices that make use of the IPQ6018 SoC and MP5496 PMIC.
  */
 
-#include "ipq6018.dtsi"
-
 &cpu0 {
 	cpu-supply = <&mp5496_s2>;
 };
@@ -35,6 +33,7 @@
 		mp5496_l2: l2 {
 			regulator-min-microvolt = <1800000>;
 			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
 		};
 	};
 };
