--- a/v2.0/src/hal/ipq60xx/nss_crypto_eip197.c
+++ b/v2.0/src/hal/ipq60xx/nss_crypto_eip197.c
@@ -455,7 +455,7 @@ free:
  * nss_crypto_eip197_ctx_fill()
  *	Fill context record specific information
  */
-int nss_crypto_eip197_ctx_fill(struct nss_crypto_ctx *ctx, struct nss_crypto_session_data *data,
+static int nss_crypto_eip197_ctx_fill(struct nss_crypto_ctx *ctx, struct nss_crypto_session_data *data,
 				struct nss_crypto_cmn_ctx *msg)
 {
 	if (data->algo > NSS_CRYPTO_CMN_ALGO_MAX)
@@ -477,7 +477,7 @@ int nss_crypto_eip197_ctx_fill(struct ns
  * nss_crypto_eip197_engine_init()
  *	allocate & initialize engine
  */
-int nss_crypto_eip197_engine_init(struct platform_device *pdev, struct device_node *np,
+static int nss_crypto_eip197_engine_init(struct platform_device *pdev, struct device_node *np,
 				struct resource *res, uint32_t offset)
 {
 	struct nss_crypto_node *node = platform_get_drvdata(pdev);
@@ -490,7 +490,7 @@ int nss_crypto_eip197_engine_init(struct
 	 * remap the I/O addresses
 	 */
 	paddr = res->start + offset;
-	vaddr = ioremap_nocache(paddr, resource_size(res));
+	vaddr = ioremap(paddr, resource_size(res));
 	if (!vaddr) {
 		nss_crypto_warn("%px: unable to remap crypto_addr(0x%px)\n", node, (void *)paddr);
 		return -EIO;
@@ -545,7 +545,7 @@ int nss_crypto_eip197_engine_init(struct
  * nss_crypto_eip197_node_init()
  *	allocate & initialize eip197 node
  */
-int nss_crypto_eip197_node_init(struct platform_device *pdev, const char *name)
+static int nss_crypto_eip197_node_init(struct platform_device *pdev, const char *name)
 {
 	struct device_node *np = of_node_get(pdev->dev.of_node);
 	struct nss_crypto_node *node;
--- a/v2.0/src/hal/ipq60xx/nss_crypto_eip197.h
+++ b/v2.0/src/hal/ipq60xx/nss_crypto_eip197.h
@@ -19,7 +19,9 @@
 #ifndef __NSS_CRYPTO_EIP197_H
 #define __NSS_CRYPTO_EIP197_H
 
+#include <linux/of_platform.h>
 #include <linux/types.h>
+#include <linux/fs.h>
 
 /*
  * Common configuration data for command and result
--- a/v2.0/src/hal/ipq60xx/nss_crypto_eip197_init.c
+++ b/v2.0/src/hal/ipq60xx/nss_crypto_eip197_init.c
@@ -437,7 +437,7 @@ static void nss_crypto_eip197_hw_setup_c
  * nss_crypto_eip197_hw_setup_cache()
  *	setup EIP197 flow and transform cache
  */
-void nss_crypto_eip197_hw_setup_cache(void __iomem *base_addr)
+static void nss_crypto_eip197_hw_setup_cache(void __iomem *base_addr)
 {
 	void __iomem *addr;
 	uint32_t val;
@@ -823,7 +823,7 @@ static void nss_crypto_eip197_hw_disable
  * nss_crypto_eip197_hw_setup()
  *	Pre initialization function for eip197
  */
-void nss_crypto_eip197_hw_setup(void __iomem *base_addr)
+static void nss_crypto_eip197_hw_setup(void __iomem *base_addr)
 {
 	/*
 	 * Reset EIP blocks and check if reset is complete
--- a/v2.0/src/hal/ipq60xx/nss_crypto_hw.c
+++ b/v2.0/src/hal/ipq60xx/nss_crypto_hw.c
@@ -15,8 +15,10 @@
  * PERFORMANCE OF THIS SOFTWARE.
  */
 
+#include <linux/of.h>
 #include <linux/of_platform.h>
 #include "nss_crypto_eip197.h"
+#include "nss_crypto_hw.h"
 
 /*
  * nss_crypto_hw_deinit()
@@ -27,7 +29,6 @@ void nss_crypto_hw_deinit(struct platfor
 	/*
 	 * TODO: Add support for putting HW into reset
 	 */
-	return;
 }
 
 /*
--- a/v2.0/src/hal/ipq807x/nss_crypto_eip197.c
+++ b/v2.0/src/hal/ipq807x/nss_crypto_eip197.c
@@ -455,7 +455,7 @@ free:
  * nss_crypto_eip197_ctx_fill()
  *	Fill context record specific information
  */
-int nss_crypto_eip197_ctx_fill(struct nss_crypto_ctx *ctx, struct nss_crypto_session_data *data,
+static int nss_crypto_eip197_ctx_fill(struct nss_crypto_ctx *ctx, struct nss_crypto_session_data *data,
 				struct nss_crypto_cmn_ctx *msg)
 {
 	if (data->algo > NSS_CRYPTO_CMN_ALGO_MAX)
@@ -477,7 +477,7 @@ int nss_crypto_eip197_ctx_fill(struct ns
  * nss_crypto_eip197_engine_init()
  *	allocate & initialize engine
  */
-int nss_crypto_eip197_engine_init(struct platform_device *pdev, struct device_node *np,
+static int nss_crypto_eip197_engine_init(struct platform_device *pdev, struct device_node *np,
 				struct resource *res, uint32_t offset)
 {
 	struct nss_crypto_node *node = platform_get_drvdata(pdev);
@@ -490,7 +490,7 @@ int nss_crypto_eip197_engine_init(struct
 	 * remap the I/O addresses
 	 */
 	paddr = res->start + offset;
-	vaddr = ioremap_nocache(paddr, resource_size(res));
+	vaddr = ioremap(paddr, resource_size(res));
 	if (!vaddr) {
 		nss_crypto_warn("%px: unable to remap crypto_addr(0x%px)\n", node, (void *)paddr);
 		return -EIO;
@@ -545,7 +545,7 @@ int nss_crypto_eip197_engine_init(struct
  * nss_crypto_eip197_node_init()
  *	allocate & initialize eip197 node
  */
-int nss_crypto_eip197_node_init(struct platform_device *pdev, const char *name)
+static int nss_crypto_eip197_node_init(struct platform_device *pdev, const char *name)
 {
 	struct device_node *np = of_node_get(pdev->dev.of_node);
 	struct nss_crypto_node *node;
--- a/v2.0/src/hal/ipq807x/nss_crypto_eip197.h
+++ b/v2.0/src/hal/ipq807x/nss_crypto_eip197.h
@@ -20,6 +20,8 @@
 #define __NSS_CRYPTO_EIP197_H
 
 #include <linux/types.h>
+#include <linux/fs.h>
+#include <linux/of_platform.h>
 
 /*
  * Common configuration data for command and result
--- a/v2.0/src/hal/ipq807x/nss_crypto_eip197_init.c
+++ b/v2.0/src/hal/ipq807x/nss_crypto_eip197_init.c
@@ -404,7 +404,7 @@ static void nss_crypto_eip197_hw_setup_c
  * nss_crypto_eip197_hw_setup_cache()
  *	setup EIP197 flow and transform cache
  */
-void nss_crypto_eip197_hw_setup_cache(void __iomem *base_addr)
+static void nss_crypto_eip197_hw_setup_cache(void __iomem *base_addr)
 {
 	void __iomem *addr;
 	uint32_t val;
@@ -785,7 +785,7 @@ static void nss_crypto_eip197_hw_disable
  * nss_crypto_eip197_hw_setup()
  *	Pre initialization function for eip197
  */
-void nss_crypto_eip197_hw_setup(void __iomem *base_addr)
+static void nss_crypto_eip197_hw_setup(void __iomem *base_addr)
 {
 	/*
 	 * Reset EIP blocks and check if reset is complete
--- a/v2.0/src/hal/ipq807x/nss_crypto_hw.c
+++ b/v2.0/src/hal/ipq807x/nss_crypto_hw.c
@@ -15,8 +15,10 @@
  * PERFORMANCE OF THIS SOFTWARE.
  */
 
+#include <linux/of.h>
 #include <linux/of_platform.h>
 #include "nss_crypto_eip197.h"
+#include "nss_crypto_hw.h"
 
 /*
  * nss_crypto_hw_deinit()
@@ -27,7 +29,6 @@ void nss_crypto_hw_deinit(struct platfor
 	/*
 	 * TODO: Add support for putting HW into reset
 	 */
-	return;
 }
 
 /*
