--- /dev/null
+++ b/include/qca-nss-phy/nss_phy_ptp.h
@@ -0,0 +1,54 @@
+#ifndef _NSS_PHY_PTP_H_
+#define _NSS_PHY_PTP_H_
+
+struct nss_phy_device;
+
+struct nss_phy_ptp_ops {
+	sw_error_t (*ptp_security_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_link_delay_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rx_crc_recalc_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_tod_uart_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_enhanced_timestamp_engine_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_pps_signal_control_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_timestamp_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_asym_correction_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_time_snapshot_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_capture_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_adjfreq_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_asym_correction_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_pkt_timestamp_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_time_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_time_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_pkt_timestamp_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_interrupt_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_trigger_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_pps_signal_control_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_capture_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rx_crc_recalc_enable)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_security_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_increment_sync_from_clock_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_tod_uart_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_enhanced_timestamp_engine_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_time_clear)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_reference_clock_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_output_waveform_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rx_timestamp_mode_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_grandmaster_mode_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_config_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_trigger_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_adjfreq_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_grandmaster_mode_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rx_timestamp_mode_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_adjtime_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_link_delay_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_increment_sync_from_clock_enable)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_config_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_output_waveform_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_interrupt_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_time_snapshot_enable)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_reference_clock_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_sync_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ptp_rtc_sync_get)(struct nss_phy_device *nss_phydev, ...);
+};
+
+#endif
--- /dev/null
+++ b/include/qca-nss-phy/nss_phy.h
@@ -0,0 +1,78 @@
+#ifndef _NSS_PHY_H_
+#define _NSS_PHY_H_
+
+#include "qca-nss-phy/nss_phy_ptp.h"
+
+#define QCA8072_PHY 0x004dd0b2
+#define QCA8075_PHY 0x004dd0b1
+#define QCA8084_PHY 0x004dd180
+#define MALIBU2PORT_PHY 0x004dd0b0
+#define SERDES_RESET 1
+
+struct phy_device;
+
+struct nss_phy_device {
+	struct phy_device *phydev;
+};
+
+struct nss_phy_ops {
+	struct nss_phy_ptp_ops *ptp_ops;
+	sw_error_t (*intr_mask_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*intr_mask_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*intr_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*function_reset)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*autoadv_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*autoadv_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*autoneg_enable)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*autoneg_restart)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*autoneg_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*cdt)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*combo_fiber_mode_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*combo_fiber_mode_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*combo_medium_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*combo_prefer_medium_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*combo_prefer_medium_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*duplex_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*duplex_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*hibernation_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*hibernation_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ieee_8023az_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ieee_8023az_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*interface_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*link_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*ldo_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*eee_adv_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*eee_adv_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*eee_partner_adv_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*eee_cap_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*eee_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*local_loopback_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*local_loopback_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*magic_frame_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*magic_frame_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*mdix_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*mdix_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*mdix_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*phyid_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*power_off)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*power_on)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*pll_off)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*pll_on)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*powersave_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*powersave_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*remote_loopback_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*remote_loopback_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*reset)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*speed_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*speed_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*stats_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*stats_status_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*stats_status_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*wol_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*wol_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*led_ctrl_source_get)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*led_ctrl_source_set)(struct nss_phy_device *nss_phydev, ...);
+	sw_error_t (*adjust_link_post)(struct nss_phy_device *nss_phydev, ...);
+};
+
+#endif
