--- a/include/qca-nss-phy/nss_phy_ptp.h
+++ b/include/qca-nss-phy/nss_phy_ptp.h
@@ -2,6 +2,23 @@
 #define _NSS_PHY_PTP_H_
 
 struct nss_phy_device;
+typedef a_uint32_t fal_ptp_asym_correction_t;
+typedef a_uint32_t fal_ptp_capture_t;
+typedef a_uint32_t fal_ptp_config_t;
+typedef a_uint32_t fal_ptp_direction_t;
+typedef a_uint32_t fal_ptp_enhanced_ts_engine_t;
+typedef a_uint32_t fal_ptp_grandmaster_mode_t;
+typedef a_uint32_t fal_ptp_interrupt_t;
+typedef a_uint32_t fal_ptp_output_waveform_t;
+typedef a_uint32_t fal_ptp_pkt_info_t;
+typedef a_uint32_t fal_ptp_pps_signal_control_t;
+typedef a_uint32_t fal_ptp_reference_clock_t;
+typedef a_uint32_t fal_ptp_rtc_src_type_t;
+typedef a_uint32_t fal_ptp_rx_timestamp_mode_t;
+typedef a_uint32_t fal_ptp_security_t;
+typedef a_uint32_t fal_ptp_time_t;
+typedef a_uint32_t fal_ptp_tod_uart_t;
+typedef a_uint32_t fal_ptp_trigger_t;
 
 struct nss_phy_ptp_ops {
 	sw_error_t (*ptp_security_set)(struct nss_phy_device *nss_phydev, ...);
