--- a/include/fal/fal_fdb.h
+++ b/include/fal/fal_fdb.h
@@ -271,6 +271,12 @@
     sw_error_t
     fal_fdb_port_learn_static_get(a_uint32_t dev_id, fal_port_t port_id,
                                   a_bool_t * enable);
+
+    sw_error_t
+    fal_fdb_rfs_set(a_uint32_t dev_id, const fal_fdb_rfs_t * entry);
+
+    sw_error_t
+    fal_fdb_rfs_del(a_uint32_t dev_id, const fal_fdb_rfs_t * entry);
 
     sw_error_t
     fal_fdb_port_add(a_uint32_t dev_id, a_uint32_t fid, fal_mac_addr_t * addr, fal_port_t port_id);
