--- a/src/ref/ref_port_ctrl.c
+++ b/src/ref/ref_port_ctrl.c
@@ -52,6 +52,16 @@
 #include <linux/time.h>
 #include "ref_port_ctrl.h"
 
+#ifndef SWITCH_PORT_SPEED_2500
+#define SWITCH_PORT_SPEED_2500 2500
+#endif
+#ifndef SWITCH_PORT_SPEED_5000
+#define SWITCH_PORT_SPEED_5000 5000
+#endif
+#ifndef SWITCH_PORT_SPEED_10000
+#define SWITCH_PORT_SPEED_10000 10000
+#endif
+
 #if defined(CONFIG_OF) && (LINUX_VERSION_CODE >= KERNEL_VERSION(3,14,0))
 #ifdef DESS
 extern struct reset_control *ess_mac_clock_disable[5];
