--- a/src/ref/ref_misc.c
+++ b/src/ref/ref_misc.c
@@ -54,6 +54,11 @@
 #include "hsl_phy.h"
 #endif
 
+#if !defined(IN_SWCONFIG)
+extern struct qca_phy_priv **qca_phy_priv_global;
+#define qca_phy_priv_get(_dev) qca_phy_priv_global[0]
+#endif
+
 int
 qca_ar8327_sw_set_max_frame_size(struct switch_dev *dev,
 										const struct switch_attr *attr,
@@ -180,8 +184,10 @@ qca_ar8327_sw_reset_switch(struct switch
 	mutex_unlock(&priv->reg_mutex);
 
 	priv->init = true;
+#if defined(IN_SWCONFIG)
 	rv += qca_ar8327_sw_hw_apply(dev);
+#endif
 	priv->init = false;
 
 	mac_mode = ssdk_dt_global_get_mac_mode(priv->device_id, 0);
