--- a/src/init/ssdk_plat.c
+++ b/src/init/ssdk_plat.c
@@ -307,8 +307,8 @@ sw_error_t qca_bitbang_mii_raw_read_try(
 	if (strncmp(bus->id, "gpio", strlen("gpio")))
 		return SW_NOT_FOUND;
 
-	if (gpio_priv && gpio_priv->sw_read) {
-		*val = gpio_priv->sw_read(bus, reg);
+	if (gpio_priv) {
+		*val = mdiobb_read_c22(bus, reg >> 5, reg & 0x1f);
 		return SW_OK;
 	}
 
@@ -323,8 +323,7 @@ sw_error_t qca_bitbang_mii_raw_write_try
 	if (strncmp(bus->id, "gpio", strlen("gpio")))
 		return SW_NOT_FOUND;
 
-	if (gpio_priv && gpio_priv->sw_write) {
-		gpio_priv->sw_write(bus, reg, val);
+	if (gpio_priv && !mdiobb_write_c22(bus, reg >> 5, reg & 0x1f, val)) {
 		return SW_OK;
 	}
 
@@ -339,13 +338,13 @@ sw_error_t qca_bitbang_mii_raw_update_tr
 	if (strncmp(bus->id, "gpio", strlen("gpio")))
 		return SW_NOT_FOUND;
 
-	if (gpio_priv && gpio_priv->sw_read && gpio_priv->sw_write) {
+	if (gpio_priv) {
 		a_uint32_t val;
 
-		val = gpio_priv->sw_read(bus, reg);
+		val = mdiobb_read_c22(bus, reg >> 5, reg & 0x1f);
 		val &= ~clear;
 		val |= set;
-		gpio_priv->sw_write(bus, reg, val);
+		mdiobb_write_c22(bus, reg >> 5, reg & 0x1f, val);
 
 		return SW_OK;
 	}
