--- a/src/init/ssdk_init.c
+++ b/src/init/ssdk_init.c
@@ -52,6 +52,7 @@
 #include <linux/gpio.h>
 #include <linux/of_gpio.h>
 #include <net/dsa.h>
+#include <linux/err.h>
 
 #if defined(IN_SWCONFIG)
 #include <linux/switch.h>
@@ -1321,6 +1322,44 @@ static struct notifier_block ssdk_dsa_no
 };
 #endif
 
+#if IS_ENABLED(CONFIG_NET_DSA) && (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+static void ssdk_dsa_netdev_link_refresh(struct net_device *dev)
+{
+	ssdk_netdev_switch_t *netdev_switch = NULL;
+	struct qca_phy_priv* priv = NULL;
+	struct net_device *conduit = NULL;
+	struct dsa_port *dp = NULL;
+
+	netdev_switch = ssdk_dts_netdev_switch_find_by_netdev(dev);
+	if (!netdev_switch) {
+		dp = dsa_port_from_netdev(dev);
+		if (IS_ERR_OR_NULL(dp) || !dp->cpu_dp)
+			return;
+
+		conduit = dsa_port_to_conduit(dp);
+		if (!conduit)
+			return;
+
+		netdev_switch = ssdk_dts_netdev_switch_find_by_netdev(conduit);
+	}
+
+	if (!netdev_switch)
+		return;
+
+	priv = ssdk_phy_priv_data_get(netdev_switch->switch_dev_id);
+	if (!priv)
+		return;
+
+	if (priv->version != QCA_VER_MHT &&
+		priv->version != QCA_VER_AR8337 &&
+		priv->version != QCA_VER_AR8327)
+		return;
+
+	priv->polling_once = A_TRUE;
+	mod_delayed_work(system_wq, &priv->qm_dwork_polling, 0);
+}
+#endif
+
 #if defined(DESS) || defined(HPPE) || defined (ISISC) || defined (ISIS) || defined(MP) || defined(MHT)
 static int ssdk_switch_register(a_uint32_t dev_id, ssdk_chip_type  chip_type)
 {
@@ -1755,8 +1794,16 @@ static int ssdk_dev_event(struct notifie
 				}
 			}
 			break;
-#ifdef MP
+		case NETDEV_UP:
+#if IS_ENABLED(CONFIG_NET_DSA) && (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+			ssdk_dsa_netdev_link_refresh(dev);
+#endif
+			break;
 		case NETDEV_CHANGE:
+#if IS_ENABLED(CONFIG_NET_DSA) && (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+			ssdk_dsa_netdev_link_refresh(dev);
+#endif
+#ifdef MP
 			if ((cfg.chip_type == CHIP_SCOMPHY) &&
 				(cfg.phy_id == MP_GEPHY)) {
 				if ((p_api == NULL) || (p_api->adpt_port_netdev_notify_set == NULL)
@@ -1774,11 +1821,14 @@ static int ssdk_dev_event(struct notifie
 					}
 				}
 			}
-			break;
 #endif
+			break;
 		case NETDEV_REGISTER:
 			if (strstr(dev->name, "eth") && !(dev->priv_flags & IFF_802_1Q_VLAN))
 				ssdk_netdev_switch_init(dev);
+#if IS_ENABLED(CONFIG_NET_DSA) && (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+				ssdk_dsa_netdev_link_refresh(dev);
+#endif
 			break;
 	}
 
